// @aws-amplify/ui-components/dist/collection/common/helpers.js
import { Hub } from '@aws-amplify/core';
import { UI_AUTH_CHANNEL, TOAST_AUTH_ERROR_EVENT, AUTH_STATE_CHANGE_EVENT, PHONE_EMPTY_ERROR_MESSAGE, } from './constants';
import { UsernameAlias } from '../common/types/auth-types';
export const hasShadowDom = (el) => {
    return !!el.shadowRoot && !!el.attachShadow;
};
export const dispatchToastHubEvent = (error) => {
    Hub.dispatch(UI_AUTH_CHANNEL, {
        event: TOAST_AUTH_ERROR_EVENT,
        message: error.message,
    });
};
export const dispatchAuthStateChangeEvent = (nextAuthState, data) => {
    Hub.dispatch(UI_AUTH_CHANNEL, {
        event: AUTH_STATE_CHANGE_EVENT,
        message: nextAuthState,
        data,
    });
};
export const composePhoneNumberInput = (phoneNumber) => {
    if (!phoneNumber.phoneNumberValue) {
        throw new Error(PHONE_EMPTY_ERROR_MESSAGE);
    }
    const sanitizedPhoneNumberValue = phoneNumber.phoneNumberValue.replace(/[-()\s]/g, '');
    return `${phoneNumber.countryDialCodeValue}${sanitizedPhoneNumberValue}`;
};
export const checkUsernameAlias = (usernameAlias) => {
    if (!(usernameAlias in UsernameAlias)) {
        throw new Error(`Invalid username Alias - ${usernameAlias}. Instead use ${Object.values(UsernameAlias)}`);
    }
};
