// @aws-amplify/ui-components/dist/esm/amplify-button_3.entry.js
import { r as registerInstance, h, g as getElement } from './index-04570516.js';
import '@aws-amplify/core';
import './auth-types-78df304e.js';
import './constants-206817b9.js';
import { h as hasShadowDom } from './helpers-5e56336b.js';

const amplifyButtonCss = ":host{--background-color:var(--amplify-primary-color);--background-color-active:var(--amplify-primary-shade);--background-color-disable:var(--amplify-primary-tint);--color:var(--amplify-primary-contrast);--border-width:0;--border-color:initial;--border-style:initial;--link-color:var(--amplify-primary-color);--link-hover:var(--amplify-primary-tint);--link-active:var(--amplify-primary-shade);width:100%;text-align:center}@media (min-width: 672px){:host{width:inherit}}.button{width:100%;min-width:153px;display:inline-block;margin-bottom:0;font-size:var(--amplify-text-xxs);font-weight:400;line-height:1.42857143;text-align:center;white-space:nowrap;vertical-align:middle;-ms-touch-action:manipulation;touch-action:manipulation;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;background-image:none;color:var(--color);text-transform:uppercase;padding:14px;letter-spacing:1.1px;background-color:var(--background-color);border-width:var(--border-width);border-color:var(--border-color);border-style:var(--border-style)}.button:active{opacity:1;background-color:var(--background-color-active)}.button:hover{opacity:0.8}.button:disabled{opacity:0.65;cursor:auto;background-color:var(--background-color-disable)}.anchor{color:var(--link-color);padding:0;border:none;font:inherit;cursor:pointer}.anchor:link{color:var(--link-color);text-decoration:none}.anchor:hover{color:var(--link-hover);text-decoration:underline}.anchor:active{color:var(--link-active);text-decoration:underline}";

const AmplifyButton = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        /** Type of the button: 'button', 'submit' or 'reset' */
        this.type = 'button';
        /** Variant of a button: 'button' | 'anchor' */
        this.variant = 'button';
        /** Disabled state of the button */
        this.disabled = false;
        this.handleClick = (ev) => {
            if (this.handleButtonClick) {
                this.handleButtonClick(ev);
            }
            else if (hasShadowDom(this.el) && this.type == 'submit') {
                // this button wants to specifically submit a form
                // climb up the dom to see if we're in a <form>
                // and if so, then use JS to submit it
                let form = this.el.closest('form');
                if (!form) {
                    // Check for form inside of form section's shadow dom
                    const formSection = this.el.closest('amplify-form-section');
                    form = formSection && formSection.shadowRoot.querySelector('form');
                }
                if (form) {
                    ev.preventDefault();
                    const fakeButton = document.createElement('button');
                    fakeButton.type = this.type;
                    fakeButton.style.display = 'none';
                    form.appendChild(fakeButton);
                    fakeButton.click();
                    fakeButton.remove();
                }
            }
        };
    }
    render() {
        return (h("button", { class: {
                [this.variant]: true,
            }, type: this.type, disabled: this.disabled, onClick: this.handleClick }, h("slot", null)));
    }
    get el() { return getElement(this); }
};
AmplifyButton.style = amplifyButtonCss;

const amplifyLoadingSpinnerCss = ".loading-spinner svg{-webkit-animation:loading-spinner 1s linear infinite;animation:loading-spinner 1s linear infinite;width:0.875rem;height:0.875rem}@-webkit-keyframes loading-spinner{from{-webkit-transform:rotate(0deg);transform:rotate(0deg)}to{-webkit-transform:rotate(360deg);transform:rotate(360deg)}}@keyframes loading-spinner{from{-webkit-transform:rotate(0deg);transform:rotate(0deg)}to{-webkit-transform:rotate(360deg);transform:rotate(360deg)}}";

const AmplifyLoadingSpinner = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
    }
    render() {
        return h("amplify-icon", { class: "loading-spinner", name: "loading" });
    }
};
AmplifyLoadingSpinner.style = amplifyLoadingSpinnerCss;

const amplifySectionCss = ":host{--background-color:var(--amplify-white);--font-family:var(--amplify-font-family)}.section{position:relative;margin-bottom:20px;background-color:var(--background-color);padding:35px 40px;text-align:left;display:inline-block;border-radius:6px;-webkit-box-shadow:1px 1px 4px 0 rgba(0, 0, 0, 0.15);box-shadow:1px 1px 4px 0 rgba(0, 0, 0, 0.15);-webkit-box-sizing:border-box;box-sizing:border-box;font-family:var(--font-family);width:100%;min-width:320px}@media (min-width: 672px){.section{width:auto;min-width:460px}}";

const AmplifySection = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.role = 'application';
    }
    render() {
        return (h("section", { class: "section", role: this.role }, h("slot", null)));
    }
    get el() { return getElement(this); }
};
AmplifySection.style = amplifySectionCss;

export { AmplifyButton as amplify_button, AmplifyLoadingSpinner as amplify_loading_spinner, AmplifySection as amplify_section };
