// @aws-amplify/ui-components/dist/esm/constants-206817b9.js
// Dictionaries
// fieldId constants
const USERNAME_SUFFIX = 'username';
const EMAIL_SUFFIX = 'email';
const CODE_SUFFIX = 'code';
const PHONE_SUFFIX = 'phone';
const PASSWORD_SUFFIX = 'password';
// Country Dial Code common constants
const COUNTRY_DIAL_CODE_SUFFIX = 'country-dial-code-select';
const COUNTRY_DIAL_CODE_DEFAULT = '+1';
// Auth Keys
const AUTH_SOURCE_KEY = 'amplify-auth-source';
const SIGNING_IN_WITH_HOSTEDUI_KEY = 'amplify-signin-with-hostedUI';
const AUTHENTICATOR_AUTHSTATE = 'amplify-authenticator-authState';
// Error message Common Constants
const PHONE_EMPTY_ERROR_MESSAGE = 'Phone number can not be empty';
const NO_AUTH_MODULE_FOUND = 'No Auth module found, please ensure @aws-amplify/auth is imported';
// TOTP Messages
const SETUP_TOTP = 'SETUP_TOTP';
// Select MFA Types Messages
const USER_NOT_SETUP_SOFTWARE_TOKEN_MFA = 'User has not set up software token mfa';
const USER_NOT_VERIFIED_SOFTWARE_TOKEN_MFA = 'User has not verified software token mfa';
// Common events
const SUCCESS = 'SUCCESS';
// Hub Events and Channels
const AUTH_CHANNEL = 'auth';
const UI_AUTH_CHANNEL = 'UI Auth';
const TOAST_AUTH_ERROR_EVENT = 'ToastAuthError';
const AUTH_STATE_CHANGE_EVENT = 'AuthStateChange';

export { AUTH_CHANNEL as A, COUNTRY_DIAL_CODE_DEFAULT as C, EMAIL_SUFFIX as E, NO_AUTH_MODULE_FOUND as N, PHONE_EMPTY_ERROR_MESSAGE as P, SIGNING_IN_WITH_HOSTEDUI_KEY as S, TOAST_AUTH_ERROR_EVENT as T, UI_AUTH_CHANNEL as U, AUTH_STATE_CHANGE_EVENT as a, AUTHENTICATOR_AUTHSTATE as b, USER_NOT_SETUP_SOFTWARE_TOKEN_MFA as c, USER_NOT_VERIFIED_SOFTWARE_TOKEN_MFA as d, COUNTRY_DIAL_CODE_SUFFIX as e, PHONE_SUFFIX as f, SETUP_TOTP as g, SUCCESS as h, AUTH_SOURCE_KEY as i, CODE_SUFFIX as j, PASSWORD_SUFFIX as k, USERNAME_SUFFIX as l };
