// @aws-amplify/ui-components/dist/cjs/amplify-authenticator.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-ba96ffec.js');
const core = require('@aws-amplify/core');
const authTypes = require('./auth-types-cd1f71d2.js');
const auth = require('@aws-amplify/auth');
const constants = require('./constants-c7bbe67d.js');

const logger = new core.Logger('Authenticator');
const AmplifyAuthenticator = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** Initial starting state of the Authenticator component. E.g. If `signup` is passed the default component is set to AmplifySignUp */
        this.initialAuthState = authTypes.AuthState.SignIn;
        this.authState = authTypes.AuthState.Loading;
        this.toastMessage = '';
        this.onAuthStateChange = (nextAuthState, data) => {
            if (nextAuthState === undefined)
                return logger.info('nextAuthState cannot be undefined');
            logger.info('Inside onAuthStateChange Method current authState:', this.authState);
            if (nextAuthState === authTypes.AuthState.SignedOut) {
                this.authState = this.initialAuthState;
            }
            else {
                this.authState = nextAuthState;
            }
            if (data !== undefined) {
                this.authData = data;
                logger.log('Auth Data was set:', this.authData);
            }
            logger.info(`authState has been updated to ${this.authState}`);
        };
    }
    async componentWillLoad() {
        core.Hub.listen(constants.AUTH_CHANNEL, ({ payload: { event, data } }) => {
            switch (event) {
                case 'cognitoHostedUI':
                    return this.onAuthStateChange(authTypes.AuthState.SignedIn, data);
                case 'cognitoHostedUI_failure':
                case 'parsingUrl_failure':
                case 'signOut':
                case 'customGreetingSignOut':
                    return this.onAuthStateChange(authTypes.AuthState.SignIn, null);
            }
        });
        core.Hub.listen(constants.UI_AUTH_CHANNEL, data => {
            const { payload } = data;
            switch (payload.event) {
                case constants.TOAST_AUTH_ERROR_EVENT:
                    if (payload.message)
                        this.toastMessage = payload.message;
                    break;
                case constants.AUTH_STATE_CHANGE_EVENT:
                    if (payload.message) {
                        this.onAuthStateChange(payload.message, payload.data);
                        this.toastMessage = '';
                    }
                    break;
                default:
                    logger.warn('Unhandled Auth Event', payload.event);
            }
        });
        auth.appendToCognitoUserAgent('amplify-authenticator');
        const byHostedUI = localStorage.getItem(constants.SIGNING_IN_WITH_HOSTEDUI_KEY);
        localStorage.removeItem(constants.SIGNING_IN_WITH_HOSTEDUI_KEY);
        if (byHostedUI !== 'true')
            await this.checkUser();
    }
    async checkUser() {
        if (!auth.Auth || typeof auth.Auth.currentAuthenticatedUser !== 'function') {
            throw new Error(constants.NO_AUTH_MODULE_FOUND);
        }
        try {
            const user = await auth.Auth.currentAuthenticatedUser();
            this.onAuthStateChange(authTypes.AuthState.SignedIn, user);
        }
        catch (error) {
            let cachedAuthState = null;
            try {
                cachedAuthState = localStorage.getItem(constants.AUTHENTICATOR_AUTHSTATE);
            }
            catch (error) {
                logger.debug('Failed to get the auth state from local storage', error);
            }
            try {
                if (cachedAuthState === authTypes.AuthState.SignedIn) {
                    await auth.Auth.signOut();
                }
                this.onAuthStateChange(this.initialAuthState);
            }
            catch (error) {
                logger.debug('Failed to sign out', error);
            }
        }
    }
    renderAuthComponent(authState) {
        switch (authState) {
            case authTypes.AuthState.SignIn:
                return (index.h("slot", { name: "sign-in" }, index.h("amplify-sign-in", { federated: this.federated, usernameAlias: this.usernameAlias })));
            case authTypes.AuthState.ConfirmSignIn:
                return (index.h("slot", { name: "confirm-sign-in" }, index.h("amplify-confirm-sign-in", { user: this.authData })));
            case authTypes.AuthState.SignUp:
                return (index.h("slot", { name: "sign-up" }, index.h("amplify-sign-up", { usernameAlias: this.usernameAlias })));
            case authTypes.AuthState.ConfirmSignUp:
                return (index.h("slot", { name: "confirm-sign-up" }, index.h("amplify-confirm-sign-up", { user: this.authData, usernameAlias: this.usernameAlias })));
            case authTypes.AuthState.ForgotPassword:
                return (index.h("slot", { name: "forgot-password" }, index.h("amplify-forgot-password", { usernameAlias: this.usernameAlias })));
            case authTypes.AuthState.ResetPassword:
                return (index.h("slot", { name: "require-new-password" }, index.h("amplify-require-new-password", { user: this.authData })));
            case authTypes.AuthState.VerifyContact:
                return (index.h("slot", { name: "verify-contact" }, index.h("amplify-verify-contact", { user: this.authData })));
            case authTypes.AuthState.TOTPSetup:
                return (index.h("slot", { name: "totp-setup" }, index.h("amplify-totp-setup", { user: this.authData })));
            case authTypes.AuthState.Loading:
                return (index.h("slot", { name: "loading" }, index.h("div", null, "Loading...")));
            case authTypes.AuthState.SignedIn:
                return [index.h("slot", { name: "greetings" }), index.h("slot", null)];
            default:
                throw new Error(`Unhandled auth state: ${authState}`);
        }
    }
    async componentDidUnload() {
        core.Hub.remove(constants.UI_AUTH_CHANNEL, data => {
            const { payload } = data;
            if (payload.event === constants.TOAST_AUTH_ERROR_EVENT && payload.message) {
                this.toastMessage = payload.message;
            }
        });
    }
    render() {
        return (index.h(index.Host, null, this.toastMessage ? (index.h("amplify-toast", { message: this.toastMessage, handleClose: () => {
                this.toastMessage = '';
            }, "data-test": "authenticator-error" })) : null, this.renderAuthComponent(this.authState)));
    }
};

exports.amplify_authenticator = AmplifyAuthenticator;
