// @aws-amplify/ui-components/dist/cjs/amplify-form-field_4.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-ba96ffec.js');

const amplifyFormFieldCss = ":host{--label-font-size:var(--amplify-text-md);--description-font-size:var(--amplify-text-sm)}.form-field{margin-bottom:15px}.form-field-label{display:block;font-size:var(--label-font-size);padding-bottom:0.5em}.form-field-description{font-size:var(--description-font-size);padding-top:0.5em}";

const AmplifyFormField = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** The input type.  Can be any HTML input type. */
        this.type = 'text';
        /** The required flag in order to make an input required prior to submitting a form */
        this.required = false;
        /** (Optional) The placeholder for the input element.  Using hints is recommended, but placeholders can also be useful to convey information to users. */
        this.placeholder = '';
    }
    render() {
        return (index.h("div", { class: "form-field" }, this.label && (index.h("div", { class: "form-field-label" }, index.h("amplify-label", { htmlFor: this.fieldId }, this.label))), this.description && (index.h("div", { id: `${this.fieldId}-description`, class: "form-field-description", "data-test": "form-field-description" }, this.description)), index.h("div", null, index.h("slot", { name: "input" }, index.h("amplify-input", { fieldId: this.fieldId, description: this.description, type: this.type, handleInputChange: this.handleInputChange, placeholder: this.placeholder, name: this.name, value: this.value, inputProps: this.inputProps, disabled: this.disabled }))), this.hint && index.h("amplify-hint", { id: `${this.fieldId}-hint` }, this.hint)));
    }
};
AmplifyFormField.style = amplifyFormFieldCss;

const amplifyHintCss = ":host{--color:var(--amplify-grey);--font-family:var(--amplify-font-family);--font-size:var(--amplify-text-xs)}.hint{color:var(--color);font-family:var(--font-family);font-weight:300;font-size:var(--font-size);margin-bottom:2.625rem}";

const AmplifyHint = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", { class: "hint" }, index.h("slot", null)));
    }
};
AmplifyHint.style = amplifyHintCss;

const amplifyInputCss = ":host{--color:var(--amplify-secondary-color);--background-color:var(--amplify-secondary-contrast);--border-color:var(--amplify-light-grey);--border-color-focus:var(--amplify-primary-color)}.input-host{width:100%}.input{display:block;width:100%;padding:16px;font-size:var(--amplify-text-sm);color:var(--color);background-color:var(--background-color);background-image:none;border:1px solid var(--border-color);border-radius:3px;-webkit-box-sizing:border-box;box-sizing:border-box;margin-bottom:10px}.input:focus{outline:none;border-color:var(--border-color-focus)}.input:disabled{opacity:0.5}";

const AmplifyInput = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** The input type.  Can be any HTML input type. */
        this.type = 'text';
        /** (Optional) The placeholder for the input element.  Using hints is recommended, but placeholders can also be useful to convey information to users. */
        this.placeholder = '';
    }
    render() {
        return (index.h(index.Host, { class: "input-host" }, index.h("input", Object.assign({ id: this.fieldId, "aria-describedby": this.fieldId && this.description ? `${this.fieldId}-description` : null, type: this.type, onChange: this.handleInputChange, placeholder: this.placeholder, name: this.name, class: "input", value: this.value, disabled: this.disabled }, this.inputProps))));
    }
};
AmplifyInput.style = amplifyInputCss;

const amplifyLabelCss = ":host{--label-color:var(--amplify-secondary-color)}.label{color:var(--label-color);font-size:var(--amplify-text-sm);margin-bottom:16px}";

const AmplifyLabel = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("label", { class: "label", htmlFor: this.htmlFor }, index.h("slot", null)));
    }
};
AmplifyLabel.style = amplifyLabelCss;

exports.amplify_form_field = AmplifyFormField;
exports.amplify_hint = AmplifyHint;
exports.amplify_input = AmplifyInput;
exports.amplify_label = AmplifyLabel;
