// @aws-sdk/client-firehose/dist/es/protocols/Aws_json1_1.js
import { __assign, __awaiter, __generator, __values } from "tslib";
import { HttpRequest as __HttpRequest } from "@aws-sdk/protocol-http";
export function serializeAws_json1_1CreateDeliveryStreamCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.CreateDeliveryStream";
            body = JSON.stringify(serializeAws_json1_1CreateDeliveryStreamInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1DeleteDeliveryStreamCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.DeleteDeliveryStream";
            body = JSON.stringify(serializeAws_json1_1DeleteDeliveryStreamInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1DescribeDeliveryStreamCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.DescribeDeliveryStream";
            body = JSON.stringify(serializeAws_json1_1DescribeDeliveryStreamInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1ListDeliveryStreamsCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.ListDeliveryStreams";
            body = JSON.stringify(serializeAws_json1_1ListDeliveryStreamsInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1ListTagsForDeliveryStreamCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.ListTagsForDeliveryStream";
            body = JSON.stringify(serializeAws_json1_1ListTagsForDeliveryStreamInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1PutRecordCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.PutRecord";
            body = JSON.stringify(serializeAws_json1_1PutRecordInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1PutRecordBatchCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.PutRecordBatch";
            body = JSON.stringify(serializeAws_json1_1PutRecordBatchInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1StartDeliveryStreamEncryptionCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.StartDeliveryStreamEncryption";
            body = JSON.stringify(serializeAws_json1_1StartDeliveryStreamEncryptionInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1StopDeliveryStreamEncryptionCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.StopDeliveryStreamEncryption";
            body = JSON.stringify(serializeAws_json1_1StopDeliveryStreamEncryptionInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1TagDeliveryStreamCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.TagDeliveryStream";
            body = JSON.stringify(serializeAws_json1_1TagDeliveryStreamInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1UntagDeliveryStreamCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.UntagDeliveryStream";
            body = JSON.stringify(serializeAws_json1_1UntagDeliveryStreamInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function serializeAws_json1_1UpdateDestinationCommand(input, context) {
    return __awaiter(this, void 0, void 0, function () {
        var headers, body;
        return __generator(this, function (_a) {
            headers = {};
            headers["Content-Type"] = "application/x-amz-json-1.1";
            headers["X-Amz-Target"] = "Firehose_20150804.UpdateDestination";
            body = JSON.stringify(serializeAws_json1_1UpdateDestinationInput(input, context));
            return [2 /*return*/, buildHttpRpcRequest(context, headers, "/", undefined, body)];
        });
    });
}
export function deserializeAws_json1_1CreateDeliveryStreamCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1CreateDeliveryStreamCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1CreateDeliveryStreamOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "CreateDeliveryStreamOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1CreateDeliveryStreamCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, _e, _f, _g, parsedBody, message;
        return __generator(this, function (_h) {
            switch (_h.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _h.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "InvalidArgumentException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#InvalidArgumentException": return [3 /*break*/, 2];
                        case "InvalidKMSResourceException": return [3 /*break*/, 4];
                        case "com.amazonaws.firehose.v20150804#InvalidKMSResourceException": return [3 /*break*/, 4];
                        case "LimitExceededException": return [3 /*break*/, 6];
                        case "com.amazonaws.firehose.v20150804#LimitExceededException": return [3 /*break*/, 6];
                        case "ResourceInUseException": return [3 /*break*/, 8];
                        case "com.amazonaws.firehose.v20150804#ResourceInUseException": return [3 /*break*/, 8];
                    }
                    return [3 /*break*/, 10];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidArgumentExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 4:
                    _e = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidKMSResourceExceptionResponse(parsedOutput, context)];
                case 5:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _e.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 6:
                    _f = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1LimitExceededExceptionResponse(parsedOutput, context)];
                case 7:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _f.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 8:
                    _g = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceInUseExceptionResponse(parsedOutput, context)];
                case 9:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _g.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 10:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _h.label = 11;
                case 11:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1DeleteDeliveryStreamCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1DeleteDeliveryStreamCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1DeleteDeliveryStreamOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "DeleteDeliveryStreamOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1DeleteDeliveryStreamCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, _e, parsedBody, message;
        return __generator(this, function (_f) {
            switch (_f.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _f.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "ResourceInUseException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#ResourceInUseException": return [3 /*break*/, 2];
                        case "ResourceNotFoundException": return [3 /*break*/, 4];
                        case "com.amazonaws.firehose.v20150804#ResourceNotFoundException": return [3 /*break*/, 4];
                    }
                    return [3 /*break*/, 6];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceInUseExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_f.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 7];
                case 4:
                    _e = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceNotFoundExceptionResponse(parsedOutput, context)];
                case 5:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _e.concat([(_f.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 7];
                case 6:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _f.label = 7;
                case 7:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1DescribeDeliveryStreamCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1DescribeDeliveryStreamCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1DescribeDeliveryStreamOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "DescribeDeliveryStreamOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1DescribeDeliveryStreamCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, parsedBody, message;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _e.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "ResourceNotFoundException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#ResourceNotFoundException": return [3 /*break*/, 2];
                    }
                    return [3 /*break*/, 4];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceNotFoundExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_e.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 5];
                case 4:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _e.label = 5;
                case 5:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1ListDeliveryStreamsCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1ListDeliveryStreamsCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1ListDeliveryStreamsOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "ListDeliveryStreamsOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1ListDeliveryStreamsCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, parsedBody, message;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _c.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    switch (errorCode) {
                        default:
                            parsedBody = parsedOutput.body;
                            errorCode = parsedBody.code || parsedBody.Code || errorCode;
                            response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    }
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1ListTagsForDeliveryStreamCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1ListTagsForDeliveryStreamCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1ListTagsForDeliveryStreamOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "ListTagsForDeliveryStreamOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1ListTagsForDeliveryStreamCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, _e, _f, parsedBody, message;
        return __generator(this, function (_g) {
            switch (_g.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _g.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "InvalidArgumentException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#InvalidArgumentException": return [3 /*break*/, 2];
                        case "LimitExceededException": return [3 /*break*/, 4];
                        case "com.amazonaws.firehose.v20150804#LimitExceededException": return [3 /*break*/, 4];
                        case "ResourceNotFoundException": return [3 /*break*/, 6];
                        case "com.amazonaws.firehose.v20150804#ResourceNotFoundException": return [3 /*break*/, 6];
                    }
                    return [3 /*break*/, 8];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidArgumentExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_g.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 9];
                case 4:
                    _e = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1LimitExceededExceptionResponse(parsedOutput, context)];
                case 5:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _e.concat([(_g.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 9];
                case 6:
                    _f = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceNotFoundExceptionResponse(parsedOutput, context)];
                case 7:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _f.concat([(_g.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 9];
                case 8:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _g.label = 9;
                case 9:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1PutRecordCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1PutRecordCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1PutRecordOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "PutRecordOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1PutRecordCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, _e, _f, _g, parsedBody, message;
        return __generator(this, function (_h) {
            switch (_h.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _h.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "InvalidArgumentException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#InvalidArgumentException": return [3 /*break*/, 2];
                        case "InvalidKMSResourceException": return [3 /*break*/, 4];
                        case "com.amazonaws.firehose.v20150804#InvalidKMSResourceException": return [3 /*break*/, 4];
                        case "ResourceNotFoundException": return [3 /*break*/, 6];
                        case "com.amazonaws.firehose.v20150804#ResourceNotFoundException": return [3 /*break*/, 6];
                        case "ServiceUnavailableException": return [3 /*break*/, 8];
                        case "com.amazonaws.firehose.v20150804#ServiceUnavailableException": return [3 /*break*/, 8];
                    }
                    return [3 /*break*/, 10];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidArgumentExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 4:
                    _e = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidKMSResourceExceptionResponse(parsedOutput, context)];
                case 5:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _e.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 6:
                    _f = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceNotFoundExceptionResponse(parsedOutput, context)];
                case 7:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _f.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 8:
                    _g = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ServiceUnavailableExceptionResponse(parsedOutput, context)];
                case 9:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _g.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 10:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _h.label = 11;
                case 11:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1PutRecordBatchCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1PutRecordBatchCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1PutRecordBatchOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "PutRecordBatchOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1PutRecordBatchCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, _e, _f, _g, parsedBody, message;
        return __generator(this, function (_h) {
            switch (_h.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _h.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "InvalidArgumentException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#InvalidArgumentException": return [3 /*break*/, 2];
                        case "InvalidKMSResourceException": return [3 /*break*/, 4];
                        case "com.amazonaws.firehose.v20150804#InvalidKMSResourceException": return [3 /*break*/, 4];
                        case "ResourceNotFoundException": return [3 /*break*/, 6];
                        case "com.amazonaws.firehose.v20150804#ResourceNotFoundException": return [3 /*break*/, 6];
                        case "ServiceUnavailableException": return [3 /*break*/, 8];
                        case "com.amazonaws.firehose.v20150804#ServiceUnavailableException": return [3 /*break*/, 8];
                    }
                    return [3 /*break*/, 10];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidArgumentExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 4:
                    _e = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidKMSResourceExceptionResponse(parsedOutput, context)];
                case 5:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _e.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 6:
                    _f = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceNotFoundExceptionResponse(parsedOutput, context)];
                case 7:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _f.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 8:
                    _g = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ServiceUnavailableExceptionResponse(parsedOutput, context)];
                case 9:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _g.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 10:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _h.label = 11;
                case 11:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1StartDeliveryStreamEncryptionCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1StartDeliveryStreamEncryptionCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1StartDeliveryStreamEncryptionOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "StartDeliveryStreamEncryptionOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1StartDeliveryStreamEncryptionCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, _e, _f, _g, _h, parsedBody, message;
        return __generator(this, function (_j) {
            switch (_j.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _j.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "InvalidArgumentException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#InvalidArgumentException": return [3 /*break*/, 2];
                        case "InvalidKMSResourceException": return [3 /*break*/, 4];
                        case "com.amazonaws.firehose.v20150804#InvalidKMSResourceException": return [3 /*break*/, 4];
                        case "LimitExceededException": return [3 /*break*/, 6];
                        case "com.amazonaws.firehose.v20150804#LimitExceededException": return [3 /*break*/, 6];
                        case "ResourceInUseException": return [3 /*break*/, 8];
                        case "com.amazonaws.firehose.v20150804#ResourceInUseException": return [3 /*break*/, 8];
                        case "ResourceNotFoundException": return [3 /*break*/, 10];
                        case "com.amazonaws.firehose.v20150804#ResourceNotFoundException": return [3 /*break*/, 10];
                    }
                    return [3 /*break*/, 12];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidArgumentExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_j.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 13];
                case 4:
                    _e = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidKMSResourceExceptionResponse(parsedOutput, context)];
                case 5:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _e.concat([(_j.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 13];
                case 6:
                    _f = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1LimitExceededExceptionResponse(parsedOutput, context)];
                case 7:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _f.concat([(_j.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 13];
                case 8:
                    _g = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceInUseExceptionResponse(parsedOutput, context)];
                case 9:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _g.concat([(_j.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 13];
                case 10:
                    _h = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceNotFoundExceptionResponse(parsedOutput, context)];
                case 11:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _h.concat([(_j.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 13];
                case 12:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _j.label = 13;
                case 13:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1StopDeliveryStreamEncryptionCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1StopDeliveryStreamEncryptionCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1StopDeliveryStreamEncryptionOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "StopDeliveryStreamEncryptionOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1StopDeliveryStreamEncryptionCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, _e, _f, _g, parsedBody, message;
        return __generator(this, function (_h) {
            switch (_h.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _h.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "InvalidArgumentException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#InvalidArgumentException": return [3 /*break*/, 2];
                        case "LimitExceededException": return [3 /*break*/, 4];
                        case "com.amazonaws.firehose.v20150804#LimitExceededException": return [3 /*break*/, 4];
                        case "ResourceInUseException": return [3 /*break*/, 6];
                        case "com.amazonaws.firehose.v20150804#ResourceInUseException": return [3 /*break*/, 6];
                        case "ResourceNotFoundException": return [3 /*break*/, 8];
                        case "com.amazonaws.firehose.v20150804#ResourceNotFoundException": return [3 /*break*/, 8];
                    }
                    return [3 /*break*/, 10];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidArgumentExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 4:
                    _e = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1LimitExceededExceptionResponse(parsedOutput, context)];
                case 5:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _e.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 6:
                    _f = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceInUseExceptionResponse(parsedOutput, context)];
                case 7:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _f.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 8:
                    _g = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceNotFoundExceptionResponse(parsedOutput, context)];
                case 9:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _g.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 10:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _h.label = 11;
                case 11:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1TagDeliveryStreamCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1TagDeliveryStreamCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1TagDeliveryStreamOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "TagDeliveryStreamOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1TagDeliveryStreamCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, _e, _f, _g, parsedBody, message;
        return __generator(this, function (_h) {
            switch (_h.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _h.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "InvalidArgumentException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#InvalidArgumentException": return [3 /*break*/, 2];
                        case "LimitExceededException": return [3 /*break*/, 4];
                        case "com.amazonaws.firehose.v20150804#LimitExceededException": return [3 /*break*/, 4];
                        case "ResourceInUseException": return [3 /*break*/, 6];
                        case "com.amazonaws.firehose.v20150804#ResourceInUseException": return [3 /*break*/, 6];
                        case "ResourceNotFoundException": return [3 /*break*/, 8];
                        case "com.amazonaws.firehose.v20150804#ResourceNotFoundException": return [3 /*break*/, 8];
                    }
                    return [3 /*break*/, 10];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidArgumentExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 4:
                    _e = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1LimitExceededExceptionResponse(parsedOutput, context)];
                case 5:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _e.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 6:
                    _f = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceInUseExceptionResponse(parsedOutput, context)];
                case 7:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _f.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 8:
                    _g = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceNotFoundExceptionResponse(parsedOutput, context)];
                case 9:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _g.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 10:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _h.label = 11;
                case 11:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1UntagDeliveryStreamCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1UntagDeliveryStreamCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1UntagDeliveryStreamOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "UntagDeliveryStreamOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1UntagDeliveryStreamCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, _e, _f, _g, parsedBody, message;
        return __generator(this, function (_h) {
            switch (_h.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _h.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "InvalidArgumentException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#InvalidArgumentException": return [3 /*break*/, 2];
                        case "LimitExceededException": return [3 /*break*/, 4];
                        case "com.amazonaws.firehose.v20150804#LimitExceededException": return [3 /*break*/, 4];
                        case "ResourceInUseException": return [3 /*break*/, 6];
                        case "com.amazonaws.firehose.v20150804#ResourceInUseException": return [3 /*break*/, 6];
                        case "ResourceNotFoundException": return [3 /*break*/, 8];
                        case "com.amazonaws.firehose.v20150804#ResourceNotFoundException": return [3 /*break*/, 8];
                    }
                    return [3 /*break*/, 10];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidArgumentExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 4:
                    _e = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1LimitExceededExceptionResponse(parsedOutput, context)];
                case 5:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _e.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 6:
                    _f = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceInUseExceptionResponse(parsedOutput, context)];
                case 7:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _f.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 8:
                    _g = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceNotFoundExceptionResponse(parsedOutput, context)];
                case 9:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _g.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 10:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _h.label = 11;
                case 11:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
export function deserializeAws_json1_1UpdateDestinationCommand(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var data, contents, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (output.statusCode >= 400) {
                        return [2 /*return*/, deserializeAws_json1_1UpdateDestinationCommandError(output, context)];
                    }
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    data = _a.sent();
                    contents = {};
                    contents = deserializeAws_json1_1UpdateDestinationOutput(data, context);
                    response = __assign({ $metadata: deserializeMetadata(output), __type: "UpdateDestinationOutput" }, contents);
                    return [2 /*return*/, Promise.resolve(response)];
            }
        });
    });
}
function deserializeAws_json1_1UpdateDestinationCommandError(output, context) {
    return __awaiter(this, void 0, void 0, function () {
        var parsedOutput, _a, _b, response, errorCode, errorTypeParts, _c, _d, _e, _f, _g, parsedBody, message;
        return __generator(this, function (_h) {
            switch (_h.label) {
                case 0:
                    _a = [__assign({}, output)];
                    _b = {};
                    return [4 /*yield*/, parseBody(output.body, context)];
                case 1:
                    parsedOutput = __assign.apply(void 0, _a.concat([(_b.body = _h.sent(), _b)]));
                    errorCode = "UnknownError";
                    errorTypeParts = parsedOutput.body["__type"].split("#");
                    errorCode =
                        errorTypeParts[1] === undefined ? errorTypeParts[0] : errorTypeParts[1];
                    _c = errorCode;
                    switch (_c) {
                        case "ConcurrentModificationException": return [3 /*break*/, 2];
                        case "com.amazonaws.firehose.v20150804#ConcurrentModificationException": return [3 /*break*/, 2];
                        case "InvalidArgumentException": return [3 /*break*/, 4];
                        case "com.amazonaws.firehose.v20150804#InvalidArgumentException": return [3 /*break*/, 4];
                        case "ResourceInUseException": return [3 /*break*/, 6];
                        case "com.amazonaws.firehose.v20150804#ResourceInUseException": return [3 /*break*/, 6];
                        case "ResourceNotFoundException": return [3 /*break*/, 8];
                        case "com.amazonaws.firehose.v20150804#ResourceNotFoundException": return [3 /*break*/, 8];
                    }
                    return [3 /*break*/, 10];
                case 2:
                    _d = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ConcurrentModificationExceptionResponse(parsedOutput, context)];
                case 3:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _d.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 4:
                    _e = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1InvalidArgumentExceptionResponse(parsedOutput, context)];
                case 5:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _e.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 6:
                    _f = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceInUseExceptionResponse(parsedOutput, context)];
                case 7:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _f.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 8:
                    _g = [{}];
                    return [4 /*yield*/, deserializeAws_json1_1ResourceNotFoundExceptionResponse(parsedOutput, context)];
                case 9:
                    response = __assign.apply(void 0, [__assign.apply(void 0, _g.concat([(_h.sent())])), { name: errorCode, $metadata: deserializeMetadata(output) }]);
                    return [3 /*break*/, 11];
                case 10:
                    parsedBody = parsedOutput.body;
                    errorCode = parsedBody.code || parsedBody.Code || errorCode;
                    response = __assign(__assign({}, parsedBody), { name: "" + errorCode, message: parsedBody.message || parsedBody.Message || errorCode, $fault: "client", $metadata: deserializeMetadata(output) });
                    _h.label = 11;
                case 11:
                    message = response.message || response.Message || errorCode;
                    response.message = message;
                    delete response.Message;
                    return [2 /*return*/, Promise.reject(Object.assign(new Error(message), response))];
            }
        });
    });
}
var deserializeAws_json1_1ConcurrentModificationExceptionResponse = function (parsedOutput, context) { return __awaiter(void 0, void 0, void 0, function () {
    var body, deserialized, contents;
    return __generator(this, function (_a) {
        body = parsedOutput.body;
        deserialized = deserializeAws_json1_1ConcurrentModificationException(body, context);
        contents = __assign({ name: "ConcurrentModificationException", $fault: "client", $metadata: deserializeMetadata(parsedOutput) }, deserialized);
        return [2 /*return*/, contents];
    });
}); };
var deserializeAws_json1_1InvalidArgumentExceptionResponse = function (parsedOutput, context) { return __awaiter(void 0, void 0, void 0, function () {
    var body, deserialized, contents;
    return __generator(this, function (_a) {
        body = parsedOutput.body;
        deserialized = deserializeAws_json1_1InvalidArgumentException(body, context);
        contents = __assign({ name: "InvalidArgumentException", $fault: "client", $metadata: deserializeMetadata(parsedOutput) }, deserialized);
        return [2 /*return*/, contents];
    });
}); };
var deserializeAws_json1_1InvalidKMSResourceExceptionResponse = function (parsedOutput, context) { return __awaiter(void 0, void 0, void 0, function () {
    var body, deserialized, contents;
    return __generator(this, function (_a) {
        body = parsedOutput.body;
        deserialized = deserializeAws_json1_1InvalidKMSResourceException(body, context);
        contents = __assign({ name: "InvalidKMSResourceException", $fault: "client", $metadata: deserializeMetadata(parsedOutput) }, deserialized);
        return [2 /*return*/, contents];
    });
}); };
var deserializeAws_json1_1LimitExceededExceptionResponse = function (parsedOutput, context) { return __awaiter(void 0, void 0, void 0, function () {
    var body, deserialized, contents;
    return __generator(this, function (_a) {
        body = parsedOutput.body;
        deserialized = deserializeAws_json1_1LimitExceededException(body, context);
        contents = __assign({ name: "LimitExceededException", $fault: "client", $metadata: deserializeMetadata(parsedOutput) }, deserialized);
        return [2 /*return*/, contents];
    });
}); };
var deserializeAws_json1_1ResourceInUseExceptionResponse = function (parsedOutput, context) { return __awaiter(void 0, void 0, void 0, function () {
    var body, deserialized, contents;
    return __generator(this, function (_a) {
        body = parsedOutput.body;
        deserialized = deserializeAws_json1_1ResourceInUseException(body, context);
        contents = __assign({ name: "ResourceInUseException", $fault: "client", $metadata: deserializeMetadata(parsedOutput) }, deserialized);
        return [2 /*return*/, contents];
    });
}); };
var deserializeAws_json1_1ResourceNotFoundExceptionResponse = function (parsedOutput, context) { return __awaiter(void 0, void 0, void 0, function () {
    var body, deserialized, contents;
    return __generator(this, function (_a) {
        body = parsedOutput.body;
        deserialized = deserializeAws_json1_1ResourceNotFoundException(body, context);
        contents = __assign({ name: "ResourceNotFoundException", $fault: "client", $metadata: deserializeMetadata(parsedOutput) }, deserialized);
        return [2 /*return*/, contents];
    });
}); };
var deserializeAws_json1_1ServiceUnavailableExceptionResponse = function (parsedOutput, context) { return __awaiter(void 0, void 0, void 0, function () {
    var body, deserialized, contents;
    return __generator(this, function (_a) {
        body = parsedOutput.body;
        deserialized = deserializeAws_json1_1ServiceUnavailableException(body, context);
        contents = __assign({ name: "ServiceUnavailableException", $fault: "server", $metadata: deserializeMetadata(parsedOutput) }, deserialized);
        return [2 /*return*/, contents];
    });
}); };
var serializeAws_json1_1BufferingHints = function (input, context) {
    var bodyParams = {};
    if (input.IntervalInSeconds !== undefined) {
        bodyParams["IntervalInSeconds"] = input.IntervalInSeconds;
    }
    if (input.SizeInMBs !== undefined) {
        bodyParams["SizeInMBs"] = input.SizeInMBs;
    }
    return bodyParams;
};
var serializeAws_json1_1CloudWatchLoggingOptions = function (input, context) {
    var bodyParams = {};
    if (input.Enabled !== undefined) {
        bodyParams["Enabled"] = input.Enabled;
    }
    if (input.LogGroupName !== undefined) {
        bodyParams["LogGroupName"] = input.LogGroupName;
    }
    if (input.LogStreamName !== undefined) {
        bodyParams["LogStreamName"] = input.LogStreamName;
    }
    return bodyParams;
};
var serializeAws_json1_1ColumnToJsonKeyMappings = function (input, context) {
    var mapParams = {};
    Object.keys(input).forEach(function (key) {
        mapParams[key] = input[key];
    });
    return mapParams;
};
var serializeAws_json1_1CopyCommand = function (input, context) {
    var bodyParams = {};
    if (input.CopyOptions !== undefined) {
        bodyParams["CopyOptions"] = input.CopyOptions;
    }
    if (input.DataTableColumns !== undefined) {
        bodyParams["DataTableColumns"] = input.DataTableColumns;
    }
    if (input.DataTableName !== undefined) {
        bodyParams["DataTableName"] = input.DataTableName;
    }
    return bodyParams;
};
var serializeAws_json1_1CreateDeliveryStreamInput = function (input, context) {
    var bodyParams = {};
    if (input.DeliveryStreamEncryptionConfigurationInput !== undefined) {
        bodyParams["DeliveryStreamEncryptionConfigurationInput"] = serializeAws_json1_1DeliveryStreamEncryptionConfigurationInput(input.DeliveryStreamEncryptionConfigurationInput, context);
    }
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    if (input.DeliveryStreamType !== undefined) {
        bodyParams["DeliveryStreamType"] = input.DeliveryStreamType;
    }
    if (input.ElasticsearchDestinationConfiguration !== undefined) {
        bodyParams["ElasticsearchDestinationConfiguration"] = serializeAws_json1_1ElasticsearchDestinationConfiguration(input.ElasticsearchDestinationConfiguration, context);
    }
    if (input.ExtendedS3DestinationConfiguration !== undefined) {
        bodyParams["ExtendedS3DestinationConfiguration"] = serializeAws_json1_1ExtendedS3DestinationConfiguration(input.ExtendedS3DestinationConfiguration, context);
    }
    if (input.KinesisStreamSourceConfiguration !== undefined) {
        bodyParams["KinesisStreamSourceConfiguration"] = serializeAws_json1_1KinesisStreamSourceConfiguration(input.KinesisStreamSourceConfiguration, context);
    }
    if (input.RedshiftDestinationConfiguration !== undefined) {
        bodyParams["RedshiftDestinationConfiguration"] = serializeAws_json1_1RedshiftDestinationConfiguration(input.RedshiftDestinationConfiguration, context);
    }
    if (input.S3DestinationConfiguration !== undefined) {
        bodyParams["S3DestinationConfiguration"] = serializeAws_json1_1S3DestinationConfiguration(input.S3DestinationConfiguration, context);
    }
    if (input.SplunkDestinationConfiguration !== undefined) {
        bodyParams["SplunkDestinationConfiguration"] = serializeAws_json1_1SplunkDestinationConfiguration(input.SplunkDestinationConfiguration, context);
    }
    if (input.Tags !== undefined) {
        bodyParams["Tags"] = serializeAws_json1_1TagDeliveryStreamInputTagList(input.Tags, context);
    }
    return bodyParams;
};
var serializeAws_json1_1DataFormatConversionConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.Enabled !== undefined) {
        bodyParams["Enabled"] = input.Enabled;
    }
    if (input.InputFormatConfiguration !== undefined) {
        bodyParams["InputFormatConfiguration"] = serializeAws_json1_1InputFormatConfiguration(input.InputFormatConfiguration, context);
    }
    if (input.OutputFormatConfiguration !== undefined) {
        bodyParams["OutputFormatConfiguration"] = serializeAws_json1_1OutputFormatConfiguration(input.OutputFormatConfiguration, context);
    }
    if (input.SchemaConfiguration !== undefined) {
        bodyParams["SchemaConfiguration"] = serializeAws_json1_1SchemaConfiguration(input.SchemaConfiguration, context);
    }
    return bodyParams;
};
var serializeAws_json1_1DeleteDeliveryStreamInput = function (input, context) {
    var bodyParams = {};
    if (input.AllowForceDelete !== undefined) {
        bodyParams["AllowForceDelete"] = input.AllowForceDelete;
    }
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    return bodyParams;
};
var serializeAws_json1_1DeliveryStreamEncryptionConfigurationInput = function (input, context) {
    var bodyParams = {};
    if (input.KeyARN !== undefined) {
        bodyParams["KeyARN"] = input.KeyARN;
    }
    if (input.KeyType !== undefined) {
        bodyParams["KeyType"] = input.KeyType;
    }
    return bodyParams;
};
var serializeAws_json1_1DescribeDeliveryStreamInput = function (input, context) {
    var bodyParams = {};
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    if (input.ExclusiveStartDestinationId !== undefined) {
        bodyParams["ExclusiveStartDestinationId"] =
            input.ExclusiveStartDestinationId;
    }
    if (input.Limit !== undefined) {
        bodyParams["Limit"] = input.Limit;
    }
    return bodyParams;
};
var serializeAws_json1_1Deserializer = function (input, context) {
    var bodyParams = {};
    if (input.HiveJsonSerDe !== undefined) {
        bodyParams["HiveJsonSerDe"] = serializeAws_json1_1HiveJsonSerDe(input.HiveJsonSerDe, context);
    }
    if (input.OpenXJsonSerDe !== undefined) {
        bodyParams["OpenXJsonSerDe"] = serializeAws_json1_1OpenXJsonSerDe(input.OpenXJsonSerDe, context);
    }
    return bodyParams;
};
var serializeAws_json1_1ElasticsearchBufferingHints = function (input, context) {
    var bodyParams = {};
    if (input.IntervalInSeconds !== undefined) {
        bodyParams["IntervalInSeconds"] = input.IntervalInSeconds;
    }
    if (input.SizeInMBs !== undefined) {
        bodyParams["SizeInMBs"] = input.SizeInMBs;
    }
    return bodyParams;
};
var serializeAws_json1_1ElasticsearchDestinationConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.BufferingHints !== undefined) {
        bodyParams["BufferingHints"] = serializeAws_json1_1ElasticsearchBufferingHints(input.BufferingHints, context);
    }
    if (input.CloudWatchLoggingOptions !== undefined) {
        bodyParams["CloudWatchLoggingOptions"] = serializeAws_json1_1CloudWatchLoggingOptions(input.CloudWatchLoggingOptions, context);
    }
    if (input.ClusterEndpoint !== undefined) {
        bodyParams["ClusterEndpoint"] = input.ClusterEndpoint;
    }
    if (input.DomainARN !== undefined) {
        bodyParams["DomainARN"] = input.DomainARN;
    }
    if (input.IndexName !== undefined) {
        bodyParams["IndexName"] = input.IndexName;
    }
    if (input.IndexRotationPeriod !== undefined) {
        bodyParams["IndexRotationPeriod"] = input.IndexRotationPeriod;
    }
    if (input.ProcessingConfiguration !== undefined) {
        bodyParams["ProcessingConfiguration"] = serializeAws_json1_1ProcessingConfiguration(input.ProcessingConfiguration, context);
    }
    if (input.RetryOptions !== undefined) {
        bodyParams["RetryOptions"] = serializeAws_json1_1ElasticsearchRetryOptions(input.RetryOptions, context);
    }
    if (input.RoleARN !== undefined) {
        bodyParams["RoleARN"] = input.RoleARN;
    }
    if (input.S3BackupMode !== undefined) {
        bodyParams["S3BackupMode"] = input.S3BackupMode;
    }
    if (input.S3Configuration !== undefined) {
        bodyParams["S3Configuration"] = serializeAws_json1_1S3DestinationConfiguration(input.S3Configuration, context);
    }
    if (input.TypeName !== undefined) {
        bodyParams["TypeName"] = input.TypeName;
    }
    return bodyParams;
};
var serializeAws_json1_1ElasticsearchDestinationUpdate = function (input, context) {
    var bodyParams = {};
    if (input.BufferingHints !== undefined) {
        bodyParams["BufferingHints"] = serializeAws_json1_1ElasticsearchBufferingHints(input.BufferingHints, context);
    }
    if (input.CloudWatchLoggingOptions !== undefined) {
        bodyParams["CloudWatchLoggingOptions"] = serializeAws_json1_1CloudWatchLoggingOptions(input.CloudWatchLoggingOptions, context);
    }
    if (input.ClusterEndpoint !== undefined) {
        bodyParams["ClusterEndpoint"] = input.ClusterEndpoint;
    }
    if (input.DomainARN !== undefined) {
        bodyParams["DomainARN"] = input.DomainARN;
    }
    if (input.IndexName !== undefined) {
        bodyParams["IndexName"] = input.IndexName;
    }
    if (input.IndexRotationPeriod !== undefined) {
        bodyParams["IndexRotationPeriod"] = input.IndexRotationPeriod;
    }
    if (input.ProcessingConfiguration !== undefined) {
        bodyParams["ProcessingConfiguration"] = serializeAws_json1_1ProcessingConfiguration(input.ProcessingConfiguration, context);
    }
    if (input.RetryOptions !== undefined) {
        bodyParams["RetryOptions"] = serializeAws_json1_1ElasticsearchRetryOptions(input.RetryOptions, context);
    }
    if (input.RoleARN !== undefined) {
        bodyParams["RoleARN"] = input.RoleARN;
    }
    if (input.S3Update !== undefined) {
        bodyParams["S3Update"] = serializeAws_json1_1S3DestinationUpdate(input.S3Update, context);
    }
    if (input.TypeName !== undefined) {
        bodyParams["TypeName"] = input.TypeName;
    }
    return bodyParams;
};
var serializeAws_json1_1ElasticsearchRetryOptions = function (input, context) {
    var bodyParams = {};
    if (input.DurationInSeconds !== undefined) {
        bodyParams["DurationInSeconds"] = input.DurationInSeconds;
    }
    return bodyParams;
};
var serializeAws_json1_1EncryptionConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.KMSEncryptionConfig !== undefined) {
        bodyParams["KMSEncryptionConfig"] = serializeAws_json1_1KMSEncryptionConfig(input.KMSEncryptionConfig, context);
    }
    if (input.NoEncryptionConfig !== undefined) {
        bodyParams["NoEncryptionConfig"] = input.NoEncryptionConfig;
    }
    return bodyParams;
};
var serializeAws_json1_1ExtendedS3DestinationConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.BucketARN !== undefined) {
        bodyParams["BucketARN"] = input.BucketARN;
    }
    if (input.BufferingHints !== undefined) {
        bodyParams["BufferingHints"] = serializeAws_json1_1BufferingHints(input.BufferingHints, context);
    }
    if (input.CloudWatchLoggingOptions !== undefined) {
        bodyParams["CloudWatchLoggingOptions"] = serializeAws_json1_1CloudWatchLoggingOptions(input.CloudWatchLoggingOptions, context);
    }
    if (input.CompressionFormat !== undefined) {
        bodyParams["CompressionFormat"] = input.CompressionFormat;
    }
    if (input.DataFormatConversionConfiguration !== undefined) {
        bodyParams["DataFormatConversionConfiguration"] = serializeAws_json1_1DataFormatConversionConfiguration(input.DataFormatConversionConfiguration, context);
    }
    if (input.EncryptionConfiguration !== undefined) {
        bodyParams["EncryptionConfiguration"] = serializeAws_json1_1EncryptionConfiguration(input.EncryptionConfiguration, context);
    }
    if (input.ErrorOutputPrefix !== undefined) {
        bodyParams["ErrorOutputPrefix"] = input.ErrorOutputPrefix;
    }
    if (input.Prefix !== undefined) {
        bodyParams["Prefix"] = input.Prefix;
    }
    if (input.ProcessingConfiguration !== undefined) {
        bodyParams["ProcessingConfiguration"] = serializeAws_json1_1ProcessingConfiguration(input.ProcessingConfiguration, context);
    }
    if (input.RoleARN !== undefined) {
        bodyParams["RoleARN"] = input.RoleARN;
    }
    if (input.S3BackupConfiguration !== undefined) {
        bodyParams["S3BackupConfiguration"] = serializeAws_json1_1S3DestinationConfiguration(input.S3BackupConfiguration, context);
    }
    if (input.S3BackupMode !== undefined) {
        bodyParams["S3BackupMode"] = input.S3BackupMode;
    }
    return bodyParams;
};
var serializeAws_json1_1ExtendedS3DestinationUpdate = function (input, context) {
    var bodyParams = {};
    if (input.BucketARN !== undefined) {
        bodyParams["BucketARN"] = input.BucketARN;
    }
    if (input.BufferingHints !== undefined) {
        bodyParams["BufferingHints"] = serializeAws_json1_1BufferingHints(input.BufferingHints, context);
    }
    if (input.CloudWatchLoggingOptions !== undefined) {
        bodyParams["CloudWatchLoggingOptions"] = serializeAws_json1_1CloudWatchLoggingOptions(input.CloudWatchLoggingOptions, context);
    }
    if (input.CompressionFormat !== undefined) {
        bodyParams["CompressionFormat"] = input.CompressionFormat;
    }
    if (input.DataFormatConversionConfiguration !== undefined) {
        bodyParams["DataFormatConversionConfiguration"] = serializeAws_json1_1DataFormatConversionConfiguration(input.DataFormatConversionConfiguration, context);
    }
    if (input.EncryptionConfiguration !== undefined) {
        bodyParams["EncryptionConfiguration"] = serializeAws_json1_1EncryptionConfiguration(input.EncryptionConfiguration, context);
    }
    if (input.ErrorOutputPrefix !== undefined) {
        bodyParams["ErrorOutputPrefix"] = input.ErrorOutputPrefix;
    }
    if (input.Prefix !== undefined) {
        bodyParams["Prefix"] = input.Prefix;
    }
    if (input.ProcessingConfiguration !== undefined) {
        bodyParams["ProcessingConfiguration"] = serializeAws_json1_1ProcessingConfiguration(input.ProcessingConfiguration, context);
    }
    if (input.RoleARN !== undefined) {
        bodyParams["RoleARN"] = input.RoleARN;
    }
    if (input.S3BackupMode !== undefined) {
        bodyParams["S3BackupMode"] = input.S3BackupMode;
    }
    if (input.S3BackupUpdate !== undefined) {
        bodyParams["S3BackupUpdate"] = serializeAws_json1_1S3DestinationUpdate(input.S3BackupUpdate, context);
    }
    return bodyParams;
};
var serializeAws_json1_1HiveJsonSerDe = function (input, context) {
    var bodyParams = {};
    if (input.TimestampFormats !== undefined) {
        bodyParams["TimestampFormats"] = serializeAws_json1_1ListOfNonEmptyStrings(input.TimestampFormats, context);
    }
    return bodyParams;
};
var serializeAws_json1_1InputFormatConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.Deserializer !== undefined) {
        bodyParams["Deserializer"] = serializeAws_json1_1Deserializer(input.Deserializer, context);
    }
    return bodyParams;
};
var serializeAws_json1_1KMSEncryptionConfig = function (input, context) {
    var bodyParams = {};
    if (input.AWSKMSKeyARN !== undefined) {
        bodyParams["AWSKMSKeyARN"] = input.AWSKMSKeyARN;
    }
    return bodyParams;
};
var serializeAws_json1_1KinesisStreamSourceConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.KinesisStreamARN !== undefined) {
        bodyParams["KinesisStreamARN"] = input.KinesisStreamARN;
    }
    if (input.RoleARN !== undefined) {
        bodyParams["RoleARN"] = input.RoleARN;
    }
    return bodyParams;
};
var serializeAws_json1_1ListDeliveryStreamsInput = function (input, context) {
    var bodyParams = {};
    if (input.DeliveryStreamType !== undefined) {
        bodyParams["DeliveryStreamType"] = input.DeliveryStreamType;
    }
    if (input.ExclusiveStartDeliveryStreamName !== undefined) {
        bodyParams["ExclusiveStartDeliveryStreamName"] =
            input.ExclusiveStartDeliveryStreamName;
    }
    if (input.Limit !== undefined) {
        bodyParams["Limit"] = input.Limit;
    }
    return bodyParams;
};
var serializeAws_json1_1ListOfNonEmptyStrings = function (input, context) {
    var e_1, _a;
    var contents = [];
    try {
        for (var input_1 = __values(input), input_1_1 = input_1.next(); !input_1_1.done; input_1_1 = input_1.next()) {
            var entry = input_1_1.value;
            contents.push(entry);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (input_1_1 && !input_1_1.done && (_a = input_1.return)) _a.call(input_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return contents;
};
var serializeAws_json1_1ListOfNonEmptyStringsWithoutWhitespace = function (input, context) {
    var e_2, _a;
    var contents = [];
    try {
        for (var input_2 = __values(input), input_2_1 = input_2.next(); !input_2_1.done; input_2_1 = input_2.next()) {
            var entry = input_2_1.value;
            contents.push(entry);
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (input_2_1 && !input_2_1.done && (_a = input_2.return)) _a.call(input_2);
        }
        finally { if (e_2) throw e_2.error; }
    }
    return contents;
};
var serializeAws_json1_1ListTagsForDeliveryStreamInput = function (input, context) {
    var bodyParams = {};
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    if (input.ExclusiveStartTagKey !== undefined) {
        bodyParams["ExclusiveStartTagKey"] = input.ExclusiveStartTagKey;
    }
    if (input.Limit !== undefined) {
        bodyParams["Limit"] = input.Limit;
    }
    return bodyParams;
};
var serializeAws_json1_1OpenXJsonSerDe = function (input, context) {
    var bodyParams = {};
    if (input.CaseInsensitive !== undefined) {
        bodyParams["CaseInsensitive"] = input.CaseInsensitive;
    }
    if (input.ColumnToJsonKeyMappings !== undefined) {
        bodyParams["ColumnToJsonKeyMappings"] = serializeAws_json1_1ColumnToJsonKeyMappings(input.ColumnToJsonKeyMappings, context);
    }
    if (input.ConvertDotsInJsonKeysToUnderscores !== undefined) {
        bodyParams["ConvertDotsInJsonKeysToUnderscores"] =
            input.ConvertDotsInJsonKeysToUnderscores;
    }
    return bodyParams;
};
var serializeAws_json1_1OrcSerDe = function (input, context) {
    var bodyParams = {};
    if (input.BlockSizeBytes !== undefined) {
        bodyParams["BlockSizeBytes"] = input.BlockSizeBytes;
    }
    if (input.BloomFilterColumns !== undefined) {
        bodyParams["BloomFilterColumns"] = serializeAws_json1_1ListOfNonEmptyStringsWithoutWhitespace(input.BloomFilterColumns, context);
    }
    if (input.BloomFilterFalsePositiveProbability !== undefined) {
        bodyParams["BloomFilterFalsePositiveProbability"] =
            input.BloomFilterFalsePositiveProbability;
    }
    if (input.Compression !== undefined) {
        bodyParams["Compression"] = input.Compression;
    }
    if (input.DictionaryKeyThreshold !== undefined) {
        bodyParams["DictionaryKeyThreshold"] = input.DictionaryKeyThreshold;
    }
    if (input.EnablePadding !== undefined) {
        bodyParams["EnablePadding"] = input.EnablePadding;
    }
    if (input.FormatVersion !== undefined) {
        bodyParams["FormatVersion"] = input.FormatVersion;
    }
    if (input.PaddingTolerance !== undefined) {
        bodyParams["PaddingTolerance"] = input.PaddingTolerance;
    }
    if (input.RowIndexStride !== undefined) {
        bodyParams["RowIndexStride"] = input.RowIndexStride;
    }
    if (input.StripeSizeBytes !== undefined) {
        bodyParams["StripeSizeBytes"] = input.StripeSizeBytes;
    }
    return bodyParams;
};
var serializeAws_json1_1OutputFormatConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.Serializer !== undefined) {
        bodyParams["Serializer"] = serializeAws_json1_1Serializer(input.Serializer, context);
    }
    return bodyParams;
};
var serializeAws_json1_1ParquetSerDe = function (input, context) {
    var bodyParams = {};
    if (input.BlockSizeBytes !== undefined) {
        bodyParams["BlockSizeBytes"] = input.BlockSizeBytes;
    }
    if (input.Compression !== undefined) {
        bodyParams["Compression"] = input.Compression;
    }
    if (input.EnableDictionaryCompression !== undefined) {
        bodyParams["EnableDictionaryCompression"] =
            input.EnableDictionaryCompression;
    }
    if (input.MaxPaddingBytes !== undefined) {
        bodyParams["MaxPaddingBytes"] = input.MaxPaddingBytes;
    }
    if (input.PageSizeBytes !== undefined) {
        bodyParams["PageSizeBytes"] = input.PageSizeBytes;
    }
    if (input.WriterVersion !== undefined) {
        bodyParams["WriterVersion"] = input.WriterVersion;
    }
    return bodyParams;
};
var serializeAws_json1_1ProcessingConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.Enabled !== undefined) {
        bodyParams["Enabled"] = input.Enabled;
    }
    if (input.Processors !== undefined) {
        bodyParams["Processors"] = serializeAws_json1_1ProcessorList(input.Processors, context);
    }
    return bodyParams;
};
var serializeAws_json1_1Processor = function (input, context) {
    var bodyParams = {};
    if (input.Parameters !== undefined) {
        bodyParams["Parameters"] = serializeAws_json1_1ProcessorParameterList(input.Parameters, context);
    }
    if (input.Type !== undefined) {
        bodyParams["Type"] = input.Type;
    }
    return bodyParams;
};
var serializeAws_json1_1ProcessorList = function (input, context) {
    var e_3, _a;
    var contents = [];
    try {
        for (var input_3 = __values(input), input_3_1 = input_3.next(); !input_3_1.done; input_3_1 = input_3.next()) {
            var entry = input_3_1.value;
            contents.push(serializeAws_json1_1Processor(entry, context));
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (input_3_1 && !input_3_1.done && (_a = input_3.return)) _a.call(input_3);
        }
        finally { if (e_3) throw e_3.error; }
    }
    return contents;
};
var serializeAws_json1_1ProcessorParameter = function (input, context) {
    var bodyParams = {};
    if (input.ParameterName !== undefined) {
        bodyParams["ParameterName"] = input.ParameterName;
    }
    if (input.ParameterValue !== undefined) {
        bodyParams["ParameterValue"] = input.ParameterValue;
    }
    return bodyParams;
};
var serializeAws_json1_1ProcessorParameterList = function (input, context) {
    var e_4, _a;
    var contents = [];
    try {
        for (var input_4 = __values(input), input_4_1 = input_4.next(); !input_4_1.done; input_4_1 = input_4.next()) {
            var entry = input_4_1.value;
            contents.push(serializeAws_json1_1ProcessorParameter(entry, context));
        }
    }
    catch (e_4_1) { e_4 = { error: e_4_1 }; }
    finally {
        try {
            if (input_4_1 && !input_4_1.done && (_a = input_4.return)) _a.call(input_4);
        }
        finally { if (e_4) throw e_4.error; }
    }
    return contents;
};
var serializeAws_json1_1PutRecordBatchInput = function (input, context) {
    var bodyParams = {};
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    if (input.Records !== undefined) {
        bodyParams["Records"] = serializeAws_json1_1PutRecordBatchRequestEntryList(input.Records, context);
    }
    return bodyParams;
};
var serializeAws_json1_1PutRecordBatchRequestEntryList = function (input, context) {
    var e_5, _a;
    var contents = [];
    try {
        for (var input_5 = __values(input), input_5_1 = input_5.next(); !input_5_1.done; input_5_1 = input_5.next()) {
            var entry = input_5_1.value;
            contents.push(serializeAws_json1_1_Record(entry, context));
        }
    }
    catch (e_5_1) { e_5 = { error: e_5_1 }; }
    finally {
        try {
            if (input_5_1 && !input_5_1.done && (_a = input_5.return)) _a.call(input_5);
        }
        finally { if (e_5) throw e_5.error; }
    }
    return contents;
};
var serializeAws_json1_1PutRecordInput = function (input, context) {
    var bodyParams = {};
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    if (input.Record !== undefined) {
        bodyParams["Record"] = serializeAws_json1_1_Record(input.Record, context);
    }
    return bodyParams;
};
var serializeAws_json1_1_Record = function (input, context) {
    var bodyParams = {};
    if (input.Data !== undefined) {
        bodyParams["Data"] = context.base64Encoder(input.Data);
    }
    return bodyParams;
};
var serializeAws_json1_1RedshiftDestinationConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.CloudWatchLoggingOptions !== undefined) {
        bodyParams["CloudWatchLoggingOptions"] = serializeAws_json1_1CloudWatchLoggingOptions(input.CloudWatchLoggingOptions, context);
    }
    if (input.ClusterJDBCURL !== undefined) {
        bodyParams["ClusterJDBCURL"] = input.ClusterJDBCURL;
    }
    if (input.CopyCommand !== undefined) {
        bodyParams["CopyCommand"] = serializeAws_json1_1CopyCommand(input.CopyCommand, context);
    }
    if (input.Password !== undefined) {
        bodyParams["Password"] = input.Password;
    }
    if (input.ProcessingConfiguration !== undefined) {
        bodyParams["ProcessingConfiguration"] = serializeAws_json1_1ProcessingConfiguration(input.ProcessingConfiguration, context);
    }
    if (input.RetryOptions !== undefined) {
        bodyParams["RetryOptions"] = serializeAws_json1_1RedshiftRetryOptions(input.RetryOptions, context);
    }
    if (input.RoleARN !== undefined) {
        bodyParams["RoleARN"] = input.RoleARN;
    }
    if (input.S3BackupConfiguration !== undefined) {
        bodyParams["S3BackupConfiguration"] = serializeAws_json1_1S3DestinationConfiguration(input.S3BackupConfiguration, context);
    }
    if (input.S3BackupMode !== undefined) {
        bodyParams["S3BackupMode"] = input.S3BackupMode;
    }
    if (input.S3Configuration !== undefined) {
        bodyParams["S3Configuration"] = serializeAws_json1_1S3DestinationConfiguration(input.S3Configuration, context);
    }
    if (input.Username !== undefined) {
        bodyParams["Username"] = input.Username;
    }
    return bodyParams;
};
var serializeAws_json1_1RedshiftDestinationUpdate = function (input, context) {
    var bodyParams = {};
    if (input.CloudWatchLoggingOptions !== undefined) {
        bodyParams["CloudWatchLoggingOptions"] = serializeAws_json1_1CloudWatchLoggingOptions(input.CloudWatchLoggingOptions, context);
    }
    if (input.ClusterJDBCURL !== undefined) {
        bodyParams["ClusterJDBCURL"] = input.ClusterJDBCURL;
    }
    if (input.CopyCommand !== undefined) {
        bodyParams["CopyCommand"] = serializeAws_json1_1CopyCommand(input.CopyCommand, context);
    }
    if (input.Password !== undefined) {
        bodyParams["Password"] = input.Password;
    }
    if (input.ProcessingConfiguration !== undefined) {
        bodyParams["ProcessingConfiguration"] = serializeAws_json1_1ProcessingConfiguration(input.ProcessingConfiguration, context);
    }
    if (input.RetryOptions !== undefined) {
        bodyParams["RetryOptions"] = serializeAws_json1_1RedshiftRetryOptions(input.RetryOptions, context);
    }
    if (input.RoleARN !== undefined) {
        bodyParams["RoleARN"] = input.RoleARN;
    }
    if (input.S3BackupMode !== undefined) {
        bodyParams["S3BackupMode"] = input.S3BackupMode;
    }
    if (input.S3BackupUpdate !== undefined) {
        bodyParams["S3BackupUpdate"] = serializeAws_json1_1S3DestinationUpdate(input.S3BackupUpdate, context);
    }
    if (input.S3Update !== undefined) {
        bodyParams["S3Update"] = serializeAws_json1_1S3DestinationUpdate(input.S3Update, context);
    }
    if (input.Username !== undefined) {
        bodyParams["Username"] = input.Username;
    }
    return bodyParams;
};
var serializeAws_json1_1RedshiftRetryOptions = function (input, context) {
    var bodyParams = {};
    if (input.DurationInSeconds !== undefined) {
        bodyParams["DurationInSeconds"] = input.DurationInSeconds;
    }
    return bodyParams;
};
var serializeAws_json1_1S3DestinationConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.BucketARN !== undefined) {
        bodyParams["BucketARN"] = input.BucketARN;
    }
    if (input.BufferingHints !== undefined) {
        bodyParams["BufferingHints"] = serializeAws_json1_1BufferingHints(input.BufferingHints, context);
    }
    if (input.CloudWatchLoggingOptions !== undefined) {
        bodyParams["CloudWatchLoggingOptions"] = serializeAws_json1_1CloudWatchLoggingOptions(input.CloudWatchLoggingOptions, context);
    }
    if (input.CompressionFormat !== undefined) {
        bodyParams["CompressionFormat"] = input.CompressionFormat;
    }
    if (input.EncryptionConfiguration !== undefined) {
        bodyParams["EncryptionConfiguration"] = serializeAws_json1_1EncryptionConfiguration(input.EncryptionConfiguration, context);
    }
    if (input.ErrorOutputPrefix !== undefined) {
        bodyParams["ErrorOutputPrefix"] = input.ErrorOutputPrefix;
    }
    if (input.Prefix !== undefined) {
        bodyParams["Prefix"] = input.Prefix;
    }
    if (input.RoleARN !== undefined) {
        bodyParams["RoleARN"] = input.RoleARN;
    }
    return bodyParams;
};
var serializeAws_json1_1S3DestinationUpdate = function (input, context) {
    var bodyParams = {};
    if (input.BucketARN !== undefined) {
        bodyParams["BucketARN"] = input.BucketARN;
    }
    if (input.BufferingHints !== undefined) {
        bodyParams["BufferingHints"] = serializeAws_json1_1BufferingHints(input.BufferingHints, context);
    }
    if (input.CloudWatchLoggingOptions !== undefined) {
        bodyParams["CloudWatchLoggingOptions"] = serializeAws_json1_1CloudWatchLoggingOptions(input.CloudWatchLoggingOptions, context);
    }
    if (input.CompressionFormat !== undefined) {
        bodyParams["CompressionFormat"] = input.CompressionFormat;
    }
    if (input.EncryptionConfiguration !== undefined) {
        bodyParams["EncryptionConfiguration"] = serializeAws_json1_1EncryptionConfiguration(input.EncryptionConfiguration, context);
    }
    if (input.ErrorOutputPrefix !== undefined) {
        bodyParams["ErrorOutputPrefix"] = input.ErrorOutputPrefix;
    }
    if (input.Prefix !== undefined) {
        bodyParams["Prefix"] = input.Prefix;
    }
    if (input.RoleARN !== undefined) {
        bodyParams["RoleARN"] = input.RoleARN;
    }
    return bodyParams;
};
var serializeAws_json1_1SchemaConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.CatalogId !== undefined) {
        bodyParams["CatalogId"] = input.CatalogId;
    }
    if (input.DatabaseName !== undefined) {
        bodyParams["DatabaseName"] = input.DatabaseName;
    }
    if (input.Region !== undefined) {
        bodyParams["Region"] = input.Region;
    }
    if (input.RoleARN !== undefined) {
        bodyParams["RoleARN"] = input.RoleARN;
    }
    if (input.TableName !== undefined) {
        bodyParams["TableName"] = input.TableName;
    }
    if (input.VersionId !== undefined) {
        bodyParams["VersionId"] = input.VersionId;
    }
    return bodyParams;
};
var serializeAws_json1_1Serializer = function (input, context) {
    var bodyParams = {};
    if (input.OrcSerDe !== undefined) {
        bodyParams["OrcSerDe"] = serializeAws_json1_1OrcSerDe(input.OrcSerDe, context);
    }
    if (input.ParquetSerDe !== undefined) {
        bodyParams["ParquetSerDe"] = serializeAws_json1_1ParquetSerDe(input.ParquetSerDe, context);
    }
    return bodyParams;
};
var serializeAws_json1_1SplunkDestinationConfiguration = function (input, context) {
    var bodyParams = {};
    if (input.CloudWatchLoggingOptions !== undefined) {
        bodyParams["CloudWatchLoggingOptions"] = serializeAws_json1_1CloudWatchLoggingOptions(input.CloudWatchLoggingOptions, context);
    }
    if (input.HECAcknowledgmentTimeoutInSeconds !== undefined) {
        bodyParams["HECAcknowledgmentTimeoutInSeconds"] =
            input.HECAcknowledgmentTimeoutInSeconds;
    }
    if (input.HECEndpoint !== undefined) {
        bodyParams["HECEndpoint"] = input.HECEndpoint;
    }
    if (input.HECEndpointType !== undefined) {
        bodyParams["HECEndpointType"] = input.HECEndpointType;
    }
    if (input.HECToken !== undefined) {
        bodyParams["HECToken"] = input.HECToken;
    }
    if (input.ProcessingConfiguration !== undefined) {
        bodyParams["ProcessingConfiguration"] = serializeAws_json1_1ProcessingConfiguration(input.ProcessingConfiguration, context);
    }
    if (input.RetryOptions !== undefined) {
        bodyParams["RetryOptions"] = serializeAws_json1_1SplunkRetryOptions(input.RetryOptions, context);
    }
    if (input.S3BackupMode !== undefined) {
        bodyParams["S3BackupMode"] = input.S3BackupMode;
    }
    if (input.S3Configuration !== undefined) {
        bodyParams["S3Configuration"] = serializeAws_json1_1S3DestinationConfiguration(input.S3Configuration, context);
    }
    return bodyParams;
};
var serializeAws_json1_1SplunkDestinationUpdate = function (input, context) {
    var bodyParams = {};
    if (input.CloudWatchLoggingOptions !== undefined) {
        bodyParams["CloudWatchLoggingOptions"] = serializeAws_json1_1CloudWatchLoggingOptions(input.CloudWatchLoggingOptions, context);
    }
    if (input.HECAcknowledgmentTimeoutInSeconds !== undefined) {
        bodyParams["HECAcknowledgmentTimeoutInSeconds"] =
            input.HECAcknowledgmentTimeoutInSeconds;
    }
    if (input.HECEndpoint !== undefined) {
        bodyParams["HECEndpoint"] = input.HECEndpoint;
    }
    if (input.HECEndpointType !== undefined) {
        bodyParams["HECEndpointType"] = input.HECEndpointType;
    }
    if (input.HECToken !== undefined) {
        bodyParams["HECToken"] = input.HECToken;
    }
    if (input.ProcessingConfiguration !== undefined) {
        bodyParams["ProcessingConfiguration"] = serializeAws_json1_1ProcessingConfiguration(input.ProcessingConfiguration, context);
    }
    if (input.RetryOptions !== undefined) {
        bodyParams["RetryOptions"] = serializeAws_json1_1SplunkRetryOptions(input.RetryOptions, context);
    }
    if (input.S3BackupMode !== undefined) {
        bodyParams["S3BackupMode"] = input.S3BackupMode;
    }
    if (input.S3Update !== undefined) {
        bodyParams["S3Update"] = serializeAws_json1_1S3DestinationUpdate(input.S3Update, context);
    }
    return bodyParams;
};
var serializeAws_json1_1SplunkRetryOptions = function (input, context) {
    var bodyParams = {};
    if (input.DurationInSeconds !== undefined) {
        bodyParams["DurationInSeconds"] = input.DurationInSeconds;
    }
    return bodyParams;
};
var serializeAws_json1_1StartDeliveryStreamEncryptionInput = function (input, context) {
    var bodyParams = {};
    if (input.DeliveryStreamEncryptionConfigurationInput !== undefined) {
        bodyParams["DeliveryStreamEncryptionConfigurationInput"] = serializeAws_json1_1DeliveryStreamEncryptionConfigurationInput(input.DeliveryStreamEncryptionConfigurationInput, context);
    }
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    return bodyParams;
};
var serializeAws_json1_1StopDeliveryStreamEncryptionInput = function (input, context) {
    var bodyParams = {};
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    return bodyParams;
};
var serializeAws_json1_1Tag = function (input, context) {
    var bodyParams = {};
    if (input.Key !== undefined) {
        bodyParams["Key"] = input.Key;
    }
    if (input.Value !== undefined) {
        bodyParams["Value"] = input.Value;
    }
    return bodyParams;
};
var serializeAws_json1_1TagDeliveryStreamInput = function (input, context) {
    var bodyParams = {};
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    if (input.Tags !== undefined) {
        bodyParams["Tags"] = serializeAws_json1_1TagDeliveryStreamInputTagList(input.Tags, context);
    }
    return bodyParams;
};
var serializeAws_json1_1TagDeliveryStreamInputTagList = function (input, context) {
    var e_6, _a;
    var contents = [];
    try {
        for (var input_6 = __values(input), input_6_1 = input_6.next(); !input_6_1.done; input_6_1 = input_6.next()) {
            var entry = input_6_1.value;
            contents.push(serializeAws_json1_1Tag(entry, context));
        }
    }
    catch (e_6_1) { e_6 = { error: e_6_1 }; }
    finally {
        try {
            if (input_6_1 && !input_6_1.done && (_a = input_6.return)) _a.call(input_6);
        }
        finally { if (e_6) throw e_6.error; }
    }
    return contents;
};
var serializeAws_json1_1TagKeyList = function (input, context) {
    var e_7, _a;
    var contents = [];
    try {
        for (var input_7 = __values(input), input_7_1 = input_7.next(); !input_7_1.done; input_7_1 = input_7.next()) {
            var entry = input_7_1.value;
            contents.push(entry);
        }
    }
    catch (e_7_1) { e_7 = { error: e_7_1 }; }
    finally {
        try {
            if (input_7_1 && !input_7_1.done && (_a = input_7.return)) _a.call(input_7);
        }
        finally { if (e_7) throw e_7.error; }
    }
    return contents;
};
var serializeAws_json1_1UntagDeliveryStreamInput = function (input, context) {
    var bodyParams = {};
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    if (input.TagKeys !== undefined) {
        bodyParams["TagKeys"] = serializeAws_json1_1TagKeyList(input.TagKeys, context);
    }
    return bodyParams;
};
var serializeAws_json1_1UpdateDestinationInput = function (input, context) {
    var bodyParams = {};
    if (input.CurrentDeliveryStreamVersionId !== undefined) {
        bodyParams["CurrentDeliveryStreamVersionId"] =
            input.CurrentDeliveryStreamVersionId;
    }
    if (input.DeliveryStreamName !== undefined) {
        bodyParams["DeliveryStreamName"] = input.DeliveryStreamName;
    }
    if (input.DestinationId !== undefined) {
        bodyParams["DestinationId"] = input.DestinationId;
    }
    if (input.ElasticsearchDestinationUpdate !== undefined) {
        bodyParams["ElasticsearchDestinationUpdate"] = serializeAws_json1_1ElasticsearchDestinationUpdate(input.ElasticsearchDestinationUpdate, context);
    }
    if (input.ExtendedS3DestinationUpdate !== undefined) {
        bodyParams["ExtendedS3DestinationUpdate"] = serializeAws_json1_1ExtendedS3DestinationUpdate(input.ExtendedS3DestinationUpdate, context);
    }
    if (input.RedshiftDestinationUpdate !== undefined) {
        bodyParams["RedshiftDestinationUpdate"] = serializeAws_json1_1RedshiftDestinationUpdate(input.RedshiftDestinationUpdate, context);
    }
    if (input.S3DestinationUpdate !== undefined) {
        bodyParams["S3DestinationUpdate"] = serializeAws_json1_1S3DestinationUpdate(input.S3DestinationUpdate, context);
    }
    if (input.SplunkDestinationUpdate !== undefined) {
        bodyParams["SplunkDestinationUpdate"] = serializeAws_json1_1SplunkDestinationUpdate(input.SplunkDestinationUpdate, context);
    }
    return bodyParams;
};
var deserializeAws_json1_1BufferingHints = function (output, context) {
    var contents = {
        __type: "BufferingHints",
        IntervalInSeconds: undefined,
        SizeInMBs: undefined
    };
    if (output.IntervalInSeconds !== undefined &&
        output.IntervalInSeconds !== null) {
        contents.IntervalInSeconds = output.IntervalInSeconds;
    }
    if (output.SizeInMBs !== undefined && output.SizeInMBs !== null) {
        contents.SizeInMBs = output.SizeInMBs;
    }
    return contents;
};
var deserializeAws_json1_1CloudWatchLoggingOptions = function (output, context) {
    var contents = {
        __type: "CloudWatchLoggingOptions",
        Enabled: undefined,
        LogGroupName: undefined,
        LogStreamName: undefined
    };
    if (output.Enabled !== undefined && output.Enabled !== null) {
        contents.Enabled = output.Enabled;
    }
    if (output.LogGroupName !== undefined && output.LogGroupName !== null) {
        contents.LogGroupName = output.LogGroupName;
    }
    if (output.LogStreamName !== undefined && output.LogStreamName !== null) {
        contents.LogStreamName = output.LogStreamName;
    }
    return contents;
};
var deserializeAws_json1_1ColumnToJsonKeyMappings = function (output, context) {
    var mapParams = {};
    Object.keys(output).forEach(function (key) {
        mapParams[key] = output[key];
    });
    return mapParams;
};
var deserializeAws_json1_1ConcurrentModificationException = function (output, context) {
    var contents = {
        __type: "ConcurrentModificationException",
        message: undefined
    };
    if (output.message !== undefined && output.message !== null) {
        contents.message = output.message;
    }
    return contents;
};
var deserializeAws_json1_1CopyCommand = function (output, context) {
    var contents = {
        __type: "CopyCommand",
        CopyOptions: undefined,
        DataTableColumns: undefined,
        DataTableName: undefined
    };
    if (output.CopyOptions !== undefined && output.CopyOptions !== null) {
        contents.CopyOptions = output.CopyOptions;
    }
    if (output.DataTableColumns !== undefined &&
        output.DataTableColumns !== null) {
        contents.DataTableColumns = output.DataTableColumns;
    }
    if (output.DataTableName !== undefined && output.DataTableName !== null) {
        contents.DataTableName = output.DataTableName;
    }
    return contents;
};
var deserializeAws_json1_1CreateDeliveryStreamOutput = function (output, context) {
    var contents = {
        __type: "CreateDeliveryStreamOutput",
        DeliveryStreamARN: undefined
    };
    if (output.DeliveryStreamARN !== undefined &&
        output.DeliveryStreamARN !== null) {
        contents.DeliveryStreamARN = output.DeliveryStreamARN;
    }
    return contents;
};
var deserializeAws_json1_1DataFormatConversionConfiguration = function (output, context) {
    var contents = {
        __type: "DataFormatConversionConfiguration",
        Enabled: undefined,
        InputFormatConfiguration: undefined,
        OutputFormatConfiguration: undefined,
        SchemaConfiguration: undefined
    };
    if (output.Enabled !== undefined && output.Enabled !== null) {
        contents.Enabled = output.Enabled;
    }
    if (output.InputFormatConfiguration !== undefined &&
        output.InputFormatConfiguration !== null) {
        contents.InputFormatConfiguration = deserializeAws_json1_1InputFormatConfiguration(output.InputFormatConfiguration, context);
    }
    if (output.OutputFormatConfiguration !== undefined &&
        output.OutputFormatConfiguration !== null) {
        contents.OutputFormatConfiguration = deserializeAws_json1_1OutputFormatConfiguration(output.OutputFormatConfiguration, context);
    }
    if (output.SchemaConfiguration !== undefined &&
        output.SchemaConfiguration !== null) {
        contents.SchemaConfiguration = deserializeAws_json1_1SchemaConfiguration(output.SchemaConfiguration, context);
    }
    return contents;
};
var deserializeAws_json1_1DeleteDeliveryStreamOutput = function (output, context) {
    var contents = {
        __type: "DeleteDeliveryStreamOutput"
    };
    return contents;
};
var deserializeAws_json1_1DeliveryStreamDescription = function (output, context) {
    var contents = {
        __type: "DeliveryStreamDescription",
        CreateTimestamp: undefined,
        DeliveryStreamARN: undefined,
        DeliveryStreamEncryptionConfiguration: undefined,
        DeliveryStreamName: undefined,
        DeliveryStreamStatus: undefined,
        DeliveryStreamType: undefined,
        Destinations: undefined,
        FailureDescription: undefined,
        HasMoreDestinations: undefined,
        LastUpdateTimestamp: undefined,
        Source: undefined,
        VersionId: undefined
    };
    if (output.CreateTimestamp !== undefined && output.CreateTimestamp !== null) {
        contents.CreateTimestamp = new Date(Math.round(output.CreateTimestamp * 1000));
    }
    if (output.DeliveryStreamARN !== undefined &&
        output.DeliveryStreamARN !== null) {
        contents.DeliveryStreamARN = output.DeliveryStreamARN;
    }
    if (output.DeliveryStreamEncryptionConfiguration !== undefined &&
        output.DeliveryStreamEncryptionConfiguration !== null) {
        contents.DeliveryStreamEncryptionConfiguration = deserializeAws_json1_1DeliveryStreamEncryptionConfiguration(output.DeliveryStreamEncryptionConfiguration, context);
    }
    if (output.DeliveryStreamName !== undefined &&
        output.DeliveryStreamName !== null) {
        contents.DeliveryStreamName = output.DeliveryStreamName;
    }
    if (output.DeliveryStreamStatus !== undefined &&
        output.DeliveryStreamStatus !== null) {
        contents.DeliveryStreamStatus = output.DeliveryStreamStatus;
    }
    if (output.DeliveryStreamType !== undefined &&
        output.DeliveryStreamType !== null) {
        contents.DeliveryStreamType = output.DeliveryStreamType;
    }
    if (output.Destinations !== undefined && output.Destinations !== null) {
        contents.Destinations = deserializeAws_json1_1DestinationDescriptionList(output.Destinations, context);
    }
    if (output.FailureDescription !== undefined &&
        output.FailureDescription !== null) {
        contents.FailureDescription = deserializeAws_json1_1FailureDescription(output.FailureDescription, context);
    }
    if (output.HasMoreDestinations !== undefined &&
        output.HasMoreDestinations !== null) {
        contents.HasMoreDestinations = output.HasMoreDestinations;
    }
    if (output.LastUpdateTimestamp !== undefined &&
        output.LastUpdateTimestamp !== null) {
        contents.LastUpdateTimestamp = new Date(Math.round(output.LastUpdateTimestamp * 1000));
    }
    if (output.Source !== undefined && output.Source !== null) {
        contents.Source = deserializeAws_json1_1SourceDescription(output.Source, context);
    }
    if (output.VersionId !== undefined && output.VersionId !== null) {
        contents.VersionId = output.VersionId;
    }
    return contents;
};
var deserializeAws_json1_1DeliveryStreamEncryptionConfiguration = function (output, context) {
    var contents = {
        __type: "DeliveryStreamEncryptionConfiguration",
        FailureDescription: undefined,
        KeyARN: undefined,
        KeyType: undefined,
        Status: undefined
    };
    if (output.FailureDescription !== undefined &&
        output.FailureDescription !== null) {
        contents.FailureDescription = deserializeAws_json1_1FailureDescription(output.FailureDescription, context);
    }
    if (output.KeyARN !== undefined && output.KeyARN !== null) {
        contents.KeyARN = output.KeyARN;
    }
    if (output.KeyType !== undefined && output.KeyType !== null) {
        contents.KeyType = output.KeyType;
    }
    if (output.Status !== undefined && output.Status !== null) {
        contents.Status = output.Status;
    }
    return contents;
};
var deserializeAws_json1_1DeliveryStreamNameList = function (output, context) {
    return (output || []).map(function (entry) { return entry; });
};
var deserializeAws_json1_1DescribeDeliveryStreamOutput = function (output, context) {
    var contents = {
        __type: "DescribeDeliveryStreamOutput",
        DeliveryStreamDescription: undefined
    };
    if (output.DeliveryStreamDescription !== undefined &&
        output.DeliveryStreamDescription !== null) {
        contents.DeliveryStreamDescription = deserializeAws_json1_1DeliveryStreamDescription(output.DeliveryStreamDescription, context);
    }
    return contents;
};
var deserializeAws_json1_1Deserializer = function (output, context) {
    var contents = {
        __type: "Deserializer",
        HiveJsonSerDe: undefined,
        OpenXJsonSerDe: undefined
    };
    if (output.HiveJsonSerDe !== undefined && output.HiveJsonSerDe !== null) {
        contents.HiveJsonSerDe = deserializeAws_json1_1HiveJsonSerDe(output.HiveJsonSerDe, context);
    }
    if (output.OpenXJsonSerDe !== undefined && output.OpenXJsonSerDe !== null) {
        contents.OpenXJsonSerDe = deserializeAws_json1_1OpenXJsonSerDe(output.OpenXJsonSerDe, context);
    }
    return contents;
};
var deserializeAws_json1_1DestinationDescription = function (output, context) {
    var contents = {
        __type: "DestinationDescription",
        DestinationId: undefined,
        ElasticsearchDestinationDescription: undefined,
        ExtendedS3DestinationDescription: undefined,
        RedshiftDestinationDescription: undefined,
        S3DestinationDescription: undefined,
        SplunkDestinationDescription: undefined
    };
    if (output.DestinationId !== undefined && output.DestinationId !== null) {
        contents.DestinationId = output.DestinationId;
    }
    if (output.ElasticsearchDestinationDescription !== undefined &&
        output.ElasticsearchDestinationDescription !== null) {
        contents.ElasticsearchDestinationDescription = deserializeAws_json1_1ElasticsearchDestinationDescription(output.ElasticsearchDestinationDescription, context);
    }
    if (output.ExtendedS3DestinationDescription !== undefined &&
        output.ExtendedS3DestinationDescription !== null) {
        contents.ExtendedS3DestinationDescription = deserializeAws_json1_1ExtendedS3DestinationDescription(output.ExtendedS3DestinationDescription, context);
    }
    if (output.RedshiftDestinationDescription !== undefined &&
        output.RedshiftDestinationDescription !== null) {
        contents.RedshiftDestinationDescription = deserializeAws_json1_1RedshiftDestinationDescription(output.RedshiftDestinationDescription, context);
    }
    if (output.S3DestinationDescription !== undefined &&
        output.S3DestinationDescription !== null) {
        contents.S3DestinationDescription = deserializeAws_json1_1S3DestinationDescription(output.S3DestinationDescription, context);
    }
    if (output.SplunkDestinationDescription !== undefined &&
        output.SplunkDestinationDescription !== null) {
        contents.SplunkDestinationDescription = deserializeAws_json1_1SplunkDestinationDescription(output.SplunkDestinationDescription, context);
    }
    return contents;
};
var deserializeAws_json1_1DestinationDescriptionList = function (output, context) {
    return (output || []).map(function (entry) {
        return deserializeAws_json1_1DestinationDescription(entry, context);
    });
};
var deserializeAws_json1_1ElasticsearchBufferingHints = function (output, context) {
    var contents = {
        __type: "ElasticsearchBufferingHints",
        IntervalInSeconds: undefined,
        SizeInMBs: undefined
    };
    if (output.IntervalInSeconds !== undefined &&
        output.IntervalInSeconds !== null) {
        contents.IntervalInSeconds = output.IntervalInSeconds;
    }
    if (output.SizeInMBs !== undefined && output.SizeInMBs !== null) {
        contents.SizeInMBs = output.SizeInMBs;
    }
    return contents;
};
var deserializeAws_json1_1ElasticsearchDestinationDescription = function (output, context) {
    var contents = {
        __type: "ElasticsearchDestinationDescription",
        BufferingHints: undefined,
        CloudWatchLoggingOptions: undefined,
        ClusterEndpoint: undefined,
        DomainARN: undefined,
        IndexName: undefined,
        IndexRotationPeriod: undefined,
        ProcessingConfiguration: undefined,
        RetryOptions: undefined,
        RoleARN: undefined,
        S3BackupMode: undefined,
        S3DestinationDescription: undefined,
        TypeName: undefined
    };
    if (output.BufferingHints !== undefined && output.BufferingHints !== null) {
        contents.BufferingHints = deserializeAws_json1_1ElasticsearchBufferingHints(output.BufferingHints, context);
    }
    if (output.CloudWatchLoggingOptions !== undefined &&
        output.CloudWatchLoggingOptions !== null) {
        contents.CloudWatchLoggingOptions = deserializeAws_json1_1CloudWatchLoggingOptions(output.CloudWatchLoggingOptions, context);
    }
    if (output.ClusterEndpoint !== undefined && output.ClusterEndpoint !== null) {
        contents.ClusterEndpoint = output.ClusterEndpoint;
    }
    if (output.DomainARN !== undefined && output.DomainARN !== null) {
        contents.DomainARN = output.DomainARN;
    }
    if (output.IndexName !== undefined && output.IndexName !== null) {
        contents.IndexName = output.IndexName;
    }
    if (output.IndexRotationPeriod !== undefined &&
        output.IndexRotationPeriod !== null) {
        contents.IndexRotationPeriod = output.IndexRotationPeriod;
    }
    if (output.ProcessingConfiguration !== undefined &&
        output.ProcessingConfiguration !== null) {
        contents.ProcessingConfiguration = deserializeAws_json1_1ProcessingConfiguration(output.ProcessingConfiguration, context);
    }
    if (output.RetryOptions !== undefined && output.RetryOptions !== null) {
        contents.RetryOptions = deserializeAws_json1_1ElasticsearchRetryOptions(output.RetryOptions, context);
    }
    if (output.RoleARN !== undefined && output.RoleARN !== null) {
        contents.RoleARN = output.RoleARN;
    }
    if (output.S3BackupMode !== undefined && output.S3BackupMode !== null) {
        contents.S3BackupMode = output.S3BackupMode;
    }
    if (output.S3DestinationDescription !== undefined &&
        output.S3DestinationDescription !== null) {
        contents.S3DestinationDescription = deserializeAws_json1_1S3DestinationDescription(output.S3DestinationDescription, context);
    }
    if (output.TypeName !== undefined && output.TypeName !== null) {
        contents.TypeName = output.TypeName;
    }
    return contents;
};
var deserializeAws_json1_1ElasticsearchRetryOptions = function (output, context) {
    var contents = {
        __type: "ElasticsearchRetryOptions",
        DurationInSeconds: undefined
    };
    if (output.DurationInSeconds !== undefined &&
        output.DurationInSeconds !== null) {
        contents.DurationInSeconds = output.DurationInSeconds;
    }
    return contents;
};
var deserializeAws_json1_1EncryptionConfiguration = function (output, context) {
    var contents = {
        __type: "EncryptionConfiguration",
        KMSEncryptionConfig: undefined,
        NoEncryptionConfig: undefined
    };
    if (output.KMSEncryptionConfig !== undefined &&
        output.KMSEncryptionConfig !== null) {
        contents.KMSEncryptionConfig = deserializeAws_json1_1KMSEncryptionConfig(output.KMSEncryptionConfig, context);
    }
    if (output.NoEncryptionConfig !== undefined &&
        output.NoEncryptionConfig !== null) {
        contents.NoEncryptionConfig = output.NoEncryptionConfig;
    }
    return contents;
};
var deserializeAws_json1_1ExtendedS3DestinationDescription = function (output, context) {
    var contents = {
        __type: "ExtendedS3DestinationDescription",
        BucketARN: undefined,
        BufferingHints: undefined,
        CloudWatchLoggingOptions: undefined,
        CompressionFormat: undefined,
        DataFormatConversionConfiguration: undefined,
        EncryptionConfiguration: undefined,
        ErrorOutputPrefix: undefined,
        Prefix: undefined,
        ProcessingConfiguration: undefined,
        RoleARN: undefined,
        S3BackupDescription: undefined,
        S3BackupMode: undefined
    };
    if (output.BucketARN !== undefined && output.BucketARN !== null) {
        contents.BucketARN = output.BucketARN;
    }
    if (output.BufferingHints !== undefined && output.BufferingHints !== null) {
        contents.BufferingHints = deserializeAws_json1_1BufferingHints(output.BufferingHints, context);
    }
    if (output.CloudWatchLoggingOptions !== undefined &&
        output.CloudWatchLoggingOptions !== null) {
        contents.CloudWatchLoggingOptions = deserializeAws_json1_1CloudWatchLoggingOptions(output.CloudWatchLoggingOptions, context);
    }
    if (output.CompressionFormat !== undefined &&
        output.CompressionFormat !== null) {
        contents.CompressionFormat = output.CompressionFormat;
    }
    if (output.DataFormatConversionConfiguration !== undefined &&
        output.DataFormatConversionConfiguration !== null) {
        contents.DataFormatConversionConfiguration = deserializeAws_json1_1DataFormatConversionConfiguration(output.DataFormatConversionConfiguration, context);
    }
    if (output.EncryptionConfiguration !== undefined &&
        output.EncryptionConfiguration !== null) {
        contents.EncryptionConfiguration = deserializeAws_json1_1EncryptionConfiguration(output.EncryptionConfiguration, context);
    }
    if (output.ErrorOutputPrefix !== undefined &&
        output.ErrorOutputPrefix !== null) {
        contents.ErrorOutputPrefix = output.ErrorOutputPrefix;
    }
    if (output.Prefix !== undefined && output.Prefix !== null) {
        contents.Prefix = output.Prefix;
    }
    if (output.ProcessingConfiguration !== undefined &&
        output.ProcessingConfiguration !== null) {
        contents.ProcessingConfiguration = deserializeAws_json1_1ProcessingConfiguration(output.ProcessingConfiguration, context);
    }
    if (output.RoleARN !== undefined && output.RoleARN !== null) {
        contents.RoleARN = output.RoleARN;
    }
    if (output.S3BackupDescription !== undefined &&
        output.S3BackupDescription !== null) {
        contents.S3BackupDescription = deserializeAws_json1_1S3DestinationDescription(output.S3BackupDescription, context);
    }
    if (output.S3BackupMode !== undefined && output.S3BackupMode !== null) {
        contents.S3BackupMode = output.S3BackupMode;
    }
    return contents;
};
var deserializeAws_json1_1FailureDescription = function (output, context) {
    var contents = {
        __type: "FailureDescription",
        Details: undefined,
        Type: undefined
    };
    if (output.Details !== undefined && output.Details !== null) {
        contents.Details = output.Details;
    }
    if (output.Type !== undefined && output.Type !== null) {
        contents.Type = output.Type;
    }
    return contents;
};
var deserializeAws_json1_1HiveJsonSerDe = function (output, context) {
    var contents = {
        __type: "HiveJsonSerDe",
        TimestampFormats: undefined
    };
    if (output.TimestampFormats !== undefined &&
        output.TimestampFormats !== null) {
        contents.TimestampFormats = deserializeAws_json1_1ListOfNonEmptyStrings(output.TimestampFormats, context);
    }
    return contents;
};
var deserializeAws_json1_1InputFormatConfiguration = function (output, context) {
    var contents = {
        __type: "InputFormatConfiguration",
        Deserializer: undefined
    };
    if (output.Deserializer !== undefined && output.Deserializer !== null) {
        contents.Deserializer = deserializeAws_json1_1Deserializer(output.Deserializer, context);
    }
    return contents;
};
var deserializeAws_json1_1InvalidArgumentException = function (output, context) {
    var contents = {
        __type: "InvalidArgumentException",
        message: undefined
    };
    if (output.message !== undefined && output.message !== null) {
        contents.message = output.message;
    }
    return contents;
};
var deserializeAws_json1_1InvalidKMSResourceException = function (output, context) {
    var contents = {
        __type: "InvalidKMSResourceException",
        code: undefined,
        message: undefined
    };
    if (output.code !== undefined && output.code !== null) {
        contents.code = output.code;
    }
    if (output.message !== undefined && output.message !== null) {
        contents.message = output.message;
    }
    return contents;
};
var deserializeAws_json1_1KMSEncryptionConfig = function (output, context) {
    var contents = {
        __type: "KMSEncryptionConfig",
        AWSKMSKeyARN: undefined
    };
    if (output.AWSKMSKeyARN !== undefined && output.AWSKMSKeyARN !== null) {
        contents.AWSKMSKeyARN = output.AWSKMSKeyARN;
    }
    return contents;
};
var deserializeAws_json1_1KinesisStreamSourceDescription = function (output, context) {
    var contents = {
        __type: "KinesisStreamSourceDescription",
        DeliveryStartTimestamp: undefined,
        KinesisStreamARN: undefined,
        RoleARN: undefined
    };
    if (output.DeliveryStartTimestamp !== undefined &&
        output.DeliveryStartTimestamp !== null) {
        contents.DeliveryStartTimestamp = new Date(Math.round(output.DeliveryStartTimestamp * 1000));
    }
    if (output.KinesisStreamARN !== undefined &&
        output.KinesisStreamARN !== null) {
        contents.KinesisStreamARN = output.KinesisStreamARN;
    }
    if (output.RoleARN !== undefined && output.RoleARN !== null) {
        contents.RoleARN = output.RoleARN;
    }
    return contents;
};
var deserializeAws_json1_1LimitExceededException = function (output, context) {
    var contents = {
        __type: "LimitExceededException",
        message: undefined
    };
    if (output.message !== undefined && output.message !== null) {
        contents.message = output.message;
    }
    return contents;
};
var deserializeAws_json1_1ListDeliveryStreamsOutput = function (output, context) {
    var contents = {
        __type: "ListDeliveryStreamsOutput",
        DeliveryStreamNames: undefined,
        HasMoreDeliveryStreams: undefined
    };
    if (output.DeliveryStreamNames !== undefined &&
        output.DeliveryStreamNames !== null) {
        contents.DeliveryStreamNames = deserializeAws_json1_1DeliveryStreamNameList(output.DeliveryStreamNames, context);
    }
    if (output.HasMoreDeliveryStreams !== undefined &&
        output.HasMoreDeliveryStreams !== null) {
        contents.HasMoreDeliveryStreams = output.HasMoreDeliveryStreams;
    }
    return contents;
};
var deserializeAws_json1_1ListOfNonEmptyStrings = function (output, context) {
    return (output || []).map(function (entry) { return entry; });
};
var deserializeAws_json1_1ListOfNonEmptyStringsWithoutWhitespace = function (output, context) {
    return (output || []).map(function (entry) { return entry; });
};
var deserializeAws_json1_1ListTagsForDeliveryStreamOutput = function (output, context) {
    var contents = {
        __type: "ListTagsForDeliveryStreamOutput",
        HasMoreTags: undefined,
        Tags: undefined
    };
    if (output.HasMoreTags !== undefined && output.HasMoreTags !== null) {
        contents.HasMoreTags = output.HasMoreTags;
    }
    if (output.Tags !== undefined && output.Tags !== null) {
        contents.Tags = deserializeAws_json1_1ListTagsForDeliveryStreamOutputTagList(output.Tags, context);
    }
    return contents;
};
var deserializeAws_json1_1ListTagsForDeliveryStreamOutputTagList = function (output, context) {
    return (output || []).map(function (entry) {
        return deserializeAws_json1_1Tag(entry, context);
    });
};
var deserializeAws_json1_1OpenXJsonSerDe = function (output, context) {
    var contents = {
        __type: "OpenXJsonSerDe",
        CaseInsensitive: undefined,
        ColumnToJsonKeyMappings: undefined,
        ConvertDotsInJsonKeysToUnderscores: undefined
    };
    if (output.CaseInsensitive !== undefined && output.CaseInsensitive !== null) {
        contents.CaseInsensitive = output.CaseInsensitive;
    }
    if (output.ColumnToJsonKeyMappings !== undefined &&
        output.ColumnToJsonKeyMappings !== null) {
        contents.ColumnToJsonKeyMappings = deserializeAws_json1_1ColumnToJsonKeyMappings(output.ColumnToJsonKeyMappings, context);
    }
    if (output.ConvertDotsInJsonKeysToUnderscores !== undefined &&
        output.ConvertDotsInJsonKeysToUnderscores !== null) {
        contents.ConvertDotsInJsonKeysToUnderscores =
            output.ConvertDotsInJsonKeysToUnderscores;
    }
    return contents;
};
var deserializeAws_json1_1OrcSerDe = function (output, context) {
    var contents = {
        __type: "OrcSerDe",
        BlockSizeBytes: undefined,
        BloomFilterColumns: undefined,
        BloomFilterFalsePositiveProbability: undefined,
        Compression: undefined,
        DictionaryKeyThreshold: undefined,
        EnablePadding: undefined,
        FormatVersion: undefined,
        PaddingTolerance: undefined,
        RowIndexStride: undefined,
        StripeSizeBytes: undefined
    };
    if (output.BlockSizeBytes !== undefined && output.BlockSizeBytes !== null) {
        contents.BlockSizeBytes = output.BlockSizeBytes;
    }
    if (output.BloomFilterColumns !== undefined &&
        output.BloomFilterColumns !== null) {
        contents.BloomFilterColumns = deserializeAws_json1_1ListOfNonEmptyStringsWithoutWhitespace(output.BloomFilterColumns, context);
    }
    if (output.BloomFilterFalsePositiveProbability !== undefined &&
        output.BloomFilterFalsePositiveProbability !== null) {
        contents.BloomFilterFalsePositiveProbability =
            output.BloomFilterFalsePositiveProbability;
    }
    if (output.Compression !== undefined && output.Compression !== null) {
        contents.Compression = output.Compression;
    }
    if (output.DictionaryKeyThreshold !== undefined &&
        output.DictionaryKeyThreshold !== null) {
        contents.DictionaryKeyThreshold = output.DictionaryKeyThreshold;
    }
    if (output.EnablePadding !== undefined && output.EnablePadding !== null) {
        contents.EnablePadding = output.EnablePadding;
    }
    if (output.FormatVersion !== undefined && output.FormatVersion !== null) {
        contents.FormatVersion = output.FormatVersion;
    }
    if (output.PaddingTolerance !== undefined &&
        output.PaddingTolerance !== null) {
        contents.PaddingTolerance = output.PaddingTolerance;
    }
    if (output.RowIndexStride !== undefined && output.RowIndexStride !== null) {
        contents.RowIndexStride = output.RowIndexStride;
    }
    if (output.StripeSizeBytes !== undefined && output.StripeSizeBytes !== null) {
        contents.StripeSizeBytes = output.StripeSizeBytes;
    }
    return contents;
};
var deserializeAws_json1_1OutputFormatConfiguration = function (output, context) {
    var contents = {
        __type: "OutputFormatConfiguration",
        Serializer: undefined
    };
    if (output.Serializer !== undefined && output.Serializer !== null) {
        contents.Serializer = deserializeAws_json1_1Serializer(output.Serializer, context);
    }
    return contents;
};
var deserializeAws_json1_1ParquetSerDe = function (output, context) {
    var contents = {
        __type: "ParquetSerDe",
        BlockSizeBytes: undefined,
        Compression: undefined,
        EnableDictionaryCompression: undefined,
        MaxPaddingBytes: undefined,
        PageSizeBytes: undefined,
        WriterVersion: undefined
    };
    if (output.BlockSizeBytes !== undefined && output.BlockSizeBytes !== null) {
        contents.BlockSizeBytes = output.BlockSizeBytes;
    }
    if (output.Compression !== undefined && output.Compression !== null) {
        contents.Compression = output.Compression;
    }
    if (output.EnableDictionaryCompression !== undefined &&
        output.EnableDictionaryCompression !== null) {
        contents.EnableDictionaryCompression = output.EnableDictionaryCompression;
    }
    if (output.MaxPaddingBytes !== undefined && output.MaxPaddingBytes !== null) {
        contents.MaxPaddingBytes = output.MaxPaddingBytes;
    }
    if (output.PageSizeBytes !== undefined && output.PageSizeBytes !== null) {
        contents.PageSizeBytes = output.PageSizeBytes;
    }
    if (output.WriterVersion !== undefined && output.WriterVersion !== null) {
        contents.WriterVersion = output.WriterVersion;
    }
    return contents;
};
var deserializeAws_json1_1ProcessingConfiguration = function (output, context) {
    var contents = {
        __type: "ProcessingConfiguration",
        Enabled: undefined,
        Processors: undefined
    };
    if (output.Enabled !== undefined && output.Enabled !== null) {
        contents.Enabled = output.Enabled;
    }
    if (output.Processors !== undefined && output.Processors !== null) {
        contents.Processors = deserializeAws_json1_1ProcessorList(output.Processors, context);
    }
    return contents;
};
var deserializeAws_json1_1Processor = function (output, context) {
    var contents = {
        __type: "Processor",
        Parameters: undefined,
        Type: undefined
    };
    if (output.Parameters !== undefined && output.Parameters !== null) {
        contents.Parameters = deserializeAws_json1_1ProcessorParameterList(output.Parameters, context);
    }
    if (output.Type !== undefined && output.Type !== null) {
        contents.Type = output.Type;
    }
    return contents;
};
var deserializeAws_json1_1ProcessorList = function (output, context) {
    return (output || []).map(function (entry) {
        return deserializeAws_json1_1Processor(entry, context);
    });
};
var deserializeAws_json1_1ProcessorParameter = function (output, context) {
    var contents = {
        __type: "ProcessorParameter",
        ParameterName: undefined,
        ParameterValue: undefined
    };
    if (output.ParameterName !== undefined && output.ParameterName !== null) {
        contents.ParameterName = output.ParameterName;
    }
    if (output.ParameterValue !== undefined && output.ParameterValue !== null) {
        contents.ParameterValue = output.ParameterValue;
    }
    return contents;
};
var deserializeAws_json1_1ProcessorParameterList = function (output, context) {
    return (output || []).map(function (entry) {
        return deserializeAws_json1_1ProcessorParameter(entry, context);
    });
};
var deserializeAws_json1_1PutRecordBatchOutput = function (output, context) {
    var contents = {
        __type: "PutRecordBatchOutput",
        Encrypted: undefined,
        FailedPutCount: undefined,
        RequestResponses: undefined
    };
    if (output.Encrypted !== undefined && output.Encrypted !== null) {
        contents.Encrypted = output.Encrypted;
    }
    if (output.FailedPutCount !== undefined && output.FailedPutCount !== null) {
        contents.FailedPutCount = output.FailedPutCount;
    }
    if (output.RequestResponses !== undefined &&
        output.RequestResponses !== null) {
        contents.RequestResponses = deserializeAws_json1_1PutRecordBatchResponseEntryList(output.RequestResponses, context);
    }
    return contents;
};
var deserializeAws_json1_1PutRecordBatchResponseEntry = function (output, context) {
    var contents = {
        __type: "PutRecordBatchResponseEntry",
        ErrorCode: undefined,
        ErrorMessage: undefined,
        RecordId: undefined
    };
    if (output.ErrorCode !== undefined && output.ErrorCode !== null) {
        contents.ErrorCode = output.ErrorCode;
    }
    if (output.ErrorMessage !== undefined && output.ErrorMessage !== null) {
        contents.ErrorMessage = output.ErrorMessage;
    }
    if (output.RecordId !== undefined && output.RecordId !== null) {
        contents.RecordId = output.RecordId;
    }
    return contents;
};
var deserializeAws_json1_1PutRecordBatchResponseEntryList = function (output, context) {
    return (output || []).map(function (entry) {
        return deserializeAws_json1_1PutRecordBatchResponseEntry(entry, context);
    });
};
var deserializeAws_json1_1PutRecordOutput = function (output, context) {
    var contents = {
        __type: "PutRecordOutput",
        Encrypted: undefined,
        RecordId: undefined
    };
    if (output.Encrypted !== undefined && output.Encrypted !== null) {
        contents.Encrypted = output.Encrypted;
    }
    if (output.RecordId !== undefined && output.RecordId !== null) {
        contents.RecordId = output.RecordId;
    }
    return contents;
};
var deserializeAws_json1_1RedshiftDestinationDescription = function (output, context) {
    var contents = {
        __type: "RedshiftDestinationDescription",
        CloudWatchLoggingOptions: undefined,
        ClusterJDBCURL: undefined,
        CopyCommand: undefined,
        ProcessingConfiguration: undefined,
        RetryOptions: undefined,
        RoleARN: undefined,
        S3BackupDescription: undefined,
        S3BackupMode: undefined,
        S3DestinationDescription: undefined,
        Username: undefined
    };
    if (output.CloudWatchLoggingOptions !== undefined &&
        output.CloudWatchLoggingOptions !== null) {
        contents.CloudWatchLoggingOptions = deserializeAws_json1_1CloudWatchLoggingOptions(output.CloudWatchLoggingOptions, context);
    }
    if (output.ClusterJDBCURL !== undefined && output.ClusterJDBCURL !== null) {
        contents.ClusterJDBCURL = output.ClusterJDBCURL;
    }
    if (output.CopyCommand !== undefined && output.CopyCommand !== null) {
        contents.CopyCommand = deserializeAws_json1_1CopyCommand(output.CopyCommand, context);
    }
    if (output.ProcessingConfiguration !== undefined &&
        output.ProcessingConfiguration !== null) {
        contents.ProcessingConfiguration = deserializeAws_json1_1ProcessingConfiguration(output.ProcessingConfiguration, context);
    }
    if (output.RetryOptions !== undefined && output.RetryOptions !== null) {
        contents.RetryOptions = deserializeAws_json1_1RedshiftRetryOptions(output.RetryOptions, context);
    }
    if (output.RoleARN !== undefined && output.RoleARN !== null) {
        contents.RoleARN = output.RoleARN;
    }
    if (output.S3BackupDescription !== undefined &&
        output.S3BackupDescription !== null) {
        contents.S3BackupDescription = deserializeAws_json1_1S3DestinationDescription(output.S3BackupDescription, context);
    }
    if (output.S3BackupMode !== undefined && output.S3BackupMode !== null) {
        contents.S3BackupMode = output.S3BackupMode;
    }
    if (output.S3DestinationDescription !== undefined &&
        output.S3DestinationDescription !== null) {
        contents.S3DestinationDescription = deserializeAws_json1_1S3DestinationDescription(output.S3DestinationDescription, context);
    }
    if (output.Username !== undefined && output.Username !== null) {
        contents.Username = output.Username;
    }
    return contents;
};
var deserializeAws_json1_1RedshiftRetryOptions = function (output, context) {
    var contents = {
        __type: "RedshiftRetryOptions",
        DurationInSeconds: undefined
    };
    if (output.DurationInSeconds !== undefined &&
        output.DurationInSeconds !== null) {
        contents.DurationInSeconds = output.DurationInSeconds;
    }
    return contents;
};
var deserializeAws_json1_1ResourceInUseException = function (output, context) {
    var contents = {
        __type: "ResourceInUseException",
        message: undefined
    };
    if (output.message !== undefined && output.message !== null) {
        contents.message = output.message;
    }
    return contents;
};
var deserializeAws_json1_1ResourceNotFoundException = function (output, context) {
    var contents = {
        __type: "ResourceNotFoundException",
        message: undefined
    };
    if (output.message !== undefined && output.message !== null) {
        contents.message = output.message;
    }
    return contents;
};
var deserializeAws_json1_1S3DestinationDescription = function (output, context) {
    var contents = {
        __type: "S3DestinationDescription",
        BucketARN: undefined,
        BufferingHints: undefined,
        CloudWatchLoggingOptions: undefined,
        CompressionFormat: undefined,
        EncryptionConfiguration: undefined,
        ErrorOutputPrefix: undefined,
        Prefix: undefined,
        RoleARN: undefined
    };
    if (output.BucketARN !== undefined && output.BucketARN !== null) {
        contents.BucketARN = output.BucketARN;
    }
    if (output.BufferingHints !== undefined && output.BufferingHints !== null) {
        contents.BufferingHints = deserializeAws_json1_1BufferingHints(output.BufferingHints, context);
    }
    if (output.CloudWatchLoggingOptions !== undefined &&
        output.CloudWatchLoggingOptions !== null) {
        contents.CloudWatchLoggingOptions = deserializeAws_json1_1CloudWatchLoggingOptions(output.CloudWatchLoggingOptions, context);
    }
    if (output.CompressionFormat !== undefined &&
        output.CompressionFormat !== null) {
        contents.CompressionFormat = output.CompressionFormat;
    }
    if (output.EncryptionConfiguration !== undefined &&
        output.EncryptionConfiguration !== null) {
        contents.EncryptionConfiguration = deserializeAws_json1_1EncryptionConfiguration(output.EncryptionConfiguration, context);
    }
    if (output.ErrorOutputPrefix !== undefined &&
        output.ErrorOutputPrefix !== null) {
        contents.ErrorOutputPrefix = output.ErrorOutputPrefix;
    }
    if (output.Prefix !== undefined && output.Prefix !== null) {
        contents.Prefix = output.Prefix;
    }
    if (output.RoleARN !== undefined && output.RoleARN !== null) {
        contents.RoleARN = output.RoleARN;
    }
    return contents;
};
var deserializeAws_json1_1SchemaConfiguration = function (output, context) {
    var contents = {
        __type: "SchemaConfiguration",
        CatalogId: undefined,
        DatabaseName: undefined,
        Region: undefined,
        RoleARN: undefined,
        TableName: undefined,
        VersionId: undefined
    };
    if (output.CatalogId !== undefined && output.CatalogId !== null) {
        contents.CatalogId = output.CatalogId;
    }
    if (output.DatabaseName !== undefined && output.DatabaseName !== null) {
        contents.DatabaseName = output.DatabaseName;
    }
    if (output.Region !== undefined && output.Region !== null) {
        contents.Region = output.Region;
    }
    if (output.RoleARN !== undefined && output.RoleARN !== null) {
        contents.RoleARN = output.RoleARN;
    }
    if (output.TableName !== undefined && output.TableName !== null) {
        contents.TableName = output.TableName;
    }
    if (output.VersionId !== undefined && output.VersionId !== null) {
        contents.VersionId = output.VersionId;
    }
    return contents;
};
var deserializeAws_json1_1Serializer = function (output, context) {
    var contents = {
        __type: "Serializer",
        OrcSerDe: undefined,
        ParquetSerDe: undefined
    };
    if (output.OrcSerDe !== undefined && output.OrcSerDe !== null) {
        contents.OrcSerDe = deserializeAws_json1_1OrcSerDe(output.OrcSerDe, context);
    }
    if (output.ParquetSerDe !== undefined && output.ParquetSerDe !== null) {
        contents.ParquetSerDe = deserializeAws_json1_1ParquetSerDe(output.ParquetSerDe, context);
    }
    return contents;
};
var deserializeAws_json1_1ServiceUnavailableException = function (output, context) {
    var contents = {
        __type: "ServiceUnavailableException",
        message: undefined
    };
    if (output.message !== undefined && output.message !== null) {
        contents.message = output.message;
    }
    return contents;
};
var deserializeAws_json1_1SourceDescription = function (output, context) {
    var contents = {
        __type: "SourceDescription",
        KinesisStreamSourceDescription: undefined
    };
    if (output.KinesisStreamSourceDescription !== undefined &&
        output.KinesisStreamSourceDescription !== null) {
        contents.KinesisStreamSourceDescription = deserializeAws_json1_1KinesisStreamSourceDescription(output.KinesisStreamSourceDescription, context);
    }
    return contents;
};
var deserializeAws_json1_1SplunkDestinationDescription = function (output, context) {
    var contents = {
        __type: "SplunkDestinationDescription",
        CloudWatchLoggingOptions: undefined,
        HECAcknowledgmentTimeoutInSeconds: undefined,
        HECEndpoint: undefined,
        HECEndpointType: undefined,
        HECToken: undefined,
        ProcessingConfiguration: undefined,
        RetryOptions: undefined,
        S3BackupMode: undefined,
        S3DestinationDescription: undefined
    };
    if (output.CloudWatchLoggingOptions !== undefined &&
        output.CloudWatchLoggingOptions !== null) {
        contents.CloudWatchLoggingOptions = deserializeAws_json1_1CloudWatchLoggingOptions(output.CloudWatchLoggingOptions, context);
    }
    if (output.HECAcknowledgmentTimeoutInSeconds !== undefined &&
        output.HECAcknowledgmentTimeoutInSeconds !== null) {
        contents.HECAcknowledgmentTimeoutInSeconds =
            output.HECAcknowledgmentTimeoutInSeconds;
    }
    if (output.HECEndpoint !== undefined && output.HECEndpoint !== null) {
        contents.HECEndpoint = output.HECEndpoint;
    }
    if (output.HECEndpointType !== undefined && output.HECEndpointType !== null) {
        contents.HECEndpointType = output.HECEndpointType;
    }
    if (output.HECToken !== undefined && output.HECToken !== null) {
        contents.HECToken = output.HECToken;
    }
    if (output.ProcessingConfiguration !== undefined &&
        output.ProcessingConfiguration !== null) {
        contents.ProcessingConfiguration = deserializeAws_json1_1ProcessingConfiguration(output.ProcessingConfiguration, context);
    }
    if (output.RetryOptions !== undefined && output.RetryOptions !== null) {
        contents.RetryOptions = deserializeAws_json1_1SplunkRetryOptions(output.RetryOptions, context);
    }
    if (output.S3BackupMode !== undefined && output.S3BackupMode !== null) {
        contents.S3BackupMode = output.S3BackupMode;
    }
    if (output.S3DestinationDescription !== undefined &&
        output.S3DestinationDescription !== null) {
        contents.S3DestinationDescription = deserializeAws_json1_1S3DestinationDescription(output.S3DestinationDescription, context);
    }
    return contents;
};
var deserializeAws_json1_1SplunkRetryOptions = function (output, context) {
    var contents = {
        __type: "SplunkRetryOptions",
        DurationInSeconds: undefined
    };
    if (output.DurationInSeconds !== undefined &&
        output.DurationInSeconds !== null) {
        contents.DurationInSeconds = output.DurationInSeconds;
    }
    return contents;
};
var deserializeAws_json1_1StartDeliveryStreamEncryptionOutput = function (output, context) {
    var contents = {
        __type: "StartDeliveryStreamEncryptionOutput"
    };
    return contents;
};
var deserializeAws_json1_1StopDeliveryStreamEncryptionOutput = function (output, context) {
    var contents = {
        __type: "StopDeliveryStreamEncryptionOutput"
    };
    return contents;
};
var deserializeAws_json1_1Tag = function (output, context) {
    var contents = {
        __type: "Tag",
        Key: undefined,
        Value: undefined
    };
    if (output.Key !== undefined && output.Key !== null) {
        contents.Key = output.Key;
    }
    if (output.Value !== undefined && output.Value !== null) {
        contents.Value = output.Value;
    }
    return contents;
};
var deserializeAws_json1_1TagDeliveryStreamOutput = function (output, context) {
    var contents = {
        __type: "TagDeliveryStreamOutput"
    };
    return contents;
};
var deserializeAws_json1_1UntagDeliveryStreamOutput = function (output, context) {
    var contents = {
        __type: "UntagDeliveryStreamOutput"
    };
    return contents;
};
var deserializeAws_json1_1UpdateDestinationOutput = function (output, context) {
    var contents = {
        __type: "UpdateDestinationOutput"
    };
    return contents;
};
var deserializeMetadata = function (output) { return ({
    httpStatusCode: output.statusCode,
    httpHeaders: output.headers,
    requestId: output.headers["x-amzn-requestid"]
}); };
// Collect low-level response body stream to Uint8Array.
var collectBody = function (streamBody, context) {
    if (streamBody instanceof Uint8Array) {
        return Promise.resolve(streamBody);
    }
    return (context.streamCollector(streamBody) || Promise.resolve(new Uint8Array()));
};
// Encode Uint8Array data into string with utf-8.
var collectBodyString = function (streamBody, context) {
    return collectBody(streamBody, context).then(function (body) {
        return context.utf8Encoder(body);
    });
};
var buildHttpRpcRequest = function (context, headers, path, resolvedHostname, body) {
    var contents = __assign(__assign({}, context.endpoint), { protocol: "https", method: "POST", path: path, headers: headers });
    if (resolvedHostname !== undefined) {
        contents.hostname = resolvedHostname;
    }
    if (body !== undefined) {
        contents.body = body;
    }
    return new __HttpRequest(contents);
};
var parseBody = function (streamBody, context) {
    return collectBodyString(streamBody, context).then(function (encoded) {
        if (encoded.length) {
            return JSON.parse(encoded);
        }
        return {};
    });
};
