// @aws-amplify/ui-components/dist/esm-es5/amplify-confirm-sign-in_8.entry.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { r as registerInstance, h } from './index-04570516.js';
import { I18n, isEmpty, Logger, ConsoleLogger } from '@aws-amplify/core';
import { M as MfaOption, C as ChallengeName, A as AuthState, a as AuthFormField } from './auth-types-78df304e.js';
import { T as Translations } from './Translations-bcdbf36c.js';
import { Auth } from '@aws-amplify/auth';
import { N as NO_AUTH_MODULE_FOUND, C as COUNTRY_DIAL_CODE_DEFAULT, e as COUNTRY_DIAL_CODE_SUFFIX, f as PHONE_SUFFIX } from './constants-206817b9.js';
import { d as dispatchAuthStateChangeEvent, a as dispatchToastHubEvent, c as checkUsernameAlias, b as composePhoneNumberInput } from './helpers-5e56336b.js';
var AmplifyConfirmSignIn = /** @class */ (function () {
    function class_1(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        /** Fires when confirm sign in form is submitted */
        this.handleSubmit = function (event) { return _this.confirm(event); };
        /** Used for header text in confirm sign in component */
        this.headerText = I18n.get(Translations.CONFIRM_SMS_CODE);
        /** Used for the submit button text in confirm sign in component */
        this.submitButtonText = I18n.get(Translations.CONFIRM);
        /** Auth state change handler for this component */
        this.handleAuthStateChange = dispatchAuthStateChangeEvent;
        /**
         * Form fields allows you to utilize our pre-built components such as username field, code field, password field, email field, etc.
         * by passing an array of strings that you would like the order of the form to be in. If you need more customization, such as changing
         * text for a label or adjust a placeholder, you can follow the structure below in order to do just that.
         * ```
         * [
         *  {
         *    type: 'username'|'password'|'email'|'code'|'default',
         *    label: string,
         *    placeholder: string,
         *    hint: string | Functional Component | null,
         *    required: boolean
         *  }
         * ]
         * ```
         */
        this.formFields = [
            {
                type: 'code',
                required: true,
                handleInputChange: function (event) { return _this.handleCodeChange(event); },
            },
        ];
        /** The MFA option to confirm with */
        this.mfaOption = MfaOption.SMS;
        /* Whether or not the confirm-sign-in component is loading */
        this.loading = false;
    }
    class_1.prototype.componentWillLoad = function () {
        if (this.user && this.user['challengeName'] === ChallengeName.SoftwareTokenMFA) {
            this.mfaOption = MfaOption.TOTP;
            // If header text is using default use TOTP string
            if (this.headerText === I18n.get(Translations.CONFIRM_SMS_CODE)) {
                this.headerText = I18n.get(Translations.CONFIRM_TOTP_CODE);
            }
        }
    };
    class_1.prototype.handleCodeChange = function (event) {
        this.code = event.target.value;
    };
    class_1.prototype.checkContact = function (user) {
        var _this = this;
        if (!Auth || typeof Auth.verifiedContact !== 'function') {
            throw new Error(NO_AUTH_MODULE_FOUND);
        }
        Auth.verifiedContact(user).then(function (data) {
            if (!isEmpty(data.verified)) {
                _this.handleAuthStateChange(AuthState.SignedIn, user);
            }
            else {
                user = Object.assign(user, data);
                _this.handleAuthStateChange(AuthState.VerifyContact, user);
            }
        });
    };
    class_1.prototype.confirm = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            var mfaType, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (event) {
                            event.preventDefault();
                        }
                        mfaType = this.user['challengeName'] === ChallengeName.SoftwareTokenMFA ? ChallengeName.SoftwareTokenMFA : null;
                        if (!Auth || typeof Auth.confirmSignIn !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        this.loading = true;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, 4, 5]);
                        return [4 /*yield*/, Auth.confirmSignIn(this.user, this.code, mfaType)];
                    case 2:
                        _a.sent();
                        this.checkContact(this.user);
                        return [3 /*break*/, 5];
                    case 3:
                        error_1 = _a.sent();
                        dispatchToastHubEvent(error_1);
                        return [3 /*break*/, 5];
                    case 4:
                        this.loading = false;
                        return [7 /*endfinally*/];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    class_1.prototype.render = function () {
        var _this = this;
        return (h("amplify-form-section", { headerText: this.headerText, handleSubmit: this.handleSubmit, submitButtonText: this.submitButtonText, loading: this.loading, secondaryFooterContent: h("span", null, h("amplify-button", { variant: "anchor", onClick: function () { return _this.handleAuthStateChange(AuthState.SignIn); } }, I18n.get(Translations.BACK_TO_SIGN_IN))) }, h("amplify-auth-fields", { formFields: this.formFields })));
    };
    return class_1;
}());
var AmplifyConfirmSignUp = /** @class */ (function () {
    function class_2(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        /** Fires when sign up form is submitted */
        this.handleSubmit = function (event) { return _this.confirmSignUp(event); };
        /** Used for header text in confirm sign up component */
        this.headerText = I18n.get(Translations.CONFIRM_SIGN_UP_HEADER_TEXT);
        /** Used for the submit button text in confirm sign up component */
        this.submitButtonText = I18n.get(Translations.CONFIRM_SIGN_UP_SUBMIT_BUTTON_TEXT);
        /** Auth state change handler for this components
         * e.g. SignIn -> 'Create Account' link -> SignUp
         */
        this.handleAuthStateChange = dispatchAuthStateChangeEvent;
        /** Username Alias is used to setup authentication with `username`, `email` or `phone_number`  */
        this.usernameAlias = 'username';
        this.loading = false;
        this.userInput = this.user ? this.user.username : null;
        this._signUpAttrs = this.user && this.user.signUpAttrs ? this.user.signUpAttrs : null;
    }
    class_2.prototype.componentWillLoad = function () {
        var _this = this;
        checkUsernameAlias(this.usernameAlias);
        this.formFields = [
            {
                type: "" + this.usernameAlias,
                required: true,
                value: this.userInput,
                disabled: this.userInput ? true : false,
            },
            {
                type: 'code',
                label: I18n.get(Translations.CONFIRM_SIGN_UP_CODE_LABEL),
                placeholder: I18n.get(Translations.CONFIRM_SIGN_UP_CODE_PLACEHOLDER),
                required: true,
                hint: (h("div", null, I18n.get(Translations.CONFIRM_SIGN_UP_LOST_CODE), ' ', h("amplify-button", { variant: "anchor", onClick: function () { return _this.resendConfirmCode(); } }, I18n.get(Translations.CONFIRM_SIGN_UP_RESEND_CODE)))),
                handleInputChange: function (event) { return _this.handleCodeChange(event); },
            },
        ];
    };
    class_2.prototype.handleCodeChange = function (event) {
        this.code = event.target.value;
    };
    class_2.prototype.resendConfirmCode = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (event) {
                            event.preventDefault();
                        }
                        if (!Auth || typeof Auth.resendSignUp !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        if (!this.userInput)
                            throw new Error('Username can not be empty');
                        return [4 /*yield*/, Auth.resendSignUp(this.userInput)];
                    case 2:
                        _a.sent();
                        this.handleAuthStateChange(AuthState.ConfirmSignUp);
                        return [3 /*break*/, 4];
                    case 3:
                        error_2 = _a.sent();
                        dispatchToastHubEvent(error_2);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    // TODO: Add validation
    // TODO: Prefix
    class_2.prototype.confirmSignUp = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            var confirmSignUpResult, user, _a, error_3;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (event) {
                            event.preventDefault();
                        }
                        if (!Auth || typeof Auth.confirmSignUp !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        this.loading = true;
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 5, 6, 7]);
                        return [4 /*yield*/, Auth.confirmSignUp(this.userInput, this.code)];
                    case 2:
                        confirmSignUpResult = _b.sent();
                        _a = confirmSignUpResult && this._signUpAttrs;
                        if (!_a) return [3 /*break*/, 4];
                        return [4 /*yield*/, Auth.signIn(this.userInput, this._signUpAttrs.password)];
                    case 3:
                        _a = (_b.sent());
                        _b.label = 4;
                    case 4:
                        user = _a;
                        this.handleAuthStateChange(AuthState.SignedIn, user);
                        return [3 /*break*/, 7];
                    case 5:
                        error_3 = _b.sent();
                        dispatchToastHubEvent(error_3);
                        return [3 /*break*/, 7];
                    case 6:
                        this.loading = false;
                        return [7 /*endfinally*/];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    class_2.prototype.render = function () {
        var _this = this;
        return (h("amplify-form-section", { headerText: this.headerText, submitButtonText: this.submitButtonText, handleSubmit: this.handleSubmit, secondaryFooterContent: h("div", null, h("span", null, h("amplify-button", { variant: "anchor", onClick: function () { return _this.handleAuthStateChange(AuthState.SignIn); } }, I18n.get(Translations.BACK_TO_SIGN_IN)))) }, h("amplify-auth-fields", { formFields: this.formFields })));
    };
    return class_2;
}());
var logger = new Logger('ForgotPassword');
var AmplifyForgotPassword = /** @class */ (function () {
    function class_3(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        /** The header text of the forgot password section */
        this.headerText = I18n.get(Translations.RESET_YOUR_PASSWORD);
        /** The text displayed inside of the submit button for the form */
        this.submitButtonText = I18n.get(Translations.SEND_CODE);
        /** The function called when making a request to reset password */
        this.handleSend = function (event) { return _this.send(event); };
        /** The function called when submitting a new password */
        this.handleSubmit = function (event) { return _this.submit(event); };
        /** Auth state change handler for this component */
        this.handleAuthStateChange = dispatchAuthStateChangeEvent;
        /** Username Alias is used to setup authentication with `username`, `email` or `phone_number`  */
        this.usernameAlias = 'username';
        this.delivery = null;
        this.loading = false;
        this.phoneNumber = {
            countryDialCodeValue: COUNTRY_DIAL_CODE_DEFAULT,
            phoneNumberValue: null,
        };
    }
    class_3.prototype.componentWillLoad = function () {
        var _this = this;
        checkUsernameAlias(this.usernameAlias);
        switch (this.usernameAlias) {
            case 'email':
                this.formFields = [
                    {
                        type: 'email',
                        required: true,
                        handleInputChange: function (event) { return _this.handleEmailChange(event); },
                        inputProps: {
                            'data-test': 'forgot-password-email-input',
                        },
                    },
                ];
                break;
            case 'phone_number':
                this.formFields = [
                    {
                        type: 'phone_number',
                        required: true,
                        handleInputChange: function (event) { return _this.handlePhoneNumberChange(event); },
                        inputProps: {
                            'data-test': 'forgot-password-phone-number-input',
                        },
                    },
                ];
                break;
            case 'username':
            default:
                this.formFields = [
                    {
                        type: 'username',
                        required: true,
                        handleInputChange: function (event) { return _this.handleUsernameChange(event); },
                        value: this.username,
                        inputProps: {
                            'data-test': 'forgot-password-username-input',
                        },
                    },
                ];
                break;
        }
    };
    class_3.prototype.handleUsernameChange = function (event) {
        this.username = event.target.value;
    };
    class_3.prototype.handleEmailChange = function (event) {
        this.email = event.target.value;
    };
    class_3.prototype.handlePhoneNumberChange = function (event) {
        var name = event.target.name;
        var value = event.target.value;
        /** Cognito expects to have a string be passed when signing up. Since the Select input is separate
         * input from the phone number input, we need to first capture both components values and combined
         * them together.
         */
        if (name === COUNTRY_DIAL_CODE_SUFFIX) {
            this.phoneNumber.countryDialCodeValue = value;
        }
        if (name === PHONE_SUFFIX) {
            this.phoneNumber.phoneNumberValue = value;
        }
    };
    class_3.prototype.handlePasswordChange = function (event) {
        this.password = event.target.value;
    };
    class_3.prototype.handleCodeChange = function (event) {
        this.code = event.target.value;
    };
    class_3.prototype.send = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            var data, error_4;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (event) {
                            event.preventDefault();
                        }
                        if (!Auth || typeof Auth.forgotPassword !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        this.loading = true;
                        switch (this.usernameAlias) {
                            case 'email':
                                this.username = this.email;
                                break;
                            case 'phone_number':
                                this.username = composePhoneNumberInput(this.phoneNumber);
                                break;
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, 4, 5]);
                        return [4 /*yield*/, Auth.forgotPassword(this.username)];
                    case 2:
                        data = _a.sent();
                        logger.debug(data);
                        this.formFields = [
                            {
                                type: 'code',
                                required: true,
                                handleInputChange: function (event) { return _this.handleCodeChange(event); },
                                value: this.code,
                                inputProps: {
                                    'data-test': 'forgot-password-code-input',
                                },
                            },
                            {
                                type: 'password',
                                required: true,
                                handleInputChange: function (event) { return _this.handlePasswordChange(event); },
                                label: 'New password',
                                placeholder: 'Enter your new password',
                                value: this.password,
                            },
                        ];
                        this.delivery = data.CodeDeliveryDetails;
                        return [3 /*break*/, 5];
                    case 3:
                        error_4 = _a.sent();
                        dispatchToastHubEvent(error_4);
                        return [3 /*break*/, 5];
                    case 4:
                        this.loading = false;
                        return [7 /*endfinally*/];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    class_3.prototype.submit = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            var data, error_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (event) {
                            event.preventDefault();
                        }
                        if (!Auth || typeof Auth.forgotPasswordSubmit !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        this.loading = true;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, 4, 5]);
                        return [4 /*yield*/, Auth.forgotPasswordSubmit(this.username, this.code, this.password)];
                    case 2:
                        data = _a.sent();
                        logger.debug(data);
                        this.handleAuthStateChange(AuthState.SignIn);
                        this.delivery = null;
                        return [3 /*break*/, 5];
                    case 3:
                        error_5 = _a.sent();
                        dispatchToastHubEvent(error_5);
                        return [3 /*break*/, 5];
                    case 4:
                        this.loading = false;
                        return [7 /*endfinally*/];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    class_3.prototype.render = function () {
        var _this = this;
        var submitFn = this.delivery ? function (event) { return _this.handleSubmit(event); } : function (event) { return _this.handleSend(event); };
        return (h("amplify-form-section", { headerText: this.headerText, handleSubmit: submitFn, loading: this.loading, secondaryFooterContent: h("amplify-button", { variant: "anchor", onClick: function () { return _this.handleAuthStateChange(AuthState.SignIn); }, "data-test": "forgot-password-back-to-sign-in-link" }, I18n.get(Translations.BACK_TO_SIGN_IN)), testDataPrefix: 'forgot-password' }, h("amplify-auth-fields", { formFields: this.formFields })));
    };
    return class_3;
}());
var logger$1 = new ConsoleLogger('amplify-require-new-password');
var AmplifyRequireNewPassword = /** @class */ (function () {
    function class_4(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        /** The header text of the forgot password section */
        this.headerText = I18n.get(Translations.CHANGE_PASSWORD);
        /** The text displayed inside of the submit button for the form */
        this.submitButtonText = I18n.get(Translations.CHANGE_PASSWORD_ACTION);
        /** The function called when submitting a new password */
        this.handleSubmit = function (event) { return _this.completeNewPassword(event); };
        /** Auth state change handler for this component */
        this.handleAuthStateChange = dispatchAuthStateChangeEvent;
        /** The form fields displayed inside of the forgot password form */
        this.formFields = [
            {
                type: AuthFormField.Password,
                required: true,
                handleInputChange: function (event) { return _this.handlePasswordChange(event); },
                label: I18n.get(Translations.NEW_PASSWORD_LABEL),
                placeholder: I18n.get(Translations.NEW_PASSWORD_PLACEHOLDER),
            },
        ];
        this.loading = false;
    }
    class_4.prototype.handlePasswordChange = function (event) {
        this.password = event.target.value;
    };
    class_4.prototype.checkContact = function (user) {
        return __awaiter(this, void 0, void 0, function () {
            var data, error_6;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!Auth || typeof Auth.verifiedContact !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, Auth.verifiedContact(user)];
                    case 2:
                        data = _a.sent();
                        if (!isEmpty(data.verified)) {
                            this.handleAuthStateChange(AuthState.SignedIn, user);
                        }
                        else {
                            user = Object.assign(user, data);
                            this.handleAuthStateChange(AuthState.VerifyContact, user);
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        error_6 = _a.sent();
                        dispatchToastHubEvent(error_6);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    class_4.prototype.completeNewPassword = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            var requiredAttributes, user, error_7;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (event) {
                            event.preventDefault();
                        }
                        if (!Auth || typeof Auth.completeNewPassword !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        this.loading = true;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, 4, 5]);
                        requiredAttributes = this.user.challengeParam.requiredAttributes;
                        return [4 /*yield*/, Auth.completeNewPassword(this.user, this.password, requiredAttributes)];
                    case 2:
                        user = _a.sent();
                        logger$1.debug('complete new password', user);
                        switch (user.challengeName) {
                            case ChallengeName.SMSMFA:
                                this.handleAuthStateChange(AuthState.ConfirmSignIn, user);
                                break;
                            case ChallengeName.MFASetup:
                                logger$1.debug('TOTP setup', user.challengeParam);
                                this.handleAuthStateChange(AuthState.TOTPSetup, user);
                                break;
                            default:
                                this.checkContact(user);
                        }
                        return [3 /*break*/, 5];
                    case 3:
                        error_7 = _a.sent();
                        dispatchToastHubEvent(error_7);
                        return [3 /*break*/, 5];
                    case 4:
                        this.loading = false;
                        return [7 /*endfinally*/];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    class_4.prototype.render = function () {
        var _this = this;
        return (h("amplify-form-section", { headerText: this.headerText, handleSubmit: this.handleSubmit, loading: this.loading, secondaryFooterContent: h("amplify-button", { variant: "anchor", onClick: function () { return _this.handleAuthStateChange(AuthState.SignIn); } }, I18n.get(Translations.BACK_TO_SIGN_IN)) }, h("amplify-auth-fields", { formFields: this.formFields })));
    };
    return class_4;
}());
var amplifySignInCss = ":host{--footer-size:var(--amplify-text-sm);--footer-color:var(--amplify-grey);--footer-font-family:var(--amplify-font-family)}.sign-in-form-footer{font-family:var(--footer-font-family);font-weight:300;font-size:var(--footer-size);color:var(--footer-color);margin-bottom:10px;display:-ms-flexbox;display:flex;-ms-flex-direction:column-reverse;flex-direction:column-reverse;-ms-flex-align:center;align-items:center;-ms-flex-pack:distribute;justify-content:space-around}@media (min-width: 672px){.sign-in-form-footer{display:-ms-flexbox;display:flex;-ms-flex-direction:row;flex-direction:row;-ms-flex-pack:justify;justify-content:space-between;-ms-flex-align:baseline;align-items:baseline}}.sign-in-form-footer *+*{margin-bottom:15px}";
var logger$2 = new Logger('SignIn');
var AmplifySignIn = /** @class */ (function () {
    function class_5(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        /** Fires when sign in form is submitted */
        this.handleSubmit = function (event) { return _this.signIn(event); };
        /** Used for header text in sign in component */
        this.headerText = I18n.get(Translations.SIGN_IN_HEADER_TEXT);
        /** Used for the submit button text in sign in component */
        this.submitButtonText = I18n.get(Translations.SIGN_IN_ACTION);
        /** Auth state change handler for this component */
        this.handleAuthStateChange = dispatchAuthStateChangeEvent;
        /** Username Alias is used to setup authentication with `username`, `email` or `phone_number`  */
        this.usernameAlias = 'username';
        /**
         * Form fields allows you to utilize our pre-built components such as username field, code field, password field, email field, etc.
         * by passing an array of strings that you would like the order of the form to be in. If you need more customization, such as changing
         * text for a label or adjust a placeholder, you can follow the structure below in order to do just that.
         * ```
         * [
         *  {
         *    type: 'username'|'password'|'email'|'code'|'default',
         *    label: string,
         *    placeholder: string,
         *    hint: string | Functional Component | null,
         *    required: boolean
         *  }
         * ]
         * ```
         */
        this.formFields = [];
        /* Whether or not the sign-in component is loading */
        this.loading = false;
        /* The username value in the sign-in form */
        this.username = '';
        /* The password value in the sign-in form */
        this.password = '';
        this.phoneNumber = {
            countryDialCodeValue: COUNTRY_DIAL_CODE_DEFAULT,
            phoneNumberValue: null,
        };
        this.userInput = this.username;
    }
    class_5.prototype.handleUsernameChange = function (event) {
        this.username = event.target.value;
    };
    class_5.prototype.handlePasswordChange = function (event) {
        this.password = event.target.value;
    };
    class_5.prototype.handleEmailChange = function (event) {
        this.email = event.target.value;
    };
    class_5.prototype.handlePhoneNumberChange = function (event) {
        var name = event.target.name;
        var value = event.target.value;
        /** Cognito expects to have a string be passed when signing up. Since the Select input is separate
         * input from the phone number input, we need to first capture both components values and combined
         * them together.
         */
        if (name === COUNTRY_DIAL_CODE_SUFFIX) {
            this.phoneNumber.countryDialCodeValue = value;
        }
        if (name === PHONE_SUFFIX) {
            this.phoneNumber.phoneNumberValue = value;
        }
    };
    class_5.prototype.checkContact = function (user) {
        var _this = this;
        if (!Auth || typeof Auth.verifiedContact !== 'function') {
            throw new Error(NO_AUTH_MODULE_FOUND);
        }
        Auth.verifiedContact(user).then(function (data) {
            if (!isEmpty(data.verified)) {
                _this.handleAuthStateChange(AuthState.SignedIn, user);
            }
            else {
                user = Object.assign(user, data);
                _this.handleAuthStateChange(AuthState.VerifyContact, user);
            }
        });
    };
    class_5.prototype.signIn = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            var user, error_8;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        // avoid submitting the form
                        if (event) {
                            event.preventDefault();
                        }
                        if (!Auth || typeof Auth.signIn !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        this.loading = true;
                        switch (this.usernameAlias) {
                            case 'email':
                                this.userInput = this.email;
                                break;
                            case 'phone_number':
                                this.userInput = composePhoneNumberInput(this.phoneNumber);
                                break;
                            case 'username':
                            default:
                                this.userInput = this.username;
                                break;
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, 4, 5]);
                        return [4 /*yield*/, Auth.signIn(this.userInput, this.password)];
                    case 2:
                        user = _a.sent();
                        logger$2.debug(user);
                        if (user.challengeName === ChallengeName.SMSMFA || user.challengeName === ChallengeName.SoftwareTokenMFA) {
                            logger$2.debug('confirm user with ' + user.challengeName);
                            this.handleAuthStateChange(AuthState.ConfirmSignIn, user);
                        }
                        else if (user.challengeName === ChallengeName.NewPasswordRequired) {
                            logger$2.debug('require new password', user.challengeParam);
                            this.handleAuthStateChange(AuthState.ResetPassword, user);
                        }
                        else if (user.challengeName === ChallengeName.MFASetup) {
                            logger$2.debug('TOTP setup', user.challengeParam);
                            this.handleAuthStateChange(AuthState.TOTPSetup, user);
                        }
                        else if (user.challengeName === ChallengeName.CustomChallenge &&
                            user.challengeParam &&
                            user.challengeParam.trigger === 'true') {
                            logger$2.debug('custom challenge', user.challengeParam);
                            this.handleAuthStateChange(AuthState.CustomConfirmSignIn, user);
                        }
                        else {
                            this.checkContact(user);
                        }
                        return [3 /*break*/, 5];
                    case 3:
                        error_8 = _a.sent();
                        dispatchToastHubEvent(error_8);
                        if (error_8.code === 'UserNotConfirmedException') {
                            logger$2.debug('the user is not confirmed');
                            this.handleAuthStateChange(AuthState.ConfirmSignUp, { username: this.userInput });
                        }
                        else if (error_8.code === 'PasswordResetRequiredException') {
                            logger$2.debug('the user requires a new password');
                            this.handleAuthStateChange(AuthState.ForgotPassword, { username: this.userInput });
                        }
                        return [3 /*break*/, 5];
                    case 4:
                        this.loading = false;
                        return [7 /*endfinally*/];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    class_5.prototype.componentWillLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            var formFieldInputs;
            var _this = this;
            return __generator(this, function (_a) {
                checkUsernameAlias(this.usernameAlias);
                formFieldInputs = [];
                switch (this.usernameAlias) {
                    case 'email':
                        formFieldInputs.push({
                            type: 'email',
                            required: true,
                            handleInputChange: function (event) { return _this.handleEmailChange(event); },
                            inputProps: {
                                'data-test': 'sign-in-email-input',
                            },
                        });
                        break;
                    case 'phone_number':
                        formFieldInputs.push({
                            type: 'phone_number',
                            required: true,
                            handleInputChange: function (event) { return _this.handlePhoneNumberChange(event); },
                            inputProps: {
                                'data-test': 'sign-in-phone-number-input',
                            },
                        });
                        break;
                    case 'username':
                    default:
                        formFieldInputs.push({
                            type: 'username',
                            required: true,
                            handleInputChange: function (event) { return _this.handleUsernameChange(event); },
                            inputProps: {
                                'data-test': 'sign-in-username-input',
                            },
                        });
                        break;
                }
                formFieldInputs.push({
                    type: 'password',
                    hint: (h("div", null, I18n.get(Translations.FORGOT_PASSWORD_TEXT), ' ', h("amplify-button", { variant: "anchor", onClick: function () { return _this.handleAuthStateChange(AuthState.ForgotPassword); }, "data-test": "sign-in-forgot-password-link" }, I18n.get(Translations.RESET_PASSWORD_TEXT)))),
                    required: true,
                    handleInputChange: function (event) { return _this.handlePasswordChange(event); },
                    inputProps: {
                        'data-test': 'sign-in-password-input',
                    },
                });
                this.formFields = formFieldInputs;
                return [2 /*return*/];
            });
        });
    };
    class_5.prototype.render = function () {
        var _this = this;
        return (h("amplify-form-section", { headerText: this.headerText, handleSubmit: this.handleSubmit, testDataPrefix: 'sign-in' }, h("amplify-federated-buttons", { handleAuthStateChange: this.handleAuthStateChange, federated: this.federated }), !isEmpty(this.federated) && h("amplify-strike", null, "or"), h("amplify-auth-fields", { formFields: this.formFields }), h("div", { slot: "amplify-form-section-footer", class: "sign-in-form-footer" }, h("span", null, I18n.get(Translations.NO_ACCOUNT_TEXT), ' ', h("amplify-button", { variant: "anchor", onClick: function () { return _this.handleAuthStateChange(AuthState.SignUp); }, "data-test": "sign-in-create-account-link" }, I18n.get(Translations.CREATE_ACCOUNT_TEXT))), h("amplify-button", { type: "submit", disabled: this.loading, "data-test": "sign-in-sign-in-button" }, h("amplify-loading-spinner", { style: { display: this.loading ? 'initial' : 'none' } }), h("span", { style: { display: this.loading ? 'none' : 'initial' } }, this.submitButtonText)))));
    };
    return class_5;
}());
AmplifySignIn.style = amplifySignInCss;
var amplifySignUpCss = ":host{--footer-font-family:var(--amplify-font-family);--footer-font-size:var(--amplify-text-sm);--footer-color:var(--amplify-grey)}.sign-up-form-footer{font-family:var(--footer-font-family);font-weight:300;font-size:var(--footer-font-size);color:var(--footer-color);display:-ms-flexbox;display:flex;-ms-flex-direction:column-reverse;flex-direction:column-reverse;-ms-flex-align:center;align-items:center;-ms-flex-pack:distribute;justify-content:space-around}@media (min-width: 672px){.sign-up-form-footer{display:-ms-flexbox;display:flex;-ms-flex-direction:row;flex-direction:row;-ms-flex-pack:justify;justify-content:space-between;-ms-flex-align:baseline;align-items:baseline}}.sign-up-form-footer *+*{margin-bottom:15px}";
var AmplifySignUp = /** @class */ (function () {
    function class_6(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        /** Fires when sign up form is submitted */
        this.handleSubmit = function (event) { return _this.signUp(event); };
        /** Used for header text in sign up component */
        this.headerText = I18n.get(Translations.SIGN_UP_HEADER_TEXT);
        /** Used for the submit button text in sign up component */
        this.submitButtonText = I18n.get(Translations.SIGN_UP_SUBMIT_BUTTON_TEXT);
        /** Used for the submit button text in sign up component */
        this.haveAccountText = I18n.get(Translations.SIGN_UP_HAVE_ACCOUNT_TEXT);
        /** Used for the submit button text in sign up component */
        this.signInText = I18n.get(Translations.SIGN_IN_TEXT);
        /**
         * Form fields allows you to utilize our pre-built components such as username field, code field, password field, email field, etc.
         * by passing an array of strings that you would like the order of the form to be in. If you need more customization, such as changing
         * text for a label or adjust a placeholder, you can follow the structure below in order to do just that.
         * ```
         * [
         *  {
         *    type: 'username'|'password'|'email'|'code'|'default',
         *    label: string,
         *    placeholder: string,
         *    hint: string | Functional Component | null,
         *    required: boolean
         *  }
         * ]
         * ```
         */
        this.formFields = [];
        /** Auth state change handler for this component
         * e.g. SignIn -> 'Create Account' link -> SignUp
         */
        this.handleAuthStateChange = dispatchAuthStateChangeEvent;
        /** Username Alias is used to setup authentication with `username`, `email` or `phone_number`  */
        this.usernameAlias = 'username';
        this.loading = false;
        this.phoneNumber = {
            countryDialCodeValue: COUNTRY_DIAL_CODE_DEFAULT,
            phoneNumberValue: null,
        };
    }
    class_6.prototype.handleUsernameChange = function (event) {
        this.username = event.target.value;
    };
    class_6.prototype.handlePasswordChange = function (event) {
        this.password = event.target.value;
    };
    class_6.prototype.handleEmailChange = function (event) {
        this.email = event.target.value;
    };
    class_6.prototype.handlePhoneNumberChange = function (event) {
        var name = event.target.name;
        var value = event.target.value;
        /** Cognito expects to have a string be passed when signing up. Since the Select input is separate
         * input from the phone number input, we need to first capture both components values and combined
         * them together.
         */
        if (name === COUNTRY_DIAL_CODE_SUFFIX) {
            this.phoneNumber.countryDialCodeValue = value;
        }
        if (name === PHONE_SUFFIX) {
            this.phoneNumber.phoneNumberValue = value;
        }
    };
    // TODO: Add validation
    // TODO: Prefix
    class_6.prototype.signUp = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            var signUpAttrs, data, error_9;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (event) {
                            event.preventDefault();
                        }
                        if (!Auth || typeof Auth.signUp !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        switch (this.usernameAlias) {
                            case 'email':
                                this.userInput = this.email;
                                break;
                            case 'phone_number':
                                this.userInput = composePhoneNumberInput(this.phoneNumber);
                                break;
                            case 'username':
                            default:
                                this.userInput = this.username;
                                break;
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        signUpAttrs = {
                            username: this.userInput,
                            password: this.password,
                            attributes: {
                                email: this.email,
                                phone_number: composePhoneNumberInput(this.phoneNumber),
                            },
                        };
                        return [4 /*yield*/, Auth.signUp(signUpAttrs)];
                    case 2:
                        data = _a.sent();
                        this.handleAuthStateChange(AuthState.ConfirmSignUp, Object.assign(Object.assign({}, data.user), { signUpAttrs: signUpAttrs }));
                        return [3 /*break*/, 4];
                    case 3:
                        error_9 = _a.sent();
                        dispatchToastHubEvent(error_9);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    class_6.prototype.componentWillLoad = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                checkUsernameAlias(this.usernameAlias);
                switch (this.usernameAlias) {
                    case 'email':
                        this.formFields = [
                            {
                                type: 'email',
                                placeholder: I18n.get(Translations.SIGN_UP_EMAIL_PLACEHOLDER),
                                required: true,
                                handleInputChange: function (event) { return _this.handleEmailChange(event); },
                                inputProps: {
                                    'data-test': 'sign-up-email-input',
                                },
                            },
                            {
                                type: 'password',
                                placeholder: I18n.get(Translations.SIGN_UP_PASSWORD_PLACEHOLDER),
                                required: true,
                                handleInputChange: function (event) { return _this.handlePasswordChange(event); },
                                inputProps: {
                                    'data-test': 'sign-up-password-input',
                                },
                            },
                            {
                                type: 'phone_number',
                                required: true,
                                handleInputChange: function (event) { return _this.handlePhoneNumberChange(event); },
                                inputProps: {
                                    'data-test': 'sign-up-phone-number-input',
                                },
                            },
                        ];
                        break;
                    case 'phone_number':
                        this.formFields = [
                            {
                                type: 'phone_number',
                                required: true,
                                handleInputChange: function (event) { return _this.handlePhoneNumberChange(event); },
                                inputProps: {
                                    'data-test': 'sign-up-phone-number-input',
                                },
                            },
                            {
                                type: 'password',
                                placeholder: I18n.get(Translations.SIGN_UP_PASSWORD_PLACEHOLDER),
                                required: true,
                                handleInputChange: function (event) { return _this.handlePasswordChange(event); },
                                inputProps: {
                                    'data-test': 'sign-up-password-input',
                                },
                            },
                            {
                                type: 'email',
                                placeholder: I18n.get(Translations.SIGN_UP_EMAIL_PLACEHOLDER),
                                required: true,
                                handleInputChange: function (event) { return _this.handleEmailChange(event); },
                                inputProps: {
                                    'data-test': 'sign-up-email-input',
                                },
                            },
                        ];
                        break;
                    case 'username':
                    default:
                        this.formFields = [
                            {
                                type: 'username',
                                placeholder: I18n.get(Translations.SIGN_UP_USERNAME_PLACEHOLDER),
                                required: true,
                                handleInputChange: function (event) { return _this.handleUsernameChange(event); },
                                inputProps: {
                                    'data-test': 'sign-up-username-input',
                                },
                            },
                            {
                                type: 'password',
                                placeholder: I18n.get(Translations.SIGN_UP_PASSWORD_PLACEHOLDER),
                                required: true,
                                handleInputChange: function (event) { return _this.handlePasswordChange(event); },
                                inputProps: {
                                    'data-test': 'sign-up-password-input',
                                },
                            },
                            {
                                type: 'email',
                                placeholder: I18n.get(Translations.SIGN_UP_EMAIL_PLACEHOLDER),
                                required: true,
                                handleInputChange: function (event) { return _this.handleEmailChange(event); },
                                inputProps: {
                                    'data-test': 'sign-up-email-input',
                                },
                            },
                            {
                                type: 'phone_number',
                                required: true,
                                handleInputChange: function (event) { return _this.handlePhoneNumberChange(event); },
                                inputProps: {
                                    'data-test': 'sign-up-phone-number-input',
                                },
                            },
                        ];
                        break;
                }
                return [2 /*return*/];
            });
        });
    };
    class_6.prototype.render = function () {
        var _this = this;
        return (h("amplify-form-section", { headerText: this.headerText, handleSubmit: this.handleSubmit, testDataPrefix: 'sign-up' }, h("amplify-auth-fields", { formFields: this.formFields }), h("div", { class: "sign-up-form-footer", slot: "amplify-form-section-footer" }, h("span", null, this.haveAccountText, ' ', h("amplify-button", { variant: "anchor", onClick: function () { return _this.handleAuthStateChange(AuthState.SignIn); }, "data-test": "sign-up-sign-in-link" }, this.signInText)), h("amplify-button", { type: "submit", "data-test": "sign-up-create-account-button" }, h("amplify-loading-spinner", { style: { display: this.loading ? 'initial' : 'none' } }), h("span", { style: { display: this.loading ? 'none' : 'initial' } }, this.submitButtonText)))));
    };
    return class_6;
}());
AmplifySignUp.style = amplifySignUpCss;
var amplifyToastCss = ":host{--background-color:var(--amplify-secondary-tint);--color:var(--amplify-white);--font-size:var(--amplify-text-sm);--font-family:var(--amplify-font-family);--close-icon-color:var(--amplify-white);--close-icon-hover:var(--amplify-red)}.toast-icon{padding-right:5px}.toast{display:-ms-flexbox;display:flex;-ms-flex-pack:justify;justify-content:space-between;position:absolute;top:0;left:0;width:100%;z-index:99;-webkit-box-shadow:0 0 5px 0 rgba(0, 0, 0, 0.3);box-shadow:0 0 5px 0 rgba(0, 0, 0, 0.3);padding:16px;background-color:var(--background-color);font-size:var(--font-size);color:var(--color);-webkit-box-sizing:border-box;box-sizing:border-box;border-radius:5px;font-family:var(--font-family)}.toast>span{margin-right:10px}.toast-close{margin-left:auto;-ms-flex-item-align:center;align-self:center;position:relative;width:18px;height:18px;overflow:hidden;cursor:pointer;background:transparent;border:none}.toast-close::before,.toast-close::after{content:\"\";position:absolute;height:2px;width:100%;top:50%;left:0;margin-top:-1px;background:var(--close-icon-color)}.toast-close:hover::before,.toast-close:hover::after{background:var(--close-icon-hover)}.toast-close::before{-webkit-transform:rotate(45deg);-moz-transform:rotate(45deg);-ms-transform:rotate(45deg);-o-transform:rotate(45deg);transform:rotate(45deg)}.toast-close::after{-webkit-transform:rotate(-45deg);-moz-transform:rotate(-45deg);-ms-transform:rotate(-45deg);-o-transform:rotate(-45deg);transform:rotate(-45deg)}";
var AmplifyToast = /** @class */ (function () {
    function AmplifyToast(hostRef) {
        registerInstance(this, hostRef);
        this.message = '';
    }
    /*
    TODO #170365145: Work on a helper function that will populate and
    update class colors for success / warning / failure messages
    */
    AmplifyToast.prototype.render = function () {
        return (h("div", { class: "toast" }, h("amplify-icon", { class: "toast-icon", name: "warning" }), this.message ? h("span", null, this.message) : null, h("slot", null), h("button", { class: "toast-close", onClick: this.handleClose })));
    };
    return AmplifyToast;
}());
AmplifyToast.style = amplifyToastCss;
var logger$3 = new Logger('AmplifyVerifyContact');
var AmplifyVerifyContact = /** @class */ (function () {
    function class_7(hostRef) {
        registerInstance(this, hostRef);
        /** Authentication state handler */
        this.handleAuthStateChange = dispatchAuthStateChangeEvent;
        this.loading = false;
    }
    class_7.prototype.handleSubmit = function (event) {
        event.preventDefault();
        var form = event.target;
        this.verifyAttr ? this.submit(form.code.value) : this.verify(form.contact.value);
    };
    class_7.prototype.submit = function (code) {
        return __awaiter(this, void 0, void 0, function () {
            var attr, data, error_10;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        attr = this.verifyAttr;
                        if (!Auth || typeof Auth.verifyCurrentUserAttributeSubmit !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, Auth.verifyCurrentUserAttributeSubmit(attr, code)];
                    case 2:
                        data = _a.sent();
                        logger$3.debug(data);
                        this.handleAuthStateChange(AuthState.SignedIn, this.user);
                        this.verifyAttr = null;
                        return [3 /*break*/, 4];
                    case 3:
                        error_10 = _a.sent();
                        logger$3.error(error_10);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    class_7.prototype.verify = function (contact) {
        return __awaiter(this, void 0, void 0, function () {
            var data, error_11;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!contact) {
                            logger$3.error('Neither Email nor Phone Number selected');
                            return [2 /*return*/];
                        }
                        if (!Auth || typeof Auth.verifyCurrentUserAttribute !== 'function') {
                            throw new Error(NO_AUTH_MODULE_FOUND);
                        }
                        this.loading = true;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, 4, 5]);
                        return [4 /*yield*/, Auth.verifyCurrentUserAttribute(contact)];
                    case 2:
                        data = _a.sent();
                        logger$3.debug(data);
                        this.verifyAttr = contact;
                        return [3 /*break*/, 5];
                    case 3:
                        error_11 = _a.sent();
                        logger$3.error(error_11);
                        return [3 /*break*/, 5];
                    case 4:
                        this.loading = false;
                        return [7 /*endfinally*/];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    class_7.prototype.renderSubmit = function () {
        return (h("div", null, h("amplify-input", { inputProps: {
                autocomplete: 'off',
            }, name: "code", placeholder: I18n.get(Translations.CODE_PLACEHOLDER) })));
    };
    class_7.prototype.renderVerify = function () {
        var user = this.user;
        if (!user) {
            logger$3.debug('No user to verify');
            return null;
        }
        var unverified = user.unverified;
        if (!unverified) {
            logger$3.debug('Unverified variable does not exist on user');
            return null;
        }
        var email = unverified.email, phone_number = unverified.phone_number;
        return (h("div", null, email && (h("amplify-radio-button", { label: I18n.get(Translations.VERIFY_CONTACT_EMAIL_LABEL), key: "email", name: "contact", value: "email" })), phone_number && (h("amplify-radio-button", { label: I18n.get(Translations.VERIFY_CONTACT_PHONE_LABEL), key: "phone_number", name: "contact", value: "phone_number" }))));
    };
    class_7.prototype.render = function () {
        var _this = this;
        return (h("amplify-form-section", { handleSubmit: function (event) { return _this.handleSubmit(event); }, headerText: I18n.get(Translations.VERIFY_CONTACT_HEADER_TEXT), loading: this.loading, secondaryFooterContent: h("span", null, h("amplify-button", { variant: "anchor", onClick: function () { return _this.handleAuthStateChange(AuthState.SignedIn, _this.user); } }, "Skip")), submitButtonText: this.verifyAttr
                ? I18n.get(Translations.VERIFY_CONTACT_SUBMIT_LABEL)
                : I18n.get(Translations.VERIFY_CONTACT_VERIFY_LABEL) }, this.verifyAttr ? this.renderSubmit() : this.renderVerify()));
    };
    return class_7;
}());
export { AmplifyConfirmSignIn as amplify_confirm_sign_in, AmplifyConfirmSignUp as amplify_confirm_sign_up, AmplifyForgotPassword as amplify_forgot_password, AmplifyRequireNewPassword as amplify_require_new_password, AmplifySignIn as amplify_sign_in, AmplifySignUp as amplify_sign_up, AmplifyToast as amplify_toast, AmplifyVerifyContact as amplify_verify_contact };
