// @aws-sdk/client-rekognition/dist/es/models/index.js
import { isa as __isa } from "@aws-sdk/smithy-client";
export var AccessDeniedException;
(function (AccessDeniedException) {
    function isa(o) {
        return __isa(o, "AccessDeniedException");
    }
    AccessDeniedException.isa = isa;
})(AccessDeniedException || (AccessDeniedException = {}));
export var AgeRange;
(function (AgeRange) {
    function isa(o) {
        return __isa(o, "AgeRange");
    }
    AgeRange.isa = isa;
})(AgeRange || (AgeRange = {}));
export var Asset;
(function (Asset) {
    function isa(o) {
        return __isa(o, "Asset");
    }
    Asset.isa = isa;
})(Asset || (Asset = {}));
export var Attribute;
(function (Attribute) {
    Attribute["ALL"] = "ALL";
    Attribute["DEFAULT"] = "DEFAULT";
})(Attribute || (Attribute = {}));
export var Beard;
(function (Beard) {
    function isa(o) {
        return __isa(o, "Beard");
    }
    Beard.isa = isa;
})(Beard || (Beard = {}));
export var BoundingBox;
(function (BoundingBox) {
    function isa(o) {
        return __isa(o, "BoundingBox");
    }
    BoundingBox.isa = isa;
})(BoundingBox || (BoundingBox = {}));
export var Celebrity;
(function (Celebrity) {
    function isa(o) {
        return __isa(o, "Celebrity");
    }
    Celebrity.isa = isa;
})(Celebrity || (Celebrity = {}));
export var CelebrityDetail;
(function (CelebrityDetail) {
    function isa(o) {
        return __isa(o, "CelebrityDetail");
    }
    CelebrityDetail.isa = isa;
})(CelebrityDetail || (CelebrityDetail = {}));
export var CelebrityRecognition;
(function (CelebrityRecognition) {
    function isa(o) {
        return __isa(o, "CelebrityRecognition");
    }
    CelebrityRecognition.isa = isa;
})(CelebrityRecognition || (CelebrityRecognition = {}));
export var CelebrityRecognitionSortBy;
(function (CelebrityRecognitionSortBy) {
    CelebrityRecognitionSortBy["ID"] = "ID";
    CelebrityRecognitionSortBy["TIMESTAMP"] = "TIMESTAMP";
})(CelebrityRecognitionSortBy || (CelebrityRecognitionSortBy = {}));
export var CompareFacesMatch;
(function (CompareFacesMatch) {
    function isa(o) {
        return __isa(o, "CompareFacesMatch");
    }
    CompareFacesMatch.isa = isa;
})(CompareFacesMatch || (CompareFacesMatch = {}));
export var CompareFacesRequest;
(function (CompareFacesRequest) {
    function isa(o) {
        return __isa(o, "CompareFacesRequest");
    }
    CompareFacesRequest.isa = isa;
})(CompareFacesRequest || (CompareFacesRequest = {}));
export var CompareFacesResponse;
(function (CompareFacesResponse) {
    function isa(o) {
        return __isa(o, "CompareFacesResponse");
    }
    CompareFacesResponse.isa = isa;
})(CompareFacesResponse || (CompareFacesResponse = {}));
export var ComparedFace;
(function (ComparedFace) {
    function isa(o) {
        return __isa(o, "ComparedFace");
    }
    ComparedFace.isa = isa;
})(ComparedFace || (ComparedFace = {}));
export var ComparedSourceImageFace;
(function (ComparedSourceImageFace) {
    function isa(o) {
        return __isa(o, "ComparedSourceImageFace");
    }
    ComparedSourceImageFace.isa = isa;
})(ComparedSourceImageFace || (ComparedSourceImageFace = {}));
export var ContentClassifier;
(function (ContentClassifier) {
    ContentClassifier["FREE_OF_ADULT_CONTENT"] = "FreeOfAdultContent";
    ContentClassifier["FREE_OF_PERSONALLY_IDENTIFIABLE_INFORMATION"] = "FreeOfPersonallyIdentifiableInformation";
})(ContentClassifier || (ContentClassifier = {}));
export var ContentModerationDetection;
(function (ContentModerationDetection) {
    function isa(o) {
        return __isa(o, "ContentModerationDetection");
    }
    ContentModerationDetection.isa = isa;
})(ContentModerationDetection || (ContentModerationDetection = {}));
export var ContentModerationSortBy;
(function (ContentModerationSortBy) {
    ContentModerationSortBy["NAME"] = "NAME";
    ContentModerationSortBy["TIMESTAMP"] = "TIMESTAMP";
})(ContentModerationSortBy || (ContentModerationSortBy = {}));
export var CreateCollectionRequest;
(function (CreateCollectionRequest) {
    function isa(o) {
        return __isa(o, "CreateCollectionRequest");
    }
    CreateCollectionRequest.isa = isa;
})(CreateCollectionRequest || (CreateCollectionRequest = {}));
export var CreateCollectionResponse;
(function (CreateCollectionResponse) {
    function isa(o) {
        return __isa(o, "CreateCollectionResponse");
    }
    CreateCollectionResponse.isa = isa;
})(CreateCollectionResponse || (CreateCollectionResponse = {}));
export var CreateProjectRequest;
(function (CreateProjectRequest) {
    function isa(o) {
        return __isa(o, "CreateProjectRequest");
    }
    CreateProjectRequest.isa = isa;
})(CreateProjectRequest || (CreateProjectRequest = {}));
export var CreateProjectResponse;
(function (CreateProjectResponse) {
    function isa(o) {
        return __isa(o, "CreateProjectResponse");
    }
    CreateProjectResponse.isa = isa;
})(CreateProjectResponse || (CreateProjectResponse = {}));
export var CreateProjectVersionRequest;
(function (CreateProjectVersionRequest) {
    function isa(o) {
        return __isa(o, "CreateProjectVersionRequest");
    }
    CreateProjectVersionRequest.isa = isa;
})(CreateProjectVersionRequest || (CreateProjectVersionRequest = {}));
export var CreateProjectVersionResponse;
(function (CreateProjectVersionResponse) {
    function isa(o) {
        return __isa(o, "CreateProjectVersionResponse");
    }
    CreateProjectVersionResponse.isa = isa;
})(CreateProjectVersionResponse || (CreateProjectVersionResponse = {}));
export var CreateStreamProcessorRequest;
(function (CreateStreamProcessorRequest) {
    function isa(o) {
        return __isa(o, "CreateStreamProcessorRequest");
    }
    CreateStreamProcessorRequest.isa = isa;
})(CreateStreamProcessorRequest || (CreateStreamProcessorRequest = {}));
export var CreateStreamProcessorResponse;
(function (CreateStreamProcessorResponse) {
    function isa(o) {
        return __isa(o, "CreateStreamProcessorResponse");
    }
    CreateStreamProcessorResponse.isa = isa;
})(CreateStreamProcessorResponse || (CreateStreamProcessorResponse = {}));
export var CustomLabel;
(function (CustomLabel) {
    function isa(o) {
        return __isa(o, "CustomLabel");
    }
    CustomLabel.isa = isa;
})(CustomLabel || (CustomLabel = {}));
export var DeleteCollectionRequest;
(function (DeleteCollectionRequest) {
    function isa(o) {
        return __isa(o, "DeleteCollectionRequest");
    }
    DeleteCollectionRequest.isa = isa;
})(DeleteCollectionRequest || (DeleteCollectionRequest = {}));
export var DeleteCollectionResponse;
(function (DeleteCollectionResponse) {
    function isa(o) {
        return __isa(o, "DeleteCollectionResponse");
    }
    DeleteCollectionResponse.isa = isa;
})(DeleteCollectionResponse || (DeleteCollectionResponse = {}));
export var DeleteFacesRequest;
(function (DeleteFacesRequest) {
    function isa(o) {
        return __isa(o, "DeleteFacesRequest");
    }
    DeleteFacesRequest.isa = isa;
})(DeleteFacesRequest || (DeleteFacesRequest = {}));
export var DeleteFacesResponse;
(function (DeleteFacesResponse) {
    function isa(o) {
        return __isa(o, "DeleteFacesResponse");
    }
    DeleteFacesResponse.isa = isa;
})(DeleteFacesResponse || (DeleteFacesResponse = {}));
export var DeleteStreamProcessorRequest;
(function (DeleteStreamProcessorRequest) {
    function isa(o) {
        return __isa(o, "DeleteStreamProcessorRequest");
    }
    DeleteStreamProcessorRequest.isa = isa;
})(DeleteStreamProcessorRequest || (DeleteStreamProcessorRequest = {}));
export var DeleteStreamProcessorResponse;
(function (DeleteStreamProcessorResponse) {
    function isa(o) {
        return __isa(o, "DeleteStreamProcessorResponse");
    }
    DeleteStreamProcessorResponse.isa = isa;
})(DeleteStreamProcessorResponse || (DeleteStreamProcessorResponse = {}));
export var DescribeCollectionRequest;
(function (DescribeCollectionRequest) {
    function isa(o) {
        return __isa(o, "DescribeCollectionRequest");
    }
    DescribeCollectionRequest.isa = isa;
})(DescribeCollectionRequest || (DescribeCollectionRequest = {}));
export var DescribeCollectionResponse;
(function (DescribeCollectionResponse) {
    function isa(o) {
        return __isa(o, "DescribeCollectionResponse");
    }
    DescribeCollectionResponse.isa = isa;
})(DescribeCollectionResponse || (DescribeCollectionResponse = {}));
export var DescribeProjectVersionsRequest;
(function (DescribeProjectVersionsRequest) {
    function isa(o) {
        return __isa(o, "DescribeProjectVersionsRequest");
    }
    DescribeProjectVersionsRequest.isa = isa;
})(DescribeProjectVersionsRequest || (DescribeProjectVersionsRequest = {}));
export var DescribeProjectVersionsResponse;
(function (DescribeProjectVersionsResponse) {
    function isa(o) {
        return __isa(o, "DescribeProjectVersionsResponse");
    }
    DescribeProjectVersionsResponse.isa = isa;
})(DescribeProjectVersionsResponse || (DescribeProjectVersionsResponse = {}));
export var DescribeProjectsRequest;
(function (DescribeProjectsRequest) {
    function isa(o) {
        return __isa(o, "DescribeProjectsRequest");
    }
    DescribeProjectsRequest.isa = isa;
})(DescribeProjectsRequest || (DescribeProjectsRequest = {}));
export var DescribeProjectsResponse;
(function (DescribeProjectsResponse) {
    function isa(o) {
        return __isa(o, "DescribeProjectsResponse");
    }
    DescribeProjectsResponse.isa = isa;
})(DescribeProjectsResponse || (DescribeProjectsResponse = {}));
export var DescribeStreamProcessorRequest;
(function (DescribeStreamProcessorRequest) {
    function isa(o) {
        return __isa(o, "DescribeStreamProcessorRequest");
    }
    DescribeStreamProcessorRequest.isa = isa;
})(DescribeStreamProcessorRequest || (DescribeStreamProcessorRequest = {}));
export var DescribeStreamProcessorResponse;
(function (DescribeStreamProcessorResponse) {
    function isa(o) {
        return __isa(o, "DescribeStreamProcessorResponse");
    }
    DescribeStreamProcessorResponse.isa = isa;
})(DescribeStreamProcessorResponse || (DescribeStreamProcessorResponse = {}));
export var DetectCustomLabelsRequest;
(function (DetectCustomLabelsRequest) {
    function isa(o) {
        return __isa(o, "DetectCustomLabelsRequest");
    }
    DetectCustomLabelsRequest.isa = isa;
})(DetectCustomLabelsRequest || (DetectCustomLabelsRequest = {}));
export var DetectCustomLabelsResponse;
(function (DetectCustomLabelsResponse) {
    function isa(o) {
        return __isa(o, "DetectCustomLabelsResponse");
    }
    DetectCustomLabelsResponse.isa = isa;
})(DetectCustomLabelsResponse || (DetectCustomLabelsResponse = {}));
export var DetectFacesRequest;
(function (DetectFacesRequest) {
    function isa(o) {
        return __isa(o, "DetectFacesRequest");
    }
    DetectFacesRequest.isa = isa;
})(DetectFacesRequest || (DetectFacesRequest = {}));
export var DetectFacesResponse;
(function (DetectFacesResponse) {
    function isa(o) {
        return __isa(o, "DetectFacesResponse");
    }
    DetectFacesResponse.isa = isa;
})(DetectFacesResponse || (DetectFacesResponse = {}));
export var DetectLabelsRequest;
(function (DetectLabelsRequest) {
    function isa(o) {
        return __isa(o, "DetectLabelsRequest");
    }
    DetectLabelsRequest.isa = isa;
})(DetectLabelsRequest || (DetectLabelsRequest = {}));
export var DetectLabelsResponse;
(function (DetectLabelsResponse) {
    function isa(o) {
        return __isa(o, "DetectLabelsResponse");
    }
    DetectLabelsResponse.isa = isa;
})(DetectLabelsResponse || (DetectLabelsResponse = {}));
export var DetectModerationLabelsRequest;
(function (DetectModerationLabelsRequest) {
    function isa(o) {
        return __isa(o, "DetectModerationLabelsRequest");
    }
    DetectModerationLabelsRequest.isa = isa;
})(DetectModerationLabelsRequest || (DetectModerationLabelsRequest = {}));
export var DetectModerationLabelsResponse;
(function (DetectModerationLabelsResponse) {
    function isa(o) {
        return __isa(o, "DetectModerationLabelsResponse");
    }
    DetectModerationLabelsResponse.isa = isa;
})(DetectModerationLabelsResponse || (DetectModerationLabelsResponse = {}));
export var DetectTextRequest;
(function (DetectTextRequest) {
    function isa(o) {
        return __isa(o, "DetectTextRequest");
    }
    DetectTextRequest.isa = isa;
})(DetectTextRequest || (DetectTextRequest = {}));
export var DetectTextResponse;
(function (DetectTextResponse) {
    function isa(o) {
        return __isa(o, "DetectTextResponse");
    }
    DetectTextResponse.isa = isa;
})(DetectTextResponse || (DetectTextResponse = {}));
export var Emotion;
(function (Emotion) {
    function isa(o) {
        return __isa(o, "Emotion");
    }
    Emotion.isa = isa;
})(Emotion || (Emotion = {}));
export var EvaluationResult;
(function (EvaluationResult) {
    function isa(o) {
        return __isa(o, "EvaluationResult");
    }
    EvaluationResult.isa = isa;
})(EvaluationResult || (EvaluationResult = {}));
export var EyeOpen;
(function (EyeOpen) {
    function isa(o) {
        return __isa(o, "EyeOpen");
    }
    EyeOpen.isa = isa;
})(EyeOpen || (EyeOpen = {}));
export var Eyeglasses;
(function (Eyeglasses) {
    function isa(o) {
        return __isa(o, "Eyeglasses");
    }
    Eyeglasses.isa = isa;
})(Eyeglasses || (Eyeglasses = {}));
export var Face;
(function (Face) {
    function isa(o) {
        return __isa(o, "Face");
    }
    Face.isa = isa;
})(Face || (Face = {}));
export var FaceAttributes;
(function (FaceAttributes) {
    FaceAttributes["ALL"] = "ALL";
    FaceAttributes["DEFAULT"] = "DEFAULT";
})(FaceAttributes || (FaceAttributes = {}));
export var FaceDetail;
(function (FaceDetail) {
    function isa(o) {
        return __isa(o, "FaceDetail");
    }
    FaceDetail.isa = isa;
})(FaceDetail || (FaceDetail = {}));
export var FaceDetection;
(function (FaceDetection) {
    function isa(o) {
        return __isa(o, "FaceDetection");
    }
    FaceDetection.isa = isa;
})(FaceDetection || (FaceDetection = {}));
export var FaceMatch;
(function (FaceMatch) {
    function isa(o) {
        return __isa(o, "FaceMatch");
    }
    FaceMatch.isa = isa;
})(FaceMatch || (FaceMatch = {}));
export var FaceRecord;
(function (FaceRecord) {
    function isa(o) {
        return __isa(o, "FaceRecord");
    }
    FaceRecord.isa = isa;
})(FaceRecord || (FaceRecord = {}));
export var FaceSearchSettings;
(function (FaceSearchSettings) {
    function isa(o) {
        return __isa(o, "FaceSearchSettings");
    }
    FaceSearchSettings.isa = isa;
})(FaceSearchSettings || (FaceSearchSettings = {}));
export var FaceSearchSortBy;
(function (FaceSearchSortBy) {
    FaceSearchSortBy["INDEX"] = "INDEX";
    FaceSearchSortBy["TIMESTAMP"] = "TIMESTAMP";
})(FaceSearchSortBy || (FaceSearchSortBy = {}));
export var Gender;
(function (Gender) {
    function isa(o) {
        return __isa(o, "Gender");
    }
    Gender.isa = isa;
})(Gender || (Gender = {}));
export var GenderType;
(function (GenderType) {
    GenderType["Female"] = "Female";
    GenderType["Male"] = "Male";
})(GenderType || (GenderType = {}));
export var Geometry;
(function (Geometry) {
    function isa(o) {
        return __isa(o, "Geometry");
    }
    Geometry.isa = isa;
})(Geometry || (Geometry = {}));
export var GetCelebrityInfoRequest;
(function (GetCelebrityInfoRequest) {
    function isa(o) {
        return __isa(o, "GetCelebrityInfoRequest");
    }
    GetCelebrityInfoRequest.isa = isa;
})(GetCelebrityInfoRequest || (GetCelebrityInfoRequest = {}));
export var GetCelebrityInfoResponse;
(function (GetCelebrityInfoResponse) {
    function isa(o) {
        return __isa(o, "GetCelebrityInfoResponse");
    }
    GetCelebrityInfoResponse.isa = isa;
})(GetCelebrityInfoResponse || (GetCelebrityInfoResponse = {}));
export var GetCelebrityRecognitionRequest;
(function (GetCelebrityRecognitionRequest) {
    function isa(o) {
        return __isa(o, "GetCelebrityRecognitionRequest");
    }
    GetCelebrityRecognitionRequest.isa = isa;
})(GetCelebrityRecognitionRequest || (GetCelebrityRecognitionRequest = {}));
export var GetCelebrityRecognitionResponse;
(function (GetCelebrityRecognitionResponse) {
    function isa(o) {
        return __isa(o, "GetCelebrityRecognitionResponse");
    }
    GetCelebrityRecognitionResponse.isa = isa;
})(GetCelebrityRecognitionResponse || (GetCelebrityRecognitionResponse = {}));
export var GetContentModerationRequest;
(function (GetContentModerationRequest) {
    function isa(o) {
        return __isa(o, "GetContentModerationRequest");
    }
    GetContentModerationRequest.isa = isa;
})(GetContentModerationRequest || (GetContentModerationRequest = {}));
export var GetContentModerationResponse;
(function (GetContentModerationResponse) {
    function isa(o) {
        return __isa(o, "GetContentModerationResponse");
    }
    GetContentModerationResponse.isa = isa;
})(GetContentModerationResponse || (GetContentModerationResponse = {}));
export var GetFaceDetectionRequest;
(function (GetFaceDetectionRequest) {
    function isa(o) {
        return __isa(o, "GetFaceDetectionRequest");
    }
    GetFaceDetectionRequest.isa = isa;
})(GetFaceDetectionRequest || (GetFaceDetectionRequest = {}));
export var GetFaceDetectionResponse;
(function (GetFaceDetectionResponse) {
    function isa(o) {
        return __isa(o, "GetFaceDetectionResponse");
    }
    GetFaceDetectionResponse.isa = isa;
})(GetFaceDetectionResponse || (GetFaceDetectionResponse = {}));
export var GetFaceSearchRequest;
(function (GetFaceSearchRequest) {
    function isa(o) {
        return __isa(o, "GetFaceSearchRequest");
    }
    GetFaceSearchRequest.isa = isa;
})(GetFaceSearchRequest || (GetFaceSearchRequest = {}));
export var GetFaceSearchResponse;
(function (GetFaceSearchResponse) {
    function isa(o) {
        return __isa(o, "GetFaceSearchResponse");
    }
    GetFaceSearchResponse.isa = isa;
})(GetFaceSearchResponse || (GetFaceSearchResponse = {}));
export var GetLabelDetectionRequest;
(function (GetLabelDetectionRequest) {
    function isa(o) {
        return __isa(o, "GetLabelDetectionRequest");
    }
    GetLabelDetectionRequest.isa = isa;
})(GetLabelDetectionRequest || (GetLabelDetectionRequest = {}));
export var GetLabelDetectionResponse;
(function (GetLabelDetectionResponse) {
    function isa(o) {
        return __isa(o, "GetLabelDetectionResponse");
    }
    GetLabelDetectionResponse.isa = isa;
})(GetLabelDetectionResponse || (GetLabelDetectionResponse = {}));
export var GetPersonTrackingRequest;
(function (GetPersonTrackingRequest) {
    function isa(o) {
        return __isa(o, "GetPersonTrackingRequest");
    }
    GetPersonTrackingRequest.isa = isa;
})(GetPersonTrackingRequest || (GetPersonTrackingRequest = {}));
export var GetPersonTrackingResponse;
(function (GetPersonTrackingResponse) {
    function isa(o) {
        return __isa(o, "GetPersonTrackingResponse");
    }
    GetPersonTrackingResponse.isa = isa;
})(GetPersonTrackingResponse || (GetPersonTrackingResponse = {}));
export var GroundTruthManifest;
(function (GroundTruthManifest) {
    function isa(o) {
        return __isa(o, "GroundTruthManifest");
    }
    GroundTruthManifest.isa = isa;
})(GroundTruthManifest || (GroundTruthManifest = {}));
export var HumanLoopActivationOutput;
(function (HumanLoopActivationOutput) {
    function isa(o) {
        return __isa(o, "HumanLoopActivationOutput");
    }
    HumanLoopActivationOutput.isa = isa;
})(HumanLoopActivationOutput || (HumanLoopActivationOutput = {}));
export var HumanLoopConfig;
(function (HumanLoopConfig) {
    function isa(o) {
        return __isa(o, "HumanLoopConfig");
    }
    HumanLoopConfig.isa = isa;
})(HumanLoopConfig || (HumanLoopConfig = {}));
export var HumanLoopDataAttributes;
(function (HumanLoopDataAttributes) {
    function isa(o) {
        return __isa(o, "HumanLoopDataAttributes");
    }
    HumanLoopDataAttributes.isa = isa;
})(HumanLoopDataAttributes || (HumanLoopDataAttributes = {}));
export var HumanLoopQuotaExceededException;
(function (HumanLoopQuotaExceededException) {
    function isa(o) {
        return __isa(o, "HumanLoopQuotaExceededException");
    }
    HumanLoopQuotaExceededException.isa = isa;
})(HumanLoopQuotaExceededException || (HumanLoopQuotaExceededException = {}));
export var IdempotentParameterMismatchException;
(function (IdempotentParameterMismatchException) {
    function isa(o) {
        return __isa(o, "IdempotentParameterMismatchException");
    }
    IdempotentParameterMismatchException.isa = isa;
})(IdempotentParameterMismatchException || (IdempotentParameterMismatchException = {}));
export var Image;
(function (Image) {
    function isa(o) {
        return __isa(o, "Image");
    }
    Image.isa = isa;
})(Image || (Image = {}));
export var ImageQuality;
(function (ImageQuality) {
    function isa(o) {
        return __isa(o, "ImageQuality");
    }
    ImageQuality.isa = isa;
})(ImageQuality || (ImageQuality = {}));
export var ImageTooLargeException;
(function (ImageTooLargeException) {
    function isa(o) {
        return __isa(o, "ImageTooLargeException");
    }
    ImageTooLargeException.isa = isa;
})(ImageTooLargeException || (ImageTooLargeException = {}));
export var IndexFacesRequest;
(function (IndexFacesRequest) {
    function isa(o) {
        return __isa(o, "IndexFacesRequest");
    }
    IndexFacesRequest.isa = isa;
})(IndexFacesRequest || (IndexFacesRequest = {}));
export var IndexFacesResponse;
(function (IndexFacesResponse) {
    function isa(o) {
        return __isa(o, "IndexFacesResponse");
    }
    IndexFacesResponse.isa = isa;
})(IndexFacesResponse || (IndexFacesResponse = {}));
export var Instance;
(function (Instance) {
    function isa(o) {
        return __isa(o, "Instance");
    }
    Instance.isa = isa;
})(Instance || (Instance = {}));
export var InternalServerError;
(function (InternalServerError) {
    function isa(o) {
        return __isa(o, "InternalServerError");
    }
    InternalServerError.isa = isa;
})(InternalServerError || (InternalServerError = {}));
export var InvalidImageFormatException;
(function (InvalidImageFormatException) {
    function isa(o) {
        return __isa(o, "InvalidImageFormatException");
    }
    InvalidImageFormatException.isa = isa;
})(InvalidImageFormatException || (InvalidImageFormatException = {}));
export var InvalidPaginationTokenException;
(function (InvalidPaginationTokenException) {
    function isa(o) {
        return __isa(o, "InvalidPaginationTokenException");
    }
    InvalidPaginationTokenException.isa = isa;
})(InvalidPaginationTokenException || (InvalidPaginationTokenException = {}));
export var InvalidParameterException;
(function (InvalidParameterException) {
    function isa(o) {
        return __isa(o, "InvalidParameterException");
    }
    InvalidParameterException.isa = isa;
})(InvalidParameterException || (InvalidParameterException = {}));
export var InvalidS3ObjectException;
(function (InvalidS3ObjectException) {
    function isa(o) {
        return __isa(o, "InvalidS3ObjectException");
    }
    InvalidS3ObjectException.isa = isa;
})(InvalidS3ObjectException || (InvalidS3ObjectException = {}));
export var KinesisDataStream;
(function (KinesisDataStream) {
    function isa(o) {
        return __isa(o, "KinesisDataStream");
    }
    KinesisDataStream.isa = isa;
})(KinesisDataStream || (KinesisDataStream = {}));
export var KinesisVideoStream;
(function (KinesisVideoStream) {
    function isa(o) {
        return __isa(o, "KinesisVideoStream");
    }
    KinesisVideoStream.isa = isa;
})(KinesisVideoStream || (KinesisVideoStream = {}));
export var Label;
(function (Label) {
    function isa(o) {
        return __isa(o, "Label");
    }
    Label.isa = isa;
})(Label || (Label = {}));
export var LabelDetection;
(function (LabelDetection) {
    function isa(o) {
        return __isa(o, "LabelDetection");
    }
    LabelDetection.isa = isa;
})(LabelDetection || (LabelDetection = {}));
export var LabelDetectionSortBy;
(function (LabelDetectionSortBy) {
    LabelDetectionSortBy["NAME"] = "NAME";
    LabelDetectionSortBy["TIMESTAMP"] = "TIMESTAMP";
})(LabelDetectionSortBy || (LabelDetectionSortBy = {}));
export var Landmark;
(function (Landmark) {
    function isa(o) {
        return __isa(o, "Landmark");
    }
    Landmark.isa = isa;
})(Landmark || (Landmark = {}));
export var LandmarkType;
(function (LandmarkType) {
    LandmarkType["chinBottom"] = "chinBottom";
    LandmarkType["eyeLeft"] = "eyeLeft";
    LandmarkType["eyeRight"] = "eyeRight";
    LandmarkType["leftEyeBrowLeft"] = "leftEyeBrowLeft";
    LandmarkType["leftEyeBrowRight"] = "leftEyeBrowRight";
    LandmarkType["leftEyeBrowUp"] = "leftEyeBrowUp";
    LandmarkType["leftEyeDown"] = "leftEyeDown";
    LandmarkType["leftEyeLeft"] = "leftEyeLeft";
    LandmarkType["leftEyeRight"] = "leftEyeRight";
    LandmarkType["leftEyeUp"] = "leftEyeUp";
    LandmarkType["leftPupil"] = "leftPupil";
    LandmarkType["midJawlineLeft"] = "midJawlineLeft";
    LandmarkType["midJawlineRight"] = "midJawlineRight";
    LandmarkType["mouthDown"] = "mouthDown";
    LandmarkType["mouthLeft"] = "mouthLeft";
    LandmarkType["mouthRight"] = "mouthRight";
    LandmarkType["mouthUp"] = "mouthUp";
    LandmarkType["nose"] = "nose";
    LandmarkType["noseLeft"] = "noseLeft";
    LandmarkType["noseRight"] = "noseRight";
    LandmarkType["rightEyeBrowLeft"] = "rightEyeBrowLeft";
    LandmarkType["rightEyeBrowRight"] = "rightEyeBrowRight";
    LandmarkType["rightEyeBrowUp"] = "rightEyeBrowUp";
    LandmarkType["rightEyeDown"] = "rightEyeDown";
    LandmarkType["rightEyeLeft"] = "rightEyeLeft";
    LandmarkType["rightEyeRight"] = "rightEyeRight";
    LandmarkType["rightEyeUp"] = "rightEyeUp";
    LandmarkType["rightPupil"] = "rightPupil";
    LandmarkType["upperJawlineLeft"] = "upperJawlineLeft";
    LandmarkType["upperJawlineRight"] = "upperJawlineRight";
})(LandmarkType || (LandmarkType = {}));
export var LimitExceededException;
(function (LimitExceededException) {
    function isa(o) {
        return __isa(o, "LimitExceededException");
    }
    LimitExceededException.isa = isa;
})(LimitExceededException || (LimitExceededException = {}));
export var ListCollectionsRequest;
(function (ListCollectionsRequest) {
    function isa(o) {
        return __isa(o, "ListCollectionsRequest");
    }
    ListCollectionsRequest.isa = isa;
})(ListCollectionsRequest || (ListCollectionsRequest = {}));
export var ListCollectionsResponse;
(function (ListCollectionsResponse) {
    function isa(o) {
        return __isa(o, "ListCollectionsResponse");
    }
    ListCollectionsResponse.isa = isa;
})(ListCollectionsResponse || (ListCollectionsResponse = {}));
export var ListFacesRequest;
(function (ListFacesRequest) {
    function isa(o) {
        return __isa(o, "ListFacesRequest");
    }
    ListFacesRequest.isa = isa;
})(ListFacesRequest || (ListFacesRequest = {}));
export var ListFacesResponse;
(function (ListFacesResponse) {
    function isa(o) {
        return __isa(o, "ListFacesResponse");
    }
    ListFacesResponse.isa = isa;
})(ListFacesResponse || (ListFacesResponse = {}));
export var ListStreamProcessorsRequest;
(function (ListStreamProcessorsRequest) {
    function isa(o) {
        return __isa(o, "ListStreamProcessorsRequest");
    }
    ListStreamProcessorsRequest.isa = isa;
})(ListStreamProcessorsRequest || (ListStreamProcessorsRequest = {}));
export var ListStreamProcessorsResponse;
(function (ListStreamProcessorsResponse) {
    function isa(o) {
        return __isa(o, "ListStreamProcessorsResponse");
    }
    ListStreamProcessorsResponse.isa = isa;
})(ListStreamProcessorsResponse || (ListStreamProcessorsResponse = {}));
export var ModerationLabel;
(function (ModerationLabel) {
    function isa(o) {
        return __isa(o, "ModerationLabel");
    }
    ModerationLabel.isa = isa;
})(ModerationLabel || (ModerationLabel = {}));
export var MouthOpen;
(function (MouthOpen) {
    function isa(o) {
        return __isa(o, "MouthOpen");
    }
    MouthOpen.isa = isa;
})(MouthOpen || (MouthOpen = {}));
export var Mustache;
(function (Mustache) {
    function isa(o) {
        return __isa(o, "Mustache");
    }
    Mustache.isa = isa;
})(Mustache || (Mustache = {}));
export var NotificationChannel;
(function (NotificationChannel) {
    function isa(o) {
        return __isa(o, "NotificationChannel");
    }
    NotificationChannel.isa = isa;
})(NotificationChannel || (NotificationChannel = {}));
export var OrientationCorrection;
(function (OrientationCorrection) {
    OrientationCorrection["ROTATE_0"] = "ROTATE_0";
    OrientationCorrection["ROTATE_180"] = "ROTATE_180";
    OrientationCorrection["ROTATE_270"] = "ROTATE_270";
    OrientationCorrection["ROTATE_90"] = "ROTATE_90";
})(OrientationCorrection || (OrientationCorrection = {}));
export var OutputConfig;
(function (OutputConfig) {
    function isa(o) {
        return __isa(o, "OutputConfig");
    }
    OutputConfig.isa = isa;
})(OutputConfig || (OutputConfig = {}));
export var Parent;
(function (Parent) {
    function isa(o) {
        return __isa(o, "Parent");
    }
    Parent.isa = isa;
})(Parent || (Parent = {}));
export var PersonDetail;
(function (PersonDetail) {
    function isa(o) {
        return __isa(o, "PersonDetail");
    }
    PersonDetail.isa = isa;
})(PersonDetail || (PersonDetail = {}));
export var PersonDetection;
(function (PersonDetection) {
    function isa(o) {
        return __isa(o, "PersonDetection");
    }
    PersonDetection.isa = isa;
})(PersonDetection || (PersonDetection = {}));
export var PersonMatch;
(function (PersonMatch) {
    function isa(o) {
        return __isa(o, "PersonMatch");
    }
    PersonMatch.isa = isa;
})(PersonMatch || (PersonMatch = {}));
export var PersonTrackingSortBy;
(function (PersonTrackingSortBy) {
    PersonTrackingSortBy["INDEX"] = "INDEX";
    PersonTrackingSortBy["TIMESTAMP"] = "TIMESTAMP";
})(PersonTrackingSortBy || (PersonTrackingSortBy = {}));
export var Point;
(function (Point) {
    function isa(o) {
        return __isa(o, "Point");
    }
    Point.isa = isa;
})(Point || (Point = {}));
export var Pose;
(function (Pose) {
    function isa(o) {
        return __isa(o, "Pose");
    }
    Pose.isa = isa;
})(Pose || (Pose = {}));
export var ProjectDescription;
(function (ProjectDescription) {
    function isa(o) {
        return __isa(o, "ProjectDescription");
    }
    ProjectDescription.isa = isa;
})(ProjectDescription || (ProjectDescription = {}));
export var ProjectStatus;
(function (ProjectStatus) {
    ProjectStatus["CREATED"] = "CREATED";
    ProjectStatus["CREATING"] = "CREATING";
    ProjectStatus["DELETING"] = "DELETING";
})(ProjectStatus || (ProjectStatus = {}));
export var ProjectVersionDescription;
(function (ProjectVersionDescription) {
    function isa(o) {
        return __isa(o, "ProjectVersionDescription");
    }
    ProjectVersionDescription.isa = isa;
})(ProjectVersionDescription || (ProjectVersionDescription = {}));
export var ProjectVersionStatus;
(function (ProjectVersionStatus) {
    ProjectVersionStatus["DELETING"] = "DELETING";
    ProjectVersionStatus["FAILED"] = "FAILED";
    ProjectVersionStatus["RUNNING"] = "RUNNING";
    ProjectVersionStatus["STARTING"] = "STARTING";
    ProjectVersionStatus["STOPPED"] = "STOPPED";
    ProjectVersionStatus["STOPPING"] = "STOPPING";
    ProjectVersionStatus["TRAINING_COMPLETED"] = "TRAINING_COMPLETED";
    ProjectVersionStatus["TRAINING_FAILED"] = "TRAINING_FAILED";
    ProjectVersionStatus["TRAINING_IN_PROGRESS"] = "TRAINING_IN_PROGRESS";
})(ProjectVersionStatus || (ProjectVersionStatus = {}));
export var ProvisionedThroughputExceededException;
(function (ProvisionedThroughputExceededException) {
    function isa(o) {
        return __isa(o, "ProvisionedThroughputExceededException");
    }
    ProvisionedThroughputExceededException.isa = isa;
})(ProvisionedThroughputExceededException || (ProvisionedThroughputExceededException = {}));
export var QualityFilter;
(function (QualityFilter) {
    QualityFilter["AUTO"] = "AUTO";
    QualityFilter["HIGH"] = "HIGH";
    QualityFilter["LOW"] = "LOW";
    QualityFilter["MEDIUM"] = "MEDIUM";
    QualityFilter["NONE"] = "NONE";
})(QualityFilter || (QualityFilter = {}));
export var Reason;
(function (Reason) {
    Reason["EXCEEDS_MAX_FACES"] = "EXCEEDS_MAX_FACES";
    Reason["EXTREME_POSE"] = "EXTREME_POSE";
    Reason["LOW_BRIGHTNESS"] = "LOW_BRIGHTNESS";
    Reason["LOW_CONFIDENCE"] = "LOW_CONFIDENCE";
    Reason["LOW_FACE_QUALITY"] = "LOW_FACE_QUALITY";
    Reason["LOW_SHARPNESS"] = "LOW_SHARPNESS";
    Reason["SMALL_BOUNDING_BOX"] = "SMALL_BOUNDING_BOX";
})(Reason || (Reason = {}));
export var RecognizeCelebritiesRequest;
(function (RecognizeCelebritiesRequest) {
    function isa(o) {
        return __isa(o, "RecognizeCelebritiesRequest");
    }
    RecognizeCelebritiesRequest.isa = isa;
})(RecognizeCelebritiesRequest || (RecognizeCelebritiesRequest = {}));
export var RecognizeCelebritiesResponse;
(function (RecognizeCelebritiesResponse) {
    function isa(o) {
        return __isa(o, "RecognizeCelebritiesResponse");
    }
    RecognizeCelebritiesResponse.isa = isa;
})(RecognizeCelebritiesResponse || (RecognizeCelebritiesResponse = {}));
export var ResourceAlreadyExistsException;
(function (ResourceAlreadyExistsException) {
    function isa(o) {
        return __isa(o, "ResourceAlreadyExistsException");
    }
    ResourceAlreadyExistsException.isa = isa;
})(ResourceAlreadyExistsException || (ResourceAlreadyExistsException = {}));
export var ResourceInUseException;
(function (ResourceInUseException) {
    function isa(o) {
        return __isa(o, "ResourceInUseException");
    }
    ResourceInUseException.isa = isa;
})(ResourceInUseException || (ResourceInUseException = {}));
export var ResourceNotFoundException;
(function (ResourceNotFoundException) {
    function isa(o) {
        return __isa(o, "ResourceNotFoundException");
    }
    ResourceNotFoundException.isa = isa;
})(ResourceNotFoundException || (ResourceNotFoundException = {}));
export var ResourceNotReadyException;
(function (ResourceNotReadyException) {
    function isa(o) {
        return __isa(o, "ResourceNotReadyException");
    }
    ResourceNotReadyException.isa = isa;
})(ResourceNotReadyException || (ResourceNotReadyException = {}));
export var S3Object;
(function (S3Object) {
    function isa(o) {
        return __isa(o, "S3Object");
    }
    S3Object.isa = isa;
})(S3Object || (S3Object = {}));
export var SearchFacesByImageRequest;
(function (SearchFacesByImageRequest) {
    function isa(o) {
        return __isa(o, "SearchFacesByImageRequest");
    }
    SearchFacesByImageRequest.isa = isa;
})(SearchFacesByImageRequest || (SearchFacesByImageRequest = {}));
export var SearchFacesByImageResponse;
(function (SearchFacesByImageResponse) {
    function isa(o) {
        return __isa(o, "SearchFacesByImageResponse");
    }
    SearchFacesByImageResponse.isa = isa;
})(SearchFacesByImageResponse || (SearchFacesByImageResponse = {}));
export var SearchFacesRequest;
(function (SearchFacesRequest) {
    function isa(o) {
        return __isa(o, "SearchFacesRequest");
    }
    SearchFacesRequest.isa = isa;
})(SearchFacesRequest || (SearchFacesRequest = {}));
export var SearchFacesResponse;
(function (SearchFacesResponse) {
    function isa(o) {
        return __isa(o, "SearchFacesResponse");
    }
    SearchFacesResponse.isa = isa;
})(SearchFacesResponse || (SearchFacesResponse = {}));
export var Smile;
(function (Smile) {
    function isa(o) {
        return __isa(o, "Smile");
    }
    Smile.isa = isa;
})(Smile || (Smile = {}));
export var StartCelebrityRecognitionRequest;
(function (StartCelebrityRecognitionRequest) {
    function isa(o) {
        return __isa(o, "StartCelebrityRecognitionRequest");
    }
    StartCelebrityRecognitionRequest.isa = isa;
})(StartCelebrityRecognitionRequest || (StartCelebrityRecognitionRequest = {}));
export var StartCelebrityRecognitionResponse;
(function (StartCelebrityRecognitionResponse) {
    function isa(o) {
        return __isa(o, "StartCelebrityRecognitionResponse");
    }
    StartCelebrityRecognitionResponse.isa = isa;
})(StartCelebrityRecognitionResponse || (StartCelebrityRecognitionResponse = {}));
export var StartContentModerationRequest;
(function (StartContentModerationRequest) {
    function isa(o) {
        return __isa(o, "StartContentModerationRequest");
    }
    StartContentModerationRequest.isa = isa;
})(StartContentModerationRequest || (StartContentModerationRequest = {}));
export var StartContentModerationResponse;
(function (StartContentModerationResponse) {
    function isa(o) {
        return __isa(o, "StartContentModerationResponse");
    }
    StartContentModerationResponse.isa = isa;
})(StartContentModerationResponse || (StartContentModerationResponse = {}));
export var StartFaceDetectionRequest;
(function (StartFaceDetectionRequest) {
    function isa(o) {
        return __isa(o, "StartFaceDetectionRequest");
    }
    StartFaceDetectionRequest.isa = isa;
})(StartFaceDetectionRequest || (StartFaceDetectionRequest = {}));
export var StartFaceDetectionResponse;
(function (StartFaceDetectionResponse) {
    function isa(o) {
        return __isa(o, "StartFaceDetectionResponse");
    }
    StartFaceDetectionResponse.isa = isa;
})(StartFaceDetectionResponse || (StartFaceDetectionResponse = {}));
export var StartFaceSearchRequest;
(function (StartFaceSearchRequest) {
    function isa(o) {
        return __isa(o, "StartFaceSearchRequest");
    }
    StartFaceSearchRequest.isa = isa;
})(StartFaceSearchRequest || (StartFaceSearchRequest = {}));
export var StartFaceSearchResponse;
(function (StartFaceSearchResponse) {
    function isa(o) {
        return __isa(o, "StartFaceSearchResponse");
    }
    StartFaceSearchResponse.isa = isa;
})(StartFaceSearchResponse || (StartFaceSearchResponse = {}));
export var StartLabelDetectionRequest;
(function (StartLabelDetectionRequest) {
    function isa(o) {
        return __isa(o, "StartLabelDetectionRequest");
    }
    StartLabelDetectionRequest.isa = isa;
})(StartLabelDetectionRequest || (StartLabelDetectionRequest = {}));
export var StartLabelDetectionResponse;
(function (StartLabelDetectionResponse) {
    function isa(o) {
        return __isa(o, "StartLabelDetectionResponse");
    }
    StartLabelDetectionResponse.isa = isa;
})(StartLabelDetectionResponse || (StartLabelDetectionResponse = {}));
export var StartPersonTrackingRequest;
(function (StartPersonTrackingRequest) {
    function isa(o) {
        return __isa(o, "StartPersonTrackingRequest");
    }
    StartPersonTrackingRequest.isa = isa;
})(StartPersonTrackingRequest || (StartPersonTrackingRequest = {}));
export var StartPersonTrackingResponse;
(function (StartPersonTrackingResponse) {
    function isa(o) {
        return __isa(o, "StartPersonTrackingResponse");
    }
    StartPersonTrackingResponse.isa = isa;
})(StartPersonTrackingResponse || (StartPersonTrackingResponse = {}));
export var StartProjectVersionRequest;
(function (StartProjectVersionRequest) {
    function isa(o) {
        return __isa(o, "StartProjectVersionRequest");
    }
    StartProjectVersionRequest.isa = isa;
})(StartProjectVersionRequest || (StartProjectVersionRequest = {}));
export var StartProjectVersionResponse;
(function (StartProjectVersionResponse) {
    function isa(o) {
        return __isa(o, "StartProjectVersionResponse");
    }
    StartProjectVersionResponse.isa = isa;
})(StartProjectVersionResponse || (StartProjectVersionResponse = {}));
export var StartStreamProcessorRequest;
(function (StartStreamProcessorRequest) {
    function isa(o) {
        return __isa(o, "StartStreamProcessorRequest");
    }
    StartStreamProcessorRequest.isa = isa;
})(StartStreamProcessorRequest || (StartStreamProcessorRequest = {}));
export var StartStreamProcessorResponse;
(function (StartStreamProcessorResponse) {
    function isa(o) {
        return __isa(o, "StartStreamProcessorResponse");
    }
    StartStreamProcessorResponse.isa = isa;
})(StartStreamProcessorResponse || (StartStreamProcessorResponse = {}));
export var StopProjectVersionRequest;
(function (StopProjectVersionRequest) {
    function isa(o) {
        return __isa(o, "StopProjectVersionRequest");
    }
    StopProjectVersionRequest.isa = isa;
})(StopProjectVersionRequest || (StopProjectVersionRequest = {}));
export var StopProjectVersionResponse;
(function (StopProjectVersionResponse) {
    function isa(o) {
        return __isa(o, "StopProjectVersionResponse");
    }
    StopProjectVersionResponse.isa = isa;
})(StopProjectVersionResponse || (StopProjectVersionResponse = {}));
export var StopStreamProcessorRequest;
(function (StopStreamProcessorRequest) {
    function isa(o) {
        return __isa(o, "StopStreamProcessorRequest");
    }
    StopStreamProcessorRequest.isa = isa;
})(StopStreamProcessorRequest || (StopStreamProcessorRequest = {}));
export var StopStreamProcessorResponse;
(function (StopStreamProcessorResponse) {
    function isa(o) {
        return __isa(o, "StopStreamProcessorResponse");
    }
    StopStreamProcessorResponse.isa = isa;
})(StopStreamProcessorResponse || (StopStreamProcessorResponse = {}));
export var StreamProcessor;
(function (StreamProcessor) {
    function isa(o) {
        return __isa(o, "StreamProcessor");
    }
    StreamProcessor.isa = isa;
})(StreamProcessor || (StreamProcessor = {}));
export var StreamProcessorInput;
(function (StreamProcessorInput) {
    function isa(o) {
        return __isa(o, "StreamProcessorInput");
    }
    StreamProcessorInput.isa = isa;
})(StreamProcessorInput || (StreamProcessorInput = {}));
export var StreamProcessorOutput;
(function (StreamProcessorOutput) {
    function isa(o) {
        return __isa(o, "StreamProcessorOutput");
    }
    StreamProcessorOutput.isa = isa;
})(StreamProcessorOutput || (StreamProcessorOutput = {}));
export var StreamProcessorSettings;
(function (StreamProcessorSettings) {
    function isa(o) {
        return __isa(o, "StreamProcessorSettings");
    }
    StreamProcessorSettings.isa = isa;
})(StreamProcessorSettings || (StreamProcessorSettings = {}));
export var StreamProcessorStatus;
(function (StreamProcessorStatus) {
    StreamProcessorStatus["FAILED"] = "FAILED";
    StreamProcessorStatus["RUNNING"] = "RUNNING";
    StreamProcessorStatus["STARTING"] = "STARTING";
    StreamProcessorStatus["STOPPED"] = "STOPPED";
    StreamProcessorStatus["STOPPING"] = "STOPPING";
})(StreamProcessorStatus || (StreamProcessorStatus = {}));
export var Summary;
(function (Summary) {
    function isa(o) {
        return __isa(o, "Summary");
    }
    Summary.isa = isa;
})(Summary || (Summary = {}));
export var Sunglasses;
(function (Sunglasses) {
    function isa(o) {
        return __isa(o, "Sunglasses");
    }
    Sunglasses.isa = isa;
})(Sunglasses || (Sunglasses = {}));
export var TestingData;
(function (TestingData) {
    function isa(o) {
        return __isa(o, "TestingData");
    }
    TestingData.isa = isa;
})(TestingData || (TestingData = {}));
export var TestingDataResult;
(function (TestingDataResult) {
    function isa(o) {
        return __isa(o, "TestingDataResult");
    }
    TestingDataResult.isa = isa;
})(TestingDataResult || (TestingDataResult = {}));
export var TextDetection;
(function (TextDetection) {
    function isa(o) {
        return __isa(o, "TextDetection");
    }
    TextDetection.isa = isa;
})(TextDetection || (TextDetection = {}));
export var TextTypes;
(function (TextTypes) {
    TextTypes["LINE"] = "LINE";
    TextTypes["WORD"] = "WORD";
})(TextTypes || (TextTypes = {}));
export var ThrottlingException;
(function (ThrottlingException) {
    function isa(o) {
        return __isa(o, "ThrottlingException");
    }
    ThrottlingException.isa = isa;
})(ThrottlingException || (ThrottlingException = {}));
export var TrainingData;
(function (TrainingData) {
    function isa(o) {
        return __isa(o, "TrainingData");
    }
    TrainingData.isa = isa;
})(TrainingData || (TrainingData = {}));
export var TrainingDataResult;
(function (TrainingDataResult) {
    function isa(o) {
        return __isa(o, "TrainingDataResult");
    }
    TrainingDataResult.isa = isa;
})(TrainingDataResult || (TrainingDataResult = {}));
export var UnindexedFace;
(function (UnindexedFace) {
    function isa(o) {
        return __isa(o, "UnindexedFace");
    }
    UnindexedFace.isa = isa;
})(UnindexedFace || (UnindexedFace = {}));
export var Video;
(function (Video) {
    function isa(o) {
        return __isa(o, "Video");
    }
    Video.isa = isa;
})(Video || (Video = {}));
export var VideoJobStatus;
(function (VideoJobStatus) {
    VideoJobStatus["FAILED"] = "FAILED";
    VideoJobStatus["IN_PROGRESS"] = "IN_PROGRESS";
    VideoJobStatus["SUCCEEDED"] = "SUCCEEDED";
})(VideoJobStatus || (VideoJobStatus = {}));
export var VideoMetadata;
(function (VideoMetadata) {
    function isa(o) {
        return __isa(o, "VideoMetadata");
    }
    VideoMetadata.isa = isa;
})(VideoMetadata || (VideoMetadata = {}));
export var VideoTooLargeException;
(function (VideoTooLargeException) {
    function isa(o) {
        return __isa(o, "VideoTooLargeException");
    }
    VideoTooLargeException.isa = isa;
})(VideoTooLargeException || (VideoTooLargeException = {}));
